\documentclass{beamer}
%\documentclass[handout]{beamer}

% language settings
%\usepackage{fontspec, polyglossia}
%\setdefaultlanguage{magyar}

% common packages
\usepackage{amsmath, multimedia, hyperref, color, multirow}
%\usepackage{graphicx}

% TikZ
\usepackage{tikz}
%\usetikzlibrary{arrows.meta, decorations.pathmorphing, decorations.pathreplacing, shapes.geometric,mindmap}
%\usetikzlibrary{shapes.geometric,fadings,bayesnet}

% beamer styles
\mode<presentation>{
\usetheme{Warsaw}
%\usetheme{Antibes}
\usecolortheme{beaver}
%\usecolortheme{seahorse}
%\usefonttheme{structureitalicserif}
\setbeamercovered{transparent}
}
\setbeamertemplate{blocks}[rounded][shadow=true]
\AtBeginSubsection[]{
  \begin{frame}<beamer>{Contents}
    \tableofcontents[currentsection,currentsubsection]
  \end{frame}
}
%\useoutertheme[]{tree}

% title, etc
\title{Somatic SNV Calling Accuracy of 5+ Callers}
%\subtitle{A subtitle may be shorter and more technical}
\author{Attila Jones, Chaggai Rosenbluh, Andy Chess}
\date{Chess Lab, Mount Sinai School of Medicine}

\begin{document}

\maketitle

\begin{frame}
\begin{enumerate}
\item callsets by 5+ callers\\
differences:
\begin{itemize}
\item specificity and sensitivity
\item bias: technical + biol (germline)
\end{itemize}
\item integration by machine learning
\begin{itemize}
\item quality annotations as input features\\
VariantMetaCaller
\item training sets: concordant and discordant calls
\end{itemize}
\item application 
\begin{itemize}
\item somatic variant calling
\item performance (precision--recall)
\end{itemize}
\end{enumerate}
\end{frame}

\begin{frame}[label=precrecall]{Measuring performance of somatic variant calling}
%{Compare call set \(C\) to truth set \(V\)}
\begin{description}
\item [\(C\)] some call set of somatic variants
\item [\(V\)] the truth set of somatic variants
\begin{enumerate}
\item mixing experiment
\item experimental validation 
\end{enumerate}
\end{description}
\vfill
\alert{Precision}\\
the fraction of calls that are true
\begin{equation}
p(C) = \frac{|C \cap V|}{|C|} = \frac{\#\text{true calls}}{\#\mathrm{calls}}
\end{equation}
\alert{Recall}\\
the fraction of variants that are called
\begin{equation}
r(C) = \frac{|C \cap V|}{|V|} = \frac{\#\text{true calls}}{\#\mathrm{variants}}
\end{equation}
\end{frame}

\begin{frame}{Experimental validation}
\includegraphics[scale=0.6]{../../notebook/2019-07-23-commonsample-precrecall/named-figure/prec-recall-common-sample-mosaicforecast.pdf}
\end{frame}

\begin{frame}{Mixing experiment}
\begin{columns}[t]
\begin{column}{0.5\textwidth}

\includegraphics[scale=0.3]{../../notebook/2018-05-01-ceph-vars-greally/figure/venn-S1-1.pdf}
\end{column}

\begin{column}{0.5\textwidth}

{\footnotesize
{\large mixing proportions}

\begin{tabular}{r|ccc}
genome & mix1 & mix2 & mix3 \\
\hline
NA12889 & 4 & 2 & 0 \\
NA12891 & 8 & 4 & 0 \\
NA12890 & 16 & 8 & 0 \\
NA12892 & 72 & 86 & 100 \\
& & & \\
\end{tabular}
}
\end{column}
\end{columns}
\end{frame}

\begin{frame}{Allele frequency}
\begin{columns}[t]
\begin{column}{0.5\textwidth}
{\tiny
\begin{tabular}{c|ccc}
genotype & \multicolumn{3}{c}{AAF (\%)} \\
& mix1 &  mix2 & mix3 \\
\hline
0001 & 36 & 43 & 50\\
0010 & 8 & 4 & 0\\
0100 & 4 & 2 & 0\\
1000 & 2 & 1 & 0\\
0002 & 72 & 86 & 100\\
... & ... & ... & ...\\
1222 & 98 & 99 & 100\\
2122 & 96 & 98 & 100\\
2212 & 92 & 96 & 100\\
2221 & 64 & 57 & 50\\
2222 & 100 & 100 & 100\\
\end{tabular}
}

\includegraphics[width=1.0\columnwidth]{../../notebook/2019-04-24-vaf-model-taejeong/named-figure/tj-vaf-hist-expon-pooled.pdf}
\end{column}

\begin{column}{0.5\textwidth}

\includegraphics[width=1.0\columnwidth]{../../notebook/2019-04-04-truth-sets-aaf/named-figure/vaf-mix1-vaf.pdf}

\includegraphics[width=1.0\columnwidth]{../../notebook/2019-04-04-truth-sets-aaf/named-figure/exp-model0-lam02-mix1.pdf}
\end{column}
\end{columns}
\end{frame}

\begin{frame}{Performance: SNVs}
\includegraphics[width=1.0\textwidth]{../../notebook/2019-08-15-benchmark-calls/named-figure/prec-recall-snp-region-autosomes-mix3_ctr.pdf}
\end{frame}

\begin{frame}
\begin{columns}[t]
\begin{column}{0.5\textwidth}

\includegraphics[scale=0.3]{../../notebook/2019-07-23-commonsample-precrecall/named-figure/prec-recall-common-sample-mosaicforecast.pdf}
\end{column}

\begin{column}{0.5\textwidth}{Discrepancy: MosaicForecast}

\includegraphics[width=1.0\columnwidth]{../../notebook/2019-08-15-benchmark-calls/named-figure/prec-recall-snp-region-autosomes-mix3_ctr.pdf}
\end{column}
\end{columns}
\end{frame}

\begin{frame}{Effect of control sample}
\includegraphics[width=1.0\textwidth]{../../notebook/2019-08-15-benchmark-calls/named-figure/prec-recall-controlsample-effect.pdf}
\end{frame}

\begin{frame}{A family of new truth sets}
\includegraphics[width=1.0\textwidth]{../../notebook/2019-04-04-truth-sets-aaf/named-figure/exp-model-mix1.pdf}
\footnotesize
\begin{tabular}{c|r|r}
s2g & SNV & indel \\
\hline
-4 & 400 & 80 \\
-3 & 4000 & 800 \\
-2 & 40000 & 8000 \\
\end{tabular}
\end{frame}

\begin{frame}{Effect of number of somatic variants}
\includegraphics[width=1.0\textwidth]{../../notebook/2019-08-15-benchmark-calls/named-figure/prec-recall-s2g-effect.pdf}
\end{frame}

\begin{frame}{Effect of allele frequency distribution}
\includegraphics[width=1.0\textwidth]{../../notebook/2019-08-15-benchmark-calls/named-figure/prec-recall-lambda-effect.pdf}
\end{frame}

\section{Callset integration and filtering by machine learning}

\begin{frame}{VCF annotations for classification (idealized figures)}
\begin{columns}[t]
\begin{column}{0.3\textwidth}

partially labeled, unfiltered, callset \(j\):

\(\{\overbrace{y_{i}}^\text{label},
\overbrace{\mathbf{x}_{ij}}^\text{annot}\}_{i=1,...,n_j}\)
\end{column}

\begin{column}{0.7\textwidth}

{\small

\tikz[baseline=-0.5ex] \path (0,0) -- node[draw, fill=blue, circle, inner
sep=2.5pt] {} (1,0); false positive: \(y_{ij}=0\)

\tikz[baseline=-0.5ex] \path (0,0) -- node[draw, fill=red, rectangle, inner
sep=3.0pt] {} (1,0); true positive: \(y_{ij}=1\)

\tikz[baseline=-0.5ex] \path (0,0) -- node[draw, diamond, inner
sep=3.0pt] {} (1,0); unlabeled: \(y_{ij}=?\)
}
\end{column}
\end{columns}
\begin{columns}[t]
\begin{column}{0.5\textwidth}
\end{column}

\begin{column}{0.5\textwidth}

\end{column}
\end{columns}
\begin{columns}[t]
\begin{column}{0.33\textwidth}
\begin{center}
\(j=1\)
\end{center}

\includegraphics[width=1\columnwidth]{figures/by-me/vcf-annot-classif/strelka2/strelka2-unlabeled.pdf}
\end{column}

\begin{column}{0.33\textwidth}
\begin{center}
\(j=...\)

\vspace{0.7in}
\large
\(\cdots\)
\normalsize
\end{center}
\end{column}

\begin{column}{0.33\textwidth}
\begin{center}
\(j=5\)
\end{center}

\includegraphics[width=1\columnwidth]{figures/by-me/vcf-annot-classif/tnseq/tnseq-unlabeled.pdf}
\end{column}
\end{columns}
\end{frame}


\begin{frame}{Probability that a call is true}
\begin{center}
weighted combination of callers
\[
P(i) = \frac{1}{5} \sum_{j=1}^5 P_j(y_{ij} | \mathbf{x}_{ij})
\]
\end{center}
\begin{columns}[t]
\begin{column}{0.33\textwidth}
\begin{center}
\(j=1\)
\end{center}

\includegraphics[width=1\columnwidth]{figures/by-me/vcf-annot-classif/strelka2/strelka2-svm-unlabeled.pdf}
\end{column}

\begin{column}{0.33\textwidth}
\begin{center}
\(j=...\)

\vspace{0.7in}
\large
\(\cdots\)
\normalsize
\end{center}
\end{column}

\begin{column}{0.33\textwidth}
\begin{center}
\(j=5\)
\end{center}

\includegraphics[width=1\columnwidth]{figures/by-me/vcf-annot-classif/tnseq/tnseq-svm-unlabeled.pdf}
\end{column}
\end{columns}
\end{frame}


\begin{frame}[plain]{VariantMetaCaller}{Gezsi et al 2015 BMC Genomics}
\includegraphics[width=1.0\textwidth]{../../notebook/2019-08-15-benchmark-calls/named-figure/prec-recall-vmc-s2g-3.pdf}
\end{frame}

%\againframe{precrecall}

\begin{frame}{Summary}
\begin{itemize}
\item callers' relative performance depends on
\begin{itemize}
\item benchmark data and truth set
\item variant allele frequency (VAF) in brain and control sample
\item distribution of VAF across variants
\end{itemize}
\item notable
\begin{itemize}
\item MosaicForecast: the best under several conditions 
\item lofreqSomatic, strelka2Somatic, MuTect2: strong performance when control sample lacks variant 
\item strelka2Germline: only useful for at high VAF
\end{itemize}
\item integrating multiple callsets
\begin{itemize}
\item machine learning (VariantMetaCaller)
\item high performance
\item precision based filtering 
\end{itemize} 
\end{itemize}
\end{frame}

\begin{frame}{Acknowledgements}
\begin{columns}[t]
\begin{column}{0.5\textwidth}
\alert{Mount Sinai, New York}

\vspace{0.5em}
Andy Chess\\
Chaggai Rosenbluh\\
Attila Jones

\vspace{0.5em}
Schahram Akbarian
\end{column}

\begin{column}{0.5\textwidth}
\alert{Harvard}

\vspace{0.5em}
Peter Park\\
Yanmei Dou\\
Minseok (Daniel) Kwon

\vspace{1.5em}
\alert{Stanford}

\vspace{0.5em}
Alex Urban\\
Xiaowei Zhu
\end{column}
\end{columns}
\end{frame}

\end{document}



\begin{columns}[t]
\begin{column}{0.5\textwidth}

\end{column}

\begin{column}{0.5\textwidth}

\end{column}
\end{columns}
