---
layout: default
tags: [BAM, alignment, genome]
featimg: "read-depth-repres-1.png"
---

All reads were mapped for three samples from a single control individual: NeuN+, NeuN-, muscle.  The coverage was even (except around centromeres) and at least as high as the nominal coverage ordered from the sequencing biotech company.

```{r echo=FALSE, warning=FALSE}
library(lme4)
library(lattice)
#library(latticeExtra)
opts_chunk$set(dpi = 144)
opts_chunk$set(out.width = "700px")
opts_chunk$set(dev = c("png", "pdf"))
lattice.options(default.args = list(as.table = TRUE))
lattice.options(default.theme = "standard.theme")
source("2017-05-24-alignment-stats.R")
```

```{r cache=TRUE}
tissues <- c("NeuN_pl", "NeuN_mn", "muscle")
fai <- get.fai()
chromosomes <- grep("^GL", fai$contig, value = TRUE, invert = TRUE)
rd.genome <- get.read.depths(fai, tissues)
rd.repres <- get.read.depths(fai, tissues, suffices = "chr1_119,500,001-121,500,000")
```

```{r cache=TRUE}
rd.repres <-
    do.call(rbind,
            lapply(regions <- c("2Mb", "10kb baseline", "10kb spike"),
                   function(x) {
                       r <- rd.repres
                       r$region <- factor(x, levels = regions, ordered = TRUE)
                       r
                   }))
rd.repres.100 <- rd.repres[seq(1, nrow(rd.repres), by = 100), ] # for test purposes
rd.genome.100 <- rd.genome[seq(1, nrow(rd.genome), by = 100), ] # for test purposes
fastaix <- "~/data/GRCh37/dna/Homo_sapiens.GRCh37.dna.fa.fai"
```

```{r depth-hist}
depth.hist <-
    lapply(tissues,
           function(x) {
               df <- get.samstats(paste0("../../results/2017-05-24-alignment-stats/MSSM179_", x, ".bam.stats"))
               df <- df[3:4]
               names(df) <- c("depth", "frequency")
               df$tissue <- factor(x, levels = tissues, ordered = TRUE)
               return(df)
           })
depth.hist <- do.call(rbind, depth.hist)
xyplot(frequency ~ depth | tissue, groups = tissue, data = depth.hist, type = "s", xlim = c(-10, 300), layout = c(1, 3), grid = TRUE)
```

```{r read-depth-genome-log, echo=FALSE, fig.asp=1.5, dev="png"}
scalingf <- 1e6
smalldata <- FALSE
rd.g <- if(smalldata) rd.genome.100 else rd.genome
depth.plot(rd.g, chromosomes, chromosomal = TRUE, scales = list(y = list(log = TRUE)))
```

```{r read-depth-genome, echo=FALSE, fig.asp=1.5, dev="png"}
depth.plot(rd.g, chromosomes, chromosomal = TRUE, ylim = c(0, 250))
```

```{r read-depth-repres, echo=FALSE, fig.asp=0.5, dev="png"}
rd.r <- if(smalldata) rd.repres.100 else rd.repres
(tp.regions <- xyplot(depth ~ pos / scalingf | region, data = rd.r, subset = tissue == "NeuN_pl" & suffix == "chr1_119,500,001-121,500,000", scales = list(x = list(rot = 90, relation = "free")), xlim = xlim.region <- list(1e6 / scalingf * c(119.5, 121.5), 1e6 / scalingf * c(120, 120.01), 1e6 / scalingf * c(121.476, 121.486)), type = "s", layout = c(3, 1), xlab = "position (Mb)", grid = TRUE))
update(tp.regions, xlim = xlim.region, ylim = c(0, 400))
```

```{r read-depth-nonchromosomal-log, echo=FALSE}
nonchromosomes <- grep("^GL", fai$contig, value = TRUE)
depth.plot(rd.genome, nonchromosomes, chromosomal = FALSE, scales = list(x = list(rot = 90), y = list(log = TRUE)))
```

```{r read-depth-nonchromosomal, echo=FALSE}
depth.plot(rd.genome, nonchromosomes, chromosomal = FALSE, ylim = c(0, 1000), scales = list(x = list(rot = 90)))
```
