---
layout: default
title: Call set concordance---test data
tags: [ strelka2, mutect2, tnseq, lofreq, somaticsniper ]
#featimg: "runtime-length-2.png"
---

Attempt to represent the concordance among call sets corresponding to various variant callers (strelka2Somatic,strelka2Germline, Tnseq/Mutect2, lofreq, somaticSniper.

```{r echo=FALSE, warning=FALSE, message=FALSE}
library(lme4)
library(lattice)
library(latticeExtra)
opts_chunk$set(dpi = 144)
opts_chunk$set(out.width = "600px")
opts_chunk$set(dev = c("png", "pdf"))
lattice.options(default.args = list(as.table = TRUE))
lattice.options(default.theme = "standard.theme")
```

```{r message=FALSE}
library(VariantAnnotation)
library(VennDiagram)
source("2018-04-08-call-set-concordance.R")
```

```{r cache=TRUE}
samples <- c("common.sample", "benchmark")
names(samples) <- samples
segs <- c(paste0(c(1, 3, 10, 30, 100), "MB"), "wgs")
names(segs) <- segs
vartypes <- c("snvs", "indels")
names(vartypes) <- vartypes
callers <- c("lofreqSomatic", "somaticSniper", "strelka2Germline", "strelka2Somatic", "Tnseq")
names(callers) <- sub("Tnseq", "Tnseq.Mutect2", callers)
# benchmark study
bm.vcf <- import.across.segments(samples[2], sgs = c("1MB", "wgs"), cal = callers)
```

```{r cache=TRUE}
# common sample
cs.vcf <- import.across.segments(samples[1], sgs = segs[- length(segs)])
```

```{r venn-common-sample-100MB-snvs}
l <- lapply(cs.vcf[["100MB"]]$snvs, function(x) row.names(x))
grid.draw(venn.diagram(l, NULL))
```

```{r venn-common-sample-100MB-indels}
l <- lapply(cs.vcf[["100MB"]]$indels, function(x) row.names(x))
grid.draw(venn.diagram(l, NULL))
```

```{r venn-benchmark-wgs-snvs}
l <- lapply(bm.vcf[["wgs"]]$snvs, function(x) row.names(x))
grid.draw(venn.diagram(l, NULL))
```

```{r venn-benchmark-wgs-indels}
l <- lapply(bm.vcf[["wgs"]]$indels, function(x) row.names(x))
grid.draw(venn.diagram(l, NULL))
```
