---
layout: default
#title: 
featimg: vaf-alt-tumor-normal-1.png
---

The VCF files produced by our somatic variant callers show variation in their precise format.  Some of these specific VCF formats are not supported by VariantMetaCaller, therefore formatting adjustments, "hacks", are required to comply with VariantMetaCaller.  This article describes these hacks, the rational behind them, and their effects on running VariantMetaCaller.  This work has been facilitated by discussions with the developer of VariantMetaCaller, András Gézsi.

## Testing VariantMetaCaller with previously tested germline callers

The VCF files of four germline callers presented in Gézsi et al 2015 were obtained from the authors.  The run of VariantMetaCaller with these input VCFs was successful.  This was expected since VariantMetaCaller was developed using the same set of germline callers.  See `2018-06-22-variant-meta-caller-test/vmc-prioritize-testvcf-germline` directory.

## With our somatic callers

### Summary

The first test runs of VariantMetaCaller failed.  As hacks were incrementally introduced the runs became successful, then faster, and finally more complete in the sense of including all 5 SNV callers instead of just the initial 4 callers.  The following hacks were made to VCFs of one or more callers (listed in parentheses):

* correct sample names (strelka2Somatic, somaticSniper)
* add sample fields (lofreqSomatic)
* move from 1 to 2 samples (strelka2Germline)
* add missing genotype---GT tags---(lofreqSomatic, strelka2Somatic)
* add annotations to the INFO field that were generated by annotatevcf.sh (all callers)

The remainging TODO list:

* BQSR for Tnseq (Don Freed suggested it)
* implement all hacks as shell scripts (ongoing)
* calculate precision and recall given gold standard (coordinate with Chaggai)
* fine tune selection of annotations (depends on precision and recall)

## Details of individual hacks

### add missing genotype---GT tags

The somatic caller of strelka2 ([see paper](https://www.nature.com/articles/s41592-018-0051-x)) assumes that 

>the normal sample is a mixture of diploid germline variation and noise, whereas the tumor sample is a mixture of the normal sample and somatic variation

It follows that, at least for the tumor sample, it makes no sense to assign a single genotype.  Therefore strelka2Somatic VCFs lack the GT (genotype) tag in the sample fields.  The problem is that VariantMetaCaller requires GT tag when it calculates the *positive threshold*, which is the "Minimum number of callers supporting a variant to consider as positive training instance for SVM" (see VariantMetaCaller's help).

The `$HOME/bin/vcf-add-GT` script solve that problem by adding GT 0/1 for TUMOR and GT 0/0 for NORMAL. This is, of course, a simplification which ignores the typically mixed genotypes in the tumor sample.  But it seems to be a reasonable simplification based on the following analysis.

The `strelka2Somatic-getsamplefields` script takes all the tags from the NORMAL and TUMOR sample fields in `strelka2Somatic.vcf` and stores those into a tsv file.

```{r engine="bash", eval=FALSE}
cd $HOME/projects/bsm/results/2018-06-22-variant-meta-caller-test/vmc-prioritize-benchmark/test08/
$HOME/projects/bsm/notebook/2018-06-22-variant-meta-caller-test/strelka2Somatic-getsamplefields ../22/strelka2Somatic.vcf > strelka2Somatic.tsv
```

```{r echo=FALSE, warning=FALSE, message=FALSE}
library(lme4)
library(lattice)
library(VennDiagram)
opts_chunk$set(dpi = 144)
opts_chunk$set(out.width = "600px")
opts_chunk$set(dev = c("png", "pdf"))
lattice.options(default.args = list(as.table = TRUE))
#lattice.options(default.theme = "standard.theme")
source("2018-06-22-variant-meta-caller-test.R")
```

The tags in the sample fields look like this:

```{r}
gt <- read.delim("~/projects/bsm/results/2018-06-22-variant-meta-caller-test/vmc-prioritize-benchmark/test08/strelka2Somatic.tsv")
str(gt)
```

Now variant allele frequencies $$v_{a,s}$$ for reference or alternative allele $$a$$ and tumor or normal sample $$s$$ are calculated as follows:

$$
\begin{equation}
v_{a,s} = \frac{\#_{a,s}}{\#_{s}},
\end{equation}
$$

where $$\#_{a,s}$$ is the count of reads carrying allele $$a$$ in sample $$s$$, whereas $$\#_{s}$$ is the total read count (depth) in sample $$s$$.  See `get.vaf` function in `2018-06-22-variant-meta-caller-test.R` for implementation details.

```{r}
vaf.long <- reshape.vaf(vaf <- get.vaf(gt))
```

The plot below shows that the alternative allele frequency is always positive (nonzero) for the tumor sample but often zero in the normal sample suggesting that the simplification introduced above is reasonable.

```{r vaf-alt-ref}
xyplot(Alt.VAF ~ Ref.VAF | Sample, data = vaf.long, pch = "+",
       xlab = "Reference Allele", ylab = "Alternative Allele", main = "Allele Frequency")
```

The next plot further supports this notion by showing that even if the alternative allele frequency is positive (nonzero) in the normal, it is even higher in the tumor sample.

```{r vaf-alt-tumor-normal}
xyplot(Alt.VAF.Normal ~ Alt.VAF.Tumor, data = vaf, pch = "+", subset = Alt.VAF.Normal >= 0,
       panel = function(x, y, ...) {
           panel.xyplot(x, y, ...)
           panel.abline(a = 0, b = 1, lty = 2)
       }, xlim = c(-0.02, 0.5), ylim = c(-0.02, 0.5), main = "Alternative Allele Frequency",
       xlab = "Tumor", ylab = "Normal")
```

The `$HOME/bin/vcf-add-GT` script assumes the normal sample has 0/0 GT for each somatic variant and the plots above do seem to suggest this assumption is reasonable.  So what about 0/1 genotype in the normal tissue?  Somatic variants in the tissue may then be of genotype 0/0 or 1/1.  But these types of variants are the more difficult to detect by strelka2's somatic calling model ([see paper](https://www.nature.com/articles/s41592-018-0051-x)) therefore these will be unlikely to end up in `strelka2Somatic.vcf`

