\documentclass[letterpaper]{article}
%\documentclass[12pt,letterpaper]{article}
%\setlength{\textwidth}{480pt}
%\setlength{\textheight}{630pt}
%\setlength{\voffset}{0pt}

\usepackage{amsmath, geometry, graphicx}
\usepackage{natbib}
%\usepackage{float}
\bibliographystyle{plainnat}

% https://tex.stackexchange.com/questions/6758/how-can-i-create-a-bibliography-like-a-section
%\usepackage{etoolbox}
%\patchcmd{\thebibliography}{\section*}{\section}{}{}

\pagestyle{plain}

\title{Involvement of Brain Somatic Mosaicism in Schizophrenia}

\author{Attila Jones\(^{1,2}\), Eduardo Maury\(^{3}\), Chaggai Rosenbluh\(^{1,4}\), ..., Andrew Chess\(^{1,2,5,6,\ast}\)}

\date{Icahn School of Medicine at Mount Sinai (ISMMS)}

\begin{document}

\maketitle

\begin{description}
\item[1] Department of Cell, Developmental and Regenerative Biology, ISMMS 
\item[2] Institute for Genomics and Multiscale Biology, Department of Genetics and Genomic Sciences, ISMMS 
\item[3] Harvard Medical School
\item[4] Department of Genetics, Hadassah Medical Center, Jerusalem, Israel
\item[5] Neuroscience Program, The Graduate School of Biomedical Sciences, ISMMS 
\item[6] Division of Psychiatric Genomics, Department of Psychiatry, ISMMS 
%\item[\(\S\)] full list of consortium members appears in the Author
Information section
\item[\(\ast\)] correspondence: andrew.chess@mssm.edu 
\end{description}

\clearpage

\section*{Abstract}

TODO

\section*{Introduction}

Schizophrenia is a common psychiatric disorder whose onset in early adulthood
manifests in psychotic episodes and social dysfunction \citep{Kahn2015}.
Early diagnosis as well as effective prevention and treatment are unavailable
but the high (80\%) heritability of the disease provides hope that genetic
research will help clarify pathomechanisms and provide biomarkers and drug
targets \citep{Breen2016,Foley2017}.  GWAS (reviewed by \cite{Visscher2017})
have been essential in clarifying the genetic of architecture of
schizophrenia~\citep{Ripke2014,Pardinas2018} and other psychiatric disorders
\citep{Sullivan2012}: they are polygenic and especially in schizophrenia much
of the heritability is due to a large set of common germline variants, mostly
SNPs, with small effect size.  GWAS of schizophrenia have so far yielded 145
genome-wide significant loci distributed roughly uniformly along the genome
involving about 700 genes, c.a 200 of which are located at the most
significant locus in the MHC region, suggesting that an impaired immune
function is responsible for the increased synaptic pruning seen in
schizophrenic brains~\citep{Sekar2016a}.  GWAS
also revealed that psychiatric disorders share much heritability with each
other~\citep{Consortium2009,PsychiatricGenomicsConsortium2013} and to a lesser
extent with some neurological disorders~\citep{Consortium2018} indicating
shared pathomechanisms and supporting the neurodevelopmental hypothesis of
schizophrenia~\citep{Nour2015}.

Beyond common SNPs other classes of genetic variation have also been shown to
carry substantial risk for schizophrenia: rare CNVs~\citep{Rees2014}, rare
germline SNVs~\citep{Purcell2014,Singh2017} as well as de novo
SNVs~\citep{Fromer2014,Rees2020}.  These studies confirmed the polygenicity of
schizophrenia and found increased rate of damaging or loss of function
mutations in several functional gene sets such as synaptic and voltage gated
calcium channel and FMRP target genes~\citep{Fromer2014,Purcell2014} and loss
of function intolerant genes~\citep{Rees2020,Singh2017} but found unaltered
genome-wide germline single nucleotide mutation rate in schizophrenics.

Besides germline variants a number of distinct genetic and epigenetic
mechanisms are hypothesized to play a role in schizophrenia and psychiatric
disorders~\citep{PsychENCODEConsortium2015}.  Somatic variants have been
called the dark matter of psychiatric genetics~\citep{Insel2014}.  Their
systematic, genome-wide study in brain disorders has just begun with the
announcement of the Brain Somatic Mosaicism
project~\citep{McConnell2017}.\footnote{TODO: cite recent work on BSM and mention
some key findings}
Within that project we and others recently developed new best practices and
tools for calling somatic variants from brain samples based on bulk tissue
sequencing data~\citep{Wang2021}.  In the present work, for the first time, we
apply these analytical tools to deep, whole genome sequencing data from the
dorsolateral prefrontal cortex (DLPFC) of schizophrenic and control
individuals.

\section*{Results}

\subsection*{Somatic SNV callsets}

We extracted NeuN+ nuclei from postmortem DLPFC samples of 40 control
and 61 schizophrenic individuals and subjected them to short read DNA
sequencing at 50--540\(\times\) coverage.  Next we called somatic SNVs using a
pipeline developed and validated by the BSMN consortium~\citep{Wang2021}.  The
genome-wide somatic SNV call rate was similar for control and schizophrenic
samples (median around 40) and showed great variability, ranging 
0---100 except for an outlier sample at 189
(Fig.~\ref{fig:genomewide-calls} top and middle).

\begin{figure}
\includegraphics[scale=0.6]{../../notebook/2020-12-09-walsh-data/named-figure/genome-wide-callrate.pdf}
\caption{
Genome-wide somatic SNV call rate for control and schizophrenic
samples and its dependence on sequencing depth.
}
\label{fig:genomewide-calls}
\end{figure}

We sought biological and technical variables that explain that high variance
of genome-wide.  Sequencing depth stood out among the observed variables
(Fig.~S\ref{fig:scattermatrix}) explaining much of the variability in call
rate (Fig.~\ref{fig:genomewide-calls} bottom). This means that the variant
calling pipeline adapts to lower coverage by applying stronger filters in
order to maintain high specificity.

The highest genome-wide rate (189 somatic SNVs) was observed in a sample from
a schizophrenic woman.  Considering that its sequencing depth is only moderate
this sample is clearly an outlier and the reason is possibly
biological.\footnote{TODO: case study of CMC\_MSSM\_224, perhaps a
supplementary table of its variants with some annotations (gene, predicted
functional consequence)?  Or a mutation spectrum analysis suggested by Andy?}

Our pipeline provided us with an estimate of allele frequency for each somatic
SNV.  This enabled us to assess, at least qualitatively, the simple model that somatic
point mutation rate is roughly constant along the cell lineages of our
samples~\citep{Rodin2021} since under this model the density of allele
frequency is quasi exponential.  We found this to be the case at high
(\(\approx 400\)) read depth (blue allele frequency density in Fig.~\ref{fig:AF}
right panel).  With decreasing read depth, however, the density of allele
frequency is increasingly shifted towards higher frequencies (orange and green
densities, Fig.~\ref{fig:AF} right panel).  These results indicate that the somatic
SNVs we called follow the expected allele frequency distribution but our
pipeline's sensitivity for low allele frequency (below c.a 0.05) is low at
low-medium sequencing coverage.

\begin{figure}
\includegraphics[scale=0.6]{../../notebook/2020-08-13-select-vars/named-figure/calls-DP-AF-jointdensity-conddensity.pdf}
\caption{
Distribution of allele frequency and read depth.  Note that even though allele frequency is
a continuous variable its estimate is discrete, which is an inherent feature
of the pipeline.
}
\label{fig:AF}
\end{figure}

\subsection*{Local rates of somatic SNVs in schizophrenia}

Next we investigated weather the rate of somatic SNVs falling into specific
functional categories is altered in schizophrenia.  We refer to this as local
rate to distinguish from the genome-wide (global) rate discussed above.  We
tested genic categories such as nonsynonymous and deleterious SNVs; regulatory
categories such as enhancers; constraint categories such as loss of function
intolerant genes; and gene sets affected by rare and common variants in
schizophrenia and related psychiatric disorders (Table~\ref{tab:modsel}).

\begin{table}
\begin{tabular}{lrlr}
group & SNV category & selected model formula & SNVs called \\
\hline
\hline
genic & coding nonsyn & 1 & 66\\
 & coding syn & 1 + Gender & 36\\
 & stop gain &  & 1\\
 & splice site &  & 1\\
 & deleterious (sift) & 1 & 17\\
 & damaging (polyphen) & 1 + DP & 8\\
regulatory & TFBS & 1 + AF & 126\\
 & CpG island & 1 + AF + BaseQRankSum & 73\\
 & enhancer & 1 & 130\\
 & genic enhancer & 1 & 12\\
constraint & evolutionary cons. (phast) & 1 + DP + AF + Gender & 276\\
 & evolutionary cons. (gerp) & 1 + Dx & 402\\
 & LoF intolerant genes & 1 + Dx & 1077\\
rare var.~gene sets  & PSD & 1 & 112\\
 & FMRP-target & 1 + AF & 227\\
 & SCZ de novo & 1 + AF & 112\\
 & Calcium channel & 1 + Dx & 11\\
 & composite & 1 & 377\\
GWAS gene set & ADHD GWAS & 1 + AF + DP & 5\\
 & Anorexia nervosa GWAS &  & 2\\
 & ASD GWAS & 1 + Dx & 22\\
 & Bipolar disorder GWAS & 1 + Dx & 91\\
 & MDD GWAS & 1 & 40\\
 & Schizophrenia GWAS & 1 + Dx & 86\\
 & Psychiatric GWAS & 1 + Dx & 176\\
\end{tabular}
\caption{
	Studied categories of SNVs, variables explaining local rate of somatic
	SNVs (selected model formula) and the number of SNVs called (the size
	of the category).
}
\label{tab:modsel}
\end{table}

For each SNV category we tested weather the local rate of somatic SNVs is
affected by disease status, Dx.  For that purpose we employed one univariate test
($\chi^2$) and two multivariate regression models---logistic and log-linear
regression (Fig.~S\ref{fig:fitted-lines}) ---, which we combined with a forward
variable selection algorithm (Fig.~S\ref{fig:varsel}) on a set of biological
and technical explanatory variables (Fig.~S\ref{fig:scattermatrix}).  These three techniques
lead to highly concordant inferences (compare p-values between the two
rightmost panels in Fig.~\ref{fig:testresults} and compare
Fig.~S\ref{fig:varsel}A to~S\ref{fig:varsel}B) suggesting that confounding variables or false
modeling assumptions did not undermine our analyses.

Three categories of somatic SNVs displayed significantly increased local rate
in schizophrenia samples after multiple test correction: Schizophrenia GWAS,
Psychiatric GWAS (both at false discovery rate, FDR, below $\alpha=0.025$ for
both the $\chi^2$ and the logistic model based test) and ASD GWAS (FDR below
$\alpha=0.025$ for only the $\chi^2$ test), see Fig.~\ref{fig:testresults},
supporting-data-file1.csv and supporting-data-file2.csv\footnote{TODO: turn
supporting-data-file2.csv into a supplementary table}.  For four
additional categories the forward algorithm selected the disease
status variable Dx (Table~\ref{tab:modsel}) but the local rate was
only nonsignificantly increased (Fig.~\ref{fig:testresults}) in
schizophrenic samples. These categories are: Bipolar disorder GWAS,
Calcium channel genes, loss of function (LoF) intolerant genes and
evolutionary conserved regions detected by GERP++.

\begin{figure}
\includegraphics[width=0.9\textwidth]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/enrichment-5-plots_scz.pdf}
\caption{Enrichment of genomic elements and gene sets in somatic SNVs in schizophrenia}
\label{fig:testresults}
\end{figure}

For the remaining categories in Table~\ref{tab:modsel} no change in local rate
was detected in schizophrenic samples relative to controls.  The
interpretation of such negative result depends on the size of the given
category, which we define as the somatic SNVs called
(Fig.~\ref{fig:testresults} left, Table~\ref{tab:modsel}).  For larger
categories (relative to, say, Schizophrenia GWAS with 72 SNVs called), such as
enhancers (99 SNVs) or FMRP-target genes (118 SNVs), this means no or small
(relative to Schizophrenia GWAS) change in local somatic SNV rate.  For
smaller categories, such as ADHD GWAS (5 SNVs) or the 17 deleterious SNVs
inferred by SIFT the interpretation is open since not enough SNVs were
called to fit sufficiently complex models well.  Moreover, for very small
categories such as Anorexia GWAS (2 SNVs), the iterative algorithm of the fit
failed to converge.  This is seen in
Fig.~S\ref{fig:binomial-QQ}-S\ref{fig:poisson-QQ} from the progressive
improvement in the goodness of fit with increasing number of somatic SNVs
called.  This analysis also showed that for the smaller sized ASD GWAS (22
SNVs) the fit is already adequate and for the relatively large Schizophrenia
GWAS (86 SNVs) and Psychiatric GWAS (176 SNVs) the fit is excellent, which
gives further support to our model based inference regarding these categories.

The moderately increased (2-5 fold, Fig.~\ref{fig:testresults}) local rate of
somatic SNVs in some of the gene set categories was expected to be a mixed
effect of a much increased rate of damaging variants and unaltered rate of
neutral variants. We attempted at testing this expectation by narrowing down
the gene set categories to nonsynonymous SNVs, since those tend to be
damaging mutations.  However, our attempt was undermined by too few
nonsynonymous SNVs even in the largest geneset, the LoF intolerant genes,
which contained only 30 nonsynonymous SNVs out of 1077 SNVs and for which the
$\chi^2$ test yielded no significant change in the local somatic SNV rate ($p
= 0.7$).\footnote{TODO: supplementary table with nonsyn SNVs for gene sets.}

Next we performed a complementary test to evade the limited number somatic SNV
calls.  Instead of narrowing down our Schizophrenia GWAS category we extended
it by including SNVs not only in genic but also in intergenic regions, in
which SNVs tend to be more neutral than those in genes.  In agreement with
that, the local somatic SNV rate was no longer significantly increased in
schizophrenic samples relative to controls
(Fig~S\ref{fig:scz-gwas-genes-vs-loci}).  Of note is that the most strongly
associated locus to schizophrenia is at the MHC region, which is only
represented in the extended Schizophrenia GWAS loci category but not in the
Schizophrenia GWAS gene set due to its structural variability.  To address
this discrepancy we narrowed down again the GWAS loci to genic regions but
this time also including the MHC locus.  We found, again, a significantly
increased local rate of somatic SNVs despite the likely inclusion of some
intergenic regions at the MHC locus (Fig.~S\ref{fig:scz-gwas-genes-vs-loci}).

Also, we ranked schizophrenia GWAS loci according to their p-values for
genome-wide association to see if the somatic SNV rate in schizophrenic
samples is associated with rank.  However, the small number of somatic SNVs
($\approx 0.6$ SNV per locus on average) only allowed us to conclude that
somatic SNVs---whether from control or schizophrenic samples---are roughly
evenly distributed across GWAS loci excluding locus 1 (the most significant
locus i.e the MHC region), which harbored 4 somatic SNVs reflecting its larger
size compared to other loci (Fig.~\ref{fig:indiv-scz-gwas-loci}).

Finally, we wondered how the observed local rates (Fig.~\ref{fig:testresults})
change when we narrow the definition of SNV categories to include only genes
highly expressed in the brain.  This analysis yielded similar relative local
rates to the ones already observed.\footnote{TODO: create supplementary figure
	or table
from 2020-11-13-functional-variants}

\section*{Discussion\footnote{TODO: In progress...}}

In this work\footnote{this paragraph might be a good basis for the abstract
and a cover letter for manuscript submission} we show, for the first time,
that somatic SNVs in certain gene sets contribute to the risk of a psychiatric
disorder, namely schizophrenia.  This finding is not only statistically sound
(false discovery rate below 0.025) but also biologically plausible
because the affected gene sets are those identified by GWAS on schizophrenia
and genetically related psychiatric disorders.

Although this general finding is confirmatory as somatic variants were
expected to play a role in psychiatric disorders~\citep{McConnell2017} it is
still remarkable if we consider the challenges of studying somatic variants in
the brain.  These include at least an order of magnitude fewer somatic
variants than germline variants; the need for very deep whole-genome sequencing
(Fig.~\ref{fig:genomewide-calls},~\ref{fig:AF}), the associated increased
cost of data processing and a customized variant calling pipeline, which has
been validated~\citep{Wang2021} in our laboratory using the same sequencing
and variant calling protocols as those used in our present work.

Studies of rare germline and de-novo SNVs showed that, although the
genome-wide ``background'' mutation rate is unaltered in schizophrenia,
specific gene sets like PSD, FMRP-target or LoF intolerant genes bear
increased burden of likely damaging~\citep{Purcell2014} or
LoF~\citep{Singh2017} SNVs in schizophrenic individuals.  Our present study
might be regarded as an extension that line of research to somatic SNVs in the
human brain since it seems plausible that a large class of somatic mutations
in the brain mimic the genetic pathomechanism of rare and de-novo variabts.
However, the above mentioned challenges prevented us to study damaging and let
alone LoF SNVs.  Despite that we found a nonsignificant but suggestive
increase in local somatic SNV rate at Calcium channel and LoF intolerant genes
as well as in evolutionary constrained regions without restricting our SNVs to
nonsynonymous or likely damaging ones.  Moreover, our more detailed analysis
of schizophrenia GWAS genes and loci suggests that not only exonic but also
many intronic somatic SNVs play a role in schizophrenia.

TODO: more questions to discuss:
\begin{itemize}
\item What insights might study of brain somatic mosaicism provide on schizophrenia
pathomechanism that go beyond the findings obtained from rare germline and
de-novo variant studies?  
\item At what cost with our current techniques~\citep{Wang2021} and how can the cost be minimized?  
\item How much do somatic SNVs contribute to schizophrenia risk relative to other
classes of variants?  Inverse relationship of different classes of risk
carrying variants~\citep{Bergen2019}
\item Clinical utility of further studies on somatic SNVs in
schizophrenia and psychiatric disorders?
\end{itemize}


\section*{Methods}

\subsection*{Study subjects, sample preparation, sequencing}

Post-mortem DLPFC samples of predominantly Caucasian subjects (61
schizophrenic and 25 control) were obtained from the Mount Sinai NIH Brain
Bank and Tissue Repository~\citep{Hoffman2019}.  In addition we included DLPFC
samples of 15 control subjects from~\citep{Rodin2021}, which did not
qualitatively change the results.

NeuN+ nuclei isolation was performed on DLPFC tissue samples using
fluorescence-activated nuclei sorting. The Illumina PCR-free TruSeq DNA
Library Prep kit was used according to the manufacturer's protocol to produce
the DNA libraries of the sorted samples. Constructed libraries were quantified
using the KAPA Library Quantification Kit with real-time PCR and were then
sequenced on the Illumina HiSeq X Ten platform to yield 150bp paired-end reads
for all samples. Sequencing was designed aiming for 200X coverage per sample.

\subsection*{Somatic SNV calling and validation}

Sequencing reads were processed by a computational pipeline described in
detail by our recent methodological study~\citep{Wang2021}.  Briefly, for each
sample, the reads were aligned to the reference sequence GRCh37d5 using bwa
v3.7.16a~\citep{Li2009b}. Duplicate reads were marked with Picard v2.12.1
(https://broadinstitute.github.io/picard/). Indel realignment and base quality
score recalibration using GATK v3.7-0~\citep{McKenna2010} gave rise to the
final BAM file for each sample.

From the final BAM files, single nucleotide variants and small indels were
called using GATK (v3.7-0) HaplotypeCaller with \texttt{--ploidy} option 2, 12, or 50
resulting in three raw call sets per sample. The raw call
sets were filtered in the following sequential steps: (1) Only PASS calls were
retained; (2) known germline variants were excluded based on the 1000 Genomes,
ExAC, GnomAD, and Kaviar databases
\citep{Auton2015,Karczewski2020,Glusman2011}, as well as ESP5600
(https://evs.gs.washington.edu/EVS/); (3) calls located in problematic
genomic regions defined as non-P bases of the 1000 Genomes were removed; (4) a
binomial test was used to remove likely heterozygous variants: for each
variant the $p$ and $N$ parameter of the binomial null distribution was set to 0.5
and the total read count (read depth) at the variant position, respectively,
and we removed the variant supporting the null hypothesis at $\alpha = 0.00001$; (5)
a similar binomial hypothesis test was used to filter variant calls at sites
with strand bias; (6) similarly, Fisher's exact test was used to filter out
calls with imbalance of strand ratios between reference (REF) and alternative
(ALT) bases; (7) multi-allelic calls were removed; (8) calls at sites with $>
2.5$ copy number, estimated using CNVnator~\citep{Abyzov2011}, were also
removed.

\subsection*{Variant annotation}

SNPnexus~\citep{Oscanoa2020}

RoadMap,... TODO: list all sources of annotation

\subsection*{Statistical analysis}

For a genomic element or gene set  \(S\) we inspected the joint distribution,
across all study subjects, of disease status Dx, two enrichment variables
\(n_S\), \(f_S\), and technical and biological variables that potentially
carry information on those enrichment variables.  The enrichment variable
\(n_S\) is the number of somatic SNVs located in \(S\), whereas the \(f_S =
	n_S/n_\mathrm{genome}\) is the fraction of somatic SNVs in \(S\)
relative to all somatic SNVs in the genome.  The aim was to formally test
enrichment in \(S\) in schizophrenia by regressing the response variables
\(n_S\) or \(f_S\) on the explanatory variable Dx
and other, ``confounding'', variables.

In case of the plot matrix (Fig~S\ref{fig:scattermatrix}) \(S\) is the set of schizophrenia GWAS genes
defined as all genes, including their introns, that are located fully or
partially in schizophrenia GWAS loci (Table S4 in \cite{Pardinas2018}).

Notable observations:
\begin{itemize}
	\item the strongsest dependence was seen between \(n_S\) and \(n_\mathrm{genome}\) as expected
		when the genome-wide background rate of somatic SNVs is
		substantial relative to enrichment in \(S\)
	\item we have already noted the strong dependence between read depth DP and
		\(n_\mathrm{genome}\)
        \item other dependencies were relatively weak
	\item for many control individuals ageOfDeath is below 50 years while for
		most schizophrenic individuals it is above 50.  This is
		because the Mount Sinai dataset contained older individuals
		that were either control or schizophrenic but the Harvard dataset contained
		only younger control individuals
\end{itemize}

\subsubsection*{Modeling the number and fraction of somatic SNV calls in a gene set}

We considered two generalized linear model families depending on the response
variable:
\begin{description}
\item[response is \(f_S\):] logistic family: binomial error distribution with
	logistic link function; see panel (A)
\item[response is \(n_S\):] log-linear family: Poisson error distribution with exponential link function; see panel (B)
\end{description}

Within each family we considered relatively simple or complex models depending
on the set of variables selected as explanatory.  For \(S =\) schizophrenia
GWAS genes, and within the
logistic family, Fig~S\ref{fig:fitted-lines}A shows fitted lines under both \(f_S \sim 1 +
	\mathrm{Dx}\) and \(f_S \sim 1 + \mathrm{Dx} + \mathrm{ageOfDeath}\)
(panel (A)), where \(1\) means intercept.  For the log-linear family panel
(Fig~S\ref{fig:fitted-lines}B)
shows fitted lines under both \(n_S \sim 1 + \mathrm{Dx}\) and \(n_S \sim 1 +
	\mathrm{Dx} + n_\mathrm{genome}\)


\subsubsection*{Variable selection for regression models}

To select the best fitting model within each family we applied forward
variable selection algorithm with the Akaike information criterion (AIC) as
optimality criterion.  (The Bayesian (BIC) information instead of AIC gave the
same results.)

In the case of \(S =\) schizophrenia GWAS genes we see that the first,
thus most informative, variable selected is Dx under the logistic model
(Fig~S\ref{fig:varsel}A).
Any additional variables increased the information criteria and thus lead to
overfitting the data.

(Fig~S\ref{fig:varsel}B) shows the same result under the log-linear model except that the
most informative variable is \(n_\mathrm{genome}\), which was expected by its
strong correlation with \(n_S\).  Nonetheless, Dx is selected also under this
model family.

\subsubsection*{The effect of ageOfDeath and that of Dataset are mediated by disease status Dx}

As we noted one of our two datasets contained relatively young individuals who
were all control.  This leads to the causal-probabilistic graphical model that
\(\mathrm{ageOfDeath} \leftarrow \mathrm{Dataset} \rightarrow \mathrm{Dx}
\rightarrow y\), where \(y\) is the response enrichment variable \(f_S\) or
\(n_S\).  It means that ageOfDeath and Dataset have indirect effects on
enrichment such that these effects are mediated by disease status Dx. If this
model is true then \(Dx\) d-separates \(y\) from Dataset and ageOfDeath so
that \(y\) becomes independent from ageOfDeath and Dataset if we condition on
Dx = control or Dx = schizophrenia.  Fig~S\ref{fig:d-separation}A illustrates
that.  The important consequence is that we should remove ageOfDeath and
Dataset from the set from our models as long as we include Dx.  Failing to do
so may confound our inference on the effect of Dx on enrichment
(Fig~S\ref{fig:d-separation}B).

\subsubsection*{Residual analysis to assess model fit }

We assessed model fit by residual analysis.  There are several types of residuals in use for generalized linear models.
The $r^\ast$ residual is approximately normally distributed even if the error
distribution for the model family is not normal, such as Poisson or
binomial~\citep{davison2003statistical} (Ch10.2 p477).

This residual is defined as follows

$$
r^\ast = r_\mathrm{D} + r_\mathrm{D}^{-1} \log
\frac{r_\mathrm{P}}{r_\mathrm{D}},
$$

where \(r_\mathrm{D}\) is the standardized deviance residual and
\(r_\mathrm{P}\) is the standardized Pearson residual.

We expressed the size of genomic element or gene set as the number of
individuals with at least one somatic SNV in that genomic element or gene set.
We sorted genomic elements and gene sets according to their size and produced
normal Q-Q plots of \(r^\ast\) from top left of Fig~S\ref{fig:binomial-QQ} (smallest) to
bottom right (largest).  This indicated poor fit for the smallest genomic
elements and gene sets (top row of plots) but good fit for medium sized and
large ones.

\subsubsection*{Multiple test correction}

\subsection*{Data availability}

All FASTQ, CRAM and VCF files generated for this work have been submitted to
collection 2965 of the National Institute of Mental Health Data Archive
(https://nda.nih.gov/) and will be released for the public.

\subsection*{Code availability}

All code developed by A.~Jones and E.~Maury is available at:\\
https://github.com/attilagk/BSM-research

\bibliography{library}


\section*{Acknowledgements}

We thank...

\section*{Author Information}

\section*{Author Contributions}

A.J...

\section*{Competing Interests}

None

\section*{Figure Legends}

\subsection*{Figure 1}
%\includegraphics[width=1.0\textwidth]{figures/by-me/commonmind-rna-seq-ms/commonmind-rna-seq-ms.pdf}

\section*{Tables}

\section*{Supplementary Material}

\setcounter{table}{0}
\makeatletter 
\renewcommand{\figurename}{Supplementary Table} % nice
\makeatother

\setcounter{figure}{0}
\makeatletter 
\renewcommand{\figurename}{Supplementary Figure} % nice
\makeatother

\begin{figure}[p]
\begin{center}
\includegraphics[scale=0.6]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/symmetric-scatter-matrix_scz.pdf}
\end{center}
\caption{
Pairwise joint distributions of response and explanatory variables TODO: fix legend
}
\label{fig:scattermatrix}
\end{figure}

\begin{figure}[p]
\begin{center}
A\includegraphics[scale=0.45]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/scz_gwas_genes-ageOfDeath-binom-fit_scz.pdf}
B\includegraphics[scale=0.45]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/scz_gwas_genes-log10_ncalls-poisson-fit_scz.pdf}
\end{center}
\caption{
Modeling the number (A) and fraction (B) of somatic SNV calls in a gene set
}
\label{fig:fitted-lines}
\end{figure}

\begin{figure}[p]
\begin{center}
A

\includegraphics[scale=0.5]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/scz_gwas_genes-fw-varsel-binom-onlyIC_scz.pdf}

B

\includegraphics[scale=0.5]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/scz_gwas_genes-fw-varsel-pois-onlyIC_scz.pdf}
\end{center}
\caption{
Variable selection for regression models. A) Binomial--logistic
family.  B) Poisson--log-linear family.
}
\label{fig:varsel}
\end{figure}

\begin{figure}[p]
\begin{center}
A\includegraphics[scale=0.45]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/Dx_Dataset_ageOfDeath_scz.pdf}
B\includegraphics[scale=0.45]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/collinearity_Dx_Dataset_ageOfDeath_scz.pdf}
\end{center}
\caption{
The effect of ageOfDeath and that of Dataset are mediated by disease status Dx
}
\label{fig:d-separation}
\end{figure}

\begin{figure}[p]
\includegraphics[width=0.9\textwidth]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/resid-binom_scz.pdf}
\caption{
Residual analysis of goodness of fit for the logistic family.
}
\label{fig:binomial-QQ}
\end{figure}

\begin{figure}[p]
\includegraphics[width=0.9\textwidth]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/resid-pois_scz.pdf}
\caption{
Residual analysis of goodness of fit for the log-linear family.
}
\label{fig:poisson-QQ}
\end{figure}

\begin{figure}[p]
\includegraphics[scale=0.6]{../../notebook/2020-07-28-CLOZUK/named-figure/rel-callrate-sczgwas_scz.pdf}
\caption{
TODO
}
\label{fig:scz-gwas-genes-vs-loci}
\end{figure}

\begin{figure}[p]
\includegraphics[scale=0.6]{../../notebook/2020-07-28-CLOZUK/named-figure/calls-in-individual-scz-gwas-loci-coding.pdf}
\caption{
Somatic SNVs at individual loci of schizophrenia GWAS
}
\label{fig:indiv-scz-gwas-loci}
\end{figure}

\end{document}
