\documentclass[letterpaper]{article}
%\documentclass[12pt,letterpaper]{article}
%\setlength{\textwidth}{480pt}
%\setlength{\textheight}{630pt}
%\setlength{\voffset}{0pt}

\usepackage{amsmath, geometry, graphicx}
\usepackage{natbib}
%\usepackage{float}
\bibliographystyle{plainnat}

% https://tex.stackexchange.com/questions/6758/how-can-i-create-a-bibliography-like-a-section
%\usepackage{etoolbox}
%\patchcmd{\thebibliography}{\section*}{\section}{}{}

\pagestyle{plain}

\title{Involvement of Brain Somatic Mosaicism in Schizophrenia}

\author{Attila Jones\(^{1,2}\), Chaggai Rosenbluh\(^{1}\), ..., Andrew Chess\(^{1,2,3,4,\ast}\)}

\date{Icahn School of Medicine at Mount Sinai (ISMMS)}

\begin{document}

\maketitle

\begin{description}
\item[1] Department of Cell, Developmental and Regenerative Biology, ISMMS 
\item[2] Institute for Genomics and Multiscale Biology, Department of Genetics and Genomic Sciences, ISMMS 
\item[3] Neuroscience Program, The Graduate School of Biomedical Sciences, ISMMS 
\item[4] Division of Psychiatric Genomics, Department of Psychiatry, ISMMS 
%\item[\(\S\)] full list of consortium members appears in the Author
Information section
\item[\(\ast\)] correspondence: andrew.chess@mssm.edu 
\end{description}

\clearpage

\section*{Abstract}

The abstract

\section*{Introduction}

Schizophrenia is a psychiatric disorder, which... heritability: 0.8 ...\citep{Kahn2015}.  
GWAS (reviewed by \cite{Visscher2017}) has been essential in clarifying the
genetic of architecture of schizophrenia~\citep{Ripke2014,Pardinas2018} and other psychiatric disorders
\citep{Sullivan2012}: they are polygenic and much of the heritability is due to
a large set of common germline variants, typically SNPs, with
small effect size.  GWAS also revealed that psychiatric disorders
typically share much heritability with each
other~\citep{Consortium2009,PsychiatricGenomicsConsortium2013} and to a lesser extent with
some neurological disorders~\citep{Consortium2018} indicating shared
pathological mechanisms and supporting the neurodevelopmental hypothesis of
schizophrenia~\citep{Nour2015}.

Beyond common SNPs other classes of genetic variation have also been shown to
carry substantial risk for schizophrenia: rare CNVs~\citep{Rees2014}, rare germline
SNVs~\citep{Purcell2014,Singh2017} as well as de novo
SNVs~\citep{Fromer2014,Rees2020}.  These studies confirmed the polygenicity of
schizophrenia and found excess mutational burden in several functional
gene sets such as synaptic and voltage gated calcium channel and FMRP target
genes\citep{Fromer2014,Purcell2014} ... and loss of function intolerant
genes~\citep{Rees2020,Singh2017} but found no evidence for increased genome
wide germline single nucleotide mutation rate in schizophrenics.

Despite much progress translation of the results of the aforementioned studies
into effective medical diagnostic and treatment methods for schizophrenia is
still lacking~\citep{Breen2016,Foley2017}. Increasing sample sizes, especially
for currently underrepresented ethnic groups, is expected to uncover new
potential risk genes and therapeutic targets~\citep{Visscher2017}, but the difficulty of interpreting
polygenic signals from SNP arrays will persist~\citep{Boyle2017}.  Rare and de
novo variants detected by sequencing afford somewhat easier interpretation.
However, besides germline variants a number of distinct genetic and epigenetic
processes are hypothesized to play a role in schizophrenia and psychiatric
disorders~\citep{PsychENCODEConsortium2015}.

Somatic variants have been called the dark matter of psychiatric
genetics~\citep{Insel2014}.  Their systematic, genome-wide study has just begun
with the announcement of the Brain Somatic Mosaicism
project~\citep{McConnell2017}.  Within that project we and others recently
developed new best practices and tools for calling somatic variants from brain
samples based on bulk tissue sequencing data~\citep{Wang2021}.  In the present
work, for the first time, we apply these analytical tools to deep, whole
genome sequencing data from the DLPFC of \(m\) schizophrenic and \(n\) control
individuals.

\section*{Results}

\subsection*{Somatic SNV callsets}

We extracted NeuN+ nuclei from postmortem samples of the DLPFC of 40 control
and 61 schizophrenic individuals and subjected them to short read DNA
sequencing at 50--540\(\times\) coverage.  Next we called somatic SNVs using a
pipeline developed and validated by the BSMN consortium~\citep{Wang2021}.  The
genome-wide somatic SNV call rate was similar for control and schizophrenic
samples (median around 40) and showed great variability, ranging 0 to 100 with
the exception of an outlier schizophrenic sample reaching \(\approx 200\)
(Fig~\ref{fig:genomewide-calls}A).

\begin{figure}
A\includegraphics[scale=0.6]{../../notebook/2020-06-10-annotate-explore-variants/named-figure/numcalls-Dx.pdf}
B\includegraphics[scale=0.6]{../../notebook/2020-12-09-walsh-data/named-figure/ncalls-depth-chess-walsh.pdf}
\caption{Genome wide somatic SNV call rate for Control and Schizophrenic samples.  TODO:
	add Harvard control samples, remove ASD}
\label{fig:genomewide-calls}
\end{figure}

The finding that somatic SNV call rate is unaltered in schizophrenics was
expected based on similar results for rare germline and de-novo
SNVs~\citep{Purcell2014,Singh2017,Fromer2014,Rees2020}.
As for the high variability of somatic SNV call rate we sought 
biological or technical variables explaining that variation.  As
Fig~\ref{fig:genomewide-calls}B
shows sequencing read depth explains much of
the variability in call rate, which indicates that the variant calling
pipeline adapts to lower coverage by applying stronger filters in order to 
maintain high specificity.

The joint distribution of read depth and estimated allele frequency across all
4141 somatic SNVs in our data set showed that increasing read depth allowed
calling SNVs at lower frequencies.  As a result the conditional density of allele frequency is
shifted to the left at higher read depth relative to lower depth.  If we
assume roughly constant somatic point mutation rate along the cell lineages of
our samples~\citep{Rodin2021} then the expected distribution of allele
frequency is quasi exponential.  With decreasing read depth the observed
conditional allele frequency distribution departed increasingly from an
exponential distribution, which indicates that the pipeline's sensitivity for
low allele frequency (below c.a 0.05) somatic SNVs is limited at even high
(\(\approx\) 400) sequencing coverage.

\begin{figure}
\includegraphics[scale=0.6]{../../notebook/2020-08-13-select-vars/named-figure/calls-DP-AF-jointdensity-conddensity.pdf}
\caption{Distribution of allele frequency and read depth.  Note that even though allele frequency is
a continuous variable its estimate is discrete, which is an inherent feature
of the pipeline.   TODO: omit upper
left graph (A) and merge (A) and (B)}
\end{figure}

\subsection*{Enrichment analysis}


Studies of rare germline and de-novo SNVs showed that,
although the genome-wide ``background'' mutation rate is unaltered in
schizophrenia, specific genomic elements and gene sets bear increased burden
of---in other words: are enriched in---SNVs in schizophrenic individuals relative to
controls.  We sought to test enrichment in several genomic elements or gene
sets (see Table below).  To that end we established statistical procedures.


\subsubsection*{Genomic elements and gene sets}

[notebook/2021-01-26-indiv-counts-modeling]


%\begin{table}
\begin{tabular}{lrlr}
category & genomic element, gene set & selected model formula & size \\
\hline
\hline
damaging & coding nonsyn & 1 & 55 \\
& coding syn & 1 + Gender & 38 \\
& missense & None & 0 \\
& stop gain & None & 1 \\
& splice site & None & 1 \\
& coding & 1 & 80 \\
& deleterious (sift) & 1 & 19 \\
& damaging (polyphen) & 1 + DP & 9 \\
\hline
regulartory & TFBS & 1 + AF & 93 \\
& CpG island & 1 + AF + BaseQRankSum & 57 \\
& enhancer & 1 & 99 \\
& genic enhancer & 1 & 22 \\
\hline
constraint & evolutionary cons. (phast) & 1 + DP + AF + Gender & 117 \\
& evolutionary cons. (gerp) & 1 + Dx & 127 \\
& LoF intolerant genes & 1 + Dx & 143 \\
\hline
SCZ gene sets & PSD & 1 & 95 \\
& FMRP-target & 1 + AF & 118 \\
& SCZ de novo & 1 + AF & 97 \\
& Calcium channel & 1 + Dx & 13 \\
& composite & 1 & 126 \\
\hline
GWAS gene sets & ADHD GWAS & 1 + AF + DP & 6 \\
& Anorexia nervosa GWAS & None & 4 \\
& ASD GWAS & 1 + Dx & 27 \\
& Bipolar disorder GWAS & 1 + Dx & 77 \\
& MDD GWAS & 1 & 48 \\
& Schizophrenia GWAS & 1 + Dx & 72 \\
& Psychiatric GWAS & 1 + Dx & 105 \\
\end{tabular}
%\caption{Genomic elements and gene sets.  Size is defined as the number of
%	individuals with at least one somatic SNV call in the genomic element
%	or gene set. TODO: remove ``coding'' since it's redundant give ``coding
%	nonsyn'' and ``coding syn''}
%\end{table}

\subsubsection*{Enrichment of genomic elements and gene sets in somatic SNVs in schizophrenia}

[notebook/2021-01-26-indiv-counts-modeling]

%\begin{figure}
\includegraphics[width=0.9\textwidth]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/enrichment-5-plots_scz.pdf}
%\caption{Enrichment of genomic elements and gene sets in somatic SNVs in schizophrenia}
%\end{figure}

\section*{Discussion}


\section*{Methods}

For a genomic element or gene set  \(S\) we inspected the joint distribution,
across all study subjects, of disease status Dx, two enrichment variables
\(n_S\), \(f_S\), and technical and biological variables that potentially
carry information on those enrichment variables.  The enrichment variable
\(n_S\) is the number of somatic SNVs located in \(S\), whereas the \(f_S =
	n_S/n_\mathrm{genome}\) is the fraction of somatic SNVs in \(S\)
relative to all somatic SNVs in the genome.  The aim was to formally test
enrichment in \(S\) in schizophrenia by regressing the response variables
\(n_S\) or \(f_S\) on the explanatory variable Dx
and other, ``confounding'', variables.

In case of the plot matrix (Fig~S\ref{fig:scattermatrix}) \(S\) is the set of schizophrenia GWAS genes
defined as all genes, including their introns, that are located fully or
partially in schizophrenia GWAS loci (Table S4 in \cite{Pardinas2018}).

Notable observations:
\begin{itemize}
	\item the strongsest dependence was seen between \(n_S\) and \(n_\mathrm{genome}\) as expected
		when the genome-wide background rate of somatic SNVs is
		substantial relative to enrichment in \(S\)
	\item we have already noted the strong dependence between read depth DP and
		\(n_\mathrm{genome}\)
        \item other dependencies were relatively weak
	\item for many control individuals ageOfDeath is below 50 years while for
		most schizophrenic individuals it is above 50.  This is
		because the Mount Sinai dataset contained older individuals
		that were either control or schizophrenic but the Harvard dataset contained
		only younger control individuals
\end{itemize}

\subsubsection*{Modeling the number and fraction of somatic SNV calls in a gene set}

We considered two generalized linear model families depending on the response
variable:
\begin{description}
\item[response is \(f_S\):] logistic family: binomial error distribution with
	logistic link function; see panel (A)
\item[response is \(n_S\):] log-linear family: Poisson error distribution with exponential link function; see panel (B)
\end{description}

Within each family we considered relatively simple or complex models depending
on the set of variables selected as explanatory.  For \(S =\) schizophrenia
GWAS genes, and within the
logistic family, Fig~S\ref{fig:fitted-lines}A shows fitted lines under both \(f_S \sim 1 +
	\mathrm{Dx}\) and \(f_S \sim 1 + \mathrm{Dx} + \mathrm{ageOfDeath}\)
(panel (A)), where \(1\) means intercept.  For the log-linear family panel
(Fig~S\ref{fig:fitted-lines}B)
shows fitted lines under both \(n_S \sim 1 + \mathrm{Dx}\) and \(n_S \sim 1 +
	\mathrm{Dx} + n_\mathrm{genome}\)


\subsubsection*{Variable selection for regression models}

To select the best fitting model within each family we applied forward
variable selection algorithm with the Akaike information criterion (AIC) as
optimality criterion.  (The Bayesian (BIC) information instead of AIC gave the
same results.)

In the case of \(S =\) schizophrenia GWAS genes we see that the first,
thus most informative, variable selected is Dx under the logistic model
(Fig~S\ref{fig:varsel}A).
Any additional variables increased the information criteria and thus lead to
overfitting the data.

(Fig~S\ref{fig:varsel}B) shows the same result under the log-linear model except that the
most informative variable is \(n_\mathrm{genome}\), which was expected by its
strong correlation with \(n_S\).  Nonetheless, Dx is selected also under this
model family.

\subsubsection*{The effect of ageOfDeath and that of Dataset are mediated by disease status Dx}

As we noted one of our two datasets contained relatively young individuals who
were all control.  This leads to the causal-probabilistic graphical model that
\(\mathrm{ageOfDeath} \leftarrow \mathrm{Dataset} \rightarrow \mathrm{Dx}
\rightarrow y\), where \(y\) is the response enrichment variable \(f_S\) or
\(n_S\).  It means that ageOfDeath and Dataset have indirect effects on
enrichment such that these effects are mediated by disease status Dx. If this
model is true then \(Dx\) d-separates \(y\) from Dataset and ageOfDeath so
that \(y\) becomes independent from ageOfDeath and Dataset if we condition on
Dx = control or Dx = schizophrenia.  Fig~S\ref{fig:d-separation}A illustrates
that.  The important consequence is that we should remove ageOfDeath and
Dataset from the set from our models as long as we include Dx.  Failing to do
so may confound our inference on the effect of Dx on enrichment
(Fig~S\ref{fig:d-separation}B).

\subsubsection*{Residual analysis to assess model fit }

We assessed model fit by residual analysis.  There are several types of residuals in use for generalized linear models.
The $r^\ast$ residual is approximately normally distributed even if the error
distribution for the model family is not normal, such as Poisson or
binomial~\citep{davison2003statistical} (Ch10.2 p477).

This residual is defined as follows

$$
r^\ast = r_\mathrm{D} + r_\mathrm{D}^{-1} \log
\frac{r_\mathrm{P}}{r_\mathrm{D}},
$$

where \(r_\mathrm{D}\) is the standardized deviance residual and
\(r_\mathrm{P}\) is the standardized Pearson residual.

We expressed the size of genomic element or gene set as the number of
individuals with at least one somatic SNV in that genomic element or gene set.
We sorted genomic elements and gene sets according to their size and produced
normal Q-Q plots of \(r^\ast\) from top left of Fig~S\ref{fig:residual-QQ} (smallest) to
bottom right (largest).  This indicated poor fit for the smallest genomic
elements and gene sets (top row of plots) but good fit for medium sized and
large ones.

\subsection*{Data availability}

Data and analytical results...

\subsection*{Code availability}

All code developed by A.~Jones is available at:\\
https://github.com/attilagk/BSM-research

\bibliography{library}
%\bibliography{bsm}


\section*{Acknowledgements}

We thank...

\section*{Author Information}

\section*{Author Contributions}

A.J...

\section*{Competing Interests}

None

\section*{Figure Legends}

\subsection*{Figure 1}
%\includegraphics[width=1.0\textwidth]{figures/by-me/commonmind-rna-seq-ms/commonmind-rna-seq-ms.pdf}

\section*{Tables}

\section*{Supplementary Figures}

\setcounter{figure}{0}
\makeatletter 
\renewcommand{\figurename}{Supplementary Figure} % nice
\makeatother

\begin{figure}[p]
\begin{center}
\includegraphics[scale=0.6]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/symmetric-scatter-matrix_scz.pdf}
\end{center}
\caption{
Pairwise joint distributions of response and explanatory variables TODO: fix legend
}
\label{fig:scattermatrix}
\end{figure}

\begin{figure}[p]
\begin{center}
A\includegraphics[scale=0.45]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/scz_gwas_genes-ageOfDeath-binom-fit_scz.pdf}
B\includegraphics[scale=0.45]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/scz_gwas_genes-log10_ncalls-poisson-fit_scz.pdf}
\end{center}
\caption{
Modeling the number (A) and fraction (B) of somatic SNV calls in a gene set
}
\label{fig:fitted-lines}
\end{figure}

\begin{figure}[p]
\begin{center}
A

\includegraphics[scale=0.5]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/scz_gwas_genes-fw-varsel-binom-onlyIC_scz.pdf}

B

\includegraphics[scale=0.5]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/scz_gwas_genes-fw-varsel-pois-onlyIC_scz.pdf}
\end{center}
\caption{
Variable selection for regression models. A) Binomial--logistic
family.  B) Poisson--log-linear family.
}
\label{fig:varsel}
\end{figure}

\begin{figure}[p]
\begin{center}
A\includegraphics[scale=0.45]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/Dx_Dataset_ageOfDeath_scz.pdf}
B\includegraphics[scale=0.45]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/collinearity_Dx_Dataset_ageOfDeath_scz.pdf}
\end{center}
\caption{
The effect of ageOfDeath and that of Dataset are mediated by disease status Dx
}
\label{fig:d-separation}
\end{figure}

\begin{figure}[p]
\includegraphics[width=0.9\textwidth]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/resid-binom.pdf}
\caption{
Residual analysis to assess model fit for genomic elements and gene
sets of increasing size (from top left to bottom right)
}
\label{fig:residual-QQ}
\end{figure}

\end{document}
