---
layout: default
title: Unexpectedly low concordance with CEPH/Utah grandparents
featimg: chess-illumina-venn-chr22-1.png
---

Two callsets are compared to each other.  Both correspond to germline variants in the CEPH/Utah maternal grandmother NA12892.  While the "ChessLab" callset was produced based on our sequencing data and our variant calling workflow (using strelka2), the "Illumina callset was generated based on Illumina's data and workflow (using the first version of strelka).  The callsets are overwhelmingly discordant, which is not explained by the minor differences listed above.  However, there's an additional difference between the two callsets: the reference genome used to indicate positions.  This difference readily explains the discrepancy.

## Introduction

The differences between the *ChessLab* callset and *Illumina* callset concern...

1. the sequencing data
1. aligner
1. the version of the strelka germline caller
1. filtering and other parts of the workflow (besides the caller)

## Preparation

```{r echo=FALSE, warning=FALSE, message=FALSE}
library(lme4)
library(lattice)
library(VennDiagram)
opts_chunk$set(dpi = 144)
opts_chunk$set(out.width = "600px")
opts_chunk$set(dev = c("png", "pdf"))
lattice.options(default.args = list(as.table = TRUE))
lattice.options(default.theme = "standard.theme")
source("../../src/mybcftools.R")
```

The script `do-mystrelka2Germline` (called on `ada`) was called to produce calls for on the whole genome based on the Chess lab Mix3A (NA12892) data.  The following script retains only chromosome 22, records that `PASS`ed strelka2's default filter, and compares the resulting call set to Illumina's callset on NA12892 (with the same filtering: chr22, `PASS`).

```{r engine="bash", eval=FALSE}
./do-concordance
```

```{r cache=TRUE}
tsv <- "/big/results/bsm/2018-05-21-NA12892-variants/chr22/isec/all.tsv"
callsets <- import.mybcftools.isec.tsv.2(tsv, gpnames <- c("ChessLab", "Illumina")) # needs cacheing
```

## Results

The Venn diagram shows that the two callsets are extremely discordant

```{r chess-illumina-venn-chr22}
my.par <- list(main.cex = 1.8, fill = trellis.par.get("superpose.line")$col[1:2], col = "gray", cat.cex = 1.4)
grid.draw(venn.diagram(callsets, NULL, main = "Germline variants in NA12892, chr22", main.cex = my.par$main.cex, col = my.par$col, fill = my.par$fill, cat.cex = my.par$cat.cex))
```

## Punchline

Actually, there is at least on more difference between the *ChessLab* and *Illumina* callsets:

1. the sequencing data
1. aligner
1. the version of the strelka germline caller
1. filtering and other parts of the workflow (besides the caller)
1. **reference genome version**

See below the reference genome version in the *ChessLab* and *Illumina* VCFs:

```{r engine="bash", eval=TRUE}
cd /big/results/bsm/2018-05-21-NA12892-variants/chr22
bcftools view -h NA12892-chesslab.vcf.gz | grep 'reference=file'
bcftools view -h NA12892-illumina.vcf.gz | grep 'reference=file'
```

This difference---the version of the reference genome---seems to explain the great discrepancy.  The next step is to remap positions in the *Illumina* callsets to remove that difference.
