{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'x' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-8f0a1f42cba4>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;36m3\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'x' is not defined"
     ]
    }
   ],
   "source": [
    "3 * x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 403,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on _Record in module vcf.model object:\n",
      "\n",
      "class _Record(builtins.object)\n",
      " |  A set of calls at a site.  Equivalent to a row in a VCF file.\n",
      " |  \n",
      " |  The standard VCF fields CHROM, POS, ID, REF, ALT, QUAL, FILTER,\n",
      " |  INFO and FORMAT are available as properties.\n",
      " |  \n",
      " |  The list of genotype calls is in the ``samples`` property.\n",
      " |  \n",
      " |  Regarding the coordinates associated with each instance:\n",
      " |  \n",
      " |  - ``POS``, per VCF specification, is the one-based index\n",
      " |    (the first base of the contig has an index of 1) of the first\n",
      " |    base of the ``REF`` sequence.\n",
      " |  - The ``start`` and ``end`` denote the coordinates of the entire\n",
      " |    ``REF`` sequence in the zero-based, half-open coordinate\n",
      " |    system (see\n",
      " |    http://genomewiki.ucsc.edu/index.php/Coordinate_Transforms),\n",
      " |    where the first base of the contig has an index of 0, and the\n",
      " |    interval runs up to, but does not include, the base at the\n",
      " |    ``end`` index. This indexing scheme is analagous to Python\n",
      " |    slice notation.\n",
      " |  - The ``affected_start`` and ``affected_end`` coordinates are\n",
      " |    also in the zero-based, half-open coordinate system. These\n",
      " |    coordinates indicate the precise region of the reference\n",
      " |    genome actually affected by the events denoted in ``ALT``\n",
      " |    (i.e., the minimum ``affected_start`` and maximum\n",
      " |    ``affected_end``).\n",
      " |  \n",
      " |    - For SNPs and structural variants, the affected region\n",
      " |      includes all bases of ``REF``, including the first base\n",
      " |      (i.e., ``affected_start = start = POS - 1``).\n",
      " |    - For deletions, the region includes all bases of ``REF``\n",
      " |      except the first base, which flanks upstream the actual\n",
      " |      deletion event, per VCF specification.\n",
      " |    - For insertions, the ``affected_start`` and ``affected_end``\n",
      " |      coordinates represent a 0 bp-length region between the two\n",
      " |      flanking bases (i.e., ``affected_start`` =\n",
      " |      ``affected_end``). This is analagous to Python slice\n",
      " |      notation (see http://stackoverflow.com/a/2947881/38140).\n",
      " |      Neither the upstream nor downstream flanking bases are\n",
      " |      included in the region.\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __cmp__(self, other)\n",
      " |      # For Python 2\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      _Records are equal if they describe the same variant (same position, alleles)\n",
      " |  \n",
      " |  __init__(self, CHROM, POS, ID, REF, ALT, QUAL, FILTER, INFO, FORMAT, sample_indexes, samples=None)\n",
      " |      Initialize self.  See help(type(self)) for accurate signature.\n",
      " |  \n",
      " |  __iter__(self)\n",
      " |  \n",
      " |  __lt__(self, other)\n",
      " |      Return self<value.\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return str(self).\n",
      " |  \n",
      " |  add_filter(self, flt)\n",
      " |  \n",
      " |  add_format(self, fmt)\n",
      " |  \n",
      " |  add_info(self, info, value=True)\n",
      " |  \n",
      " |  genotype(self, name)\n",
      " |      Lookup a ``_Call`` for the sample given in ``name``\n",
      " |  \n",
      " |  get_hets(self)\n",
      " |      The list of het genotypes\n",
      " |  \n",
      " |  get_hom_alts(self)\n",
      " |      The list of hom alt genotypes\n",
      " |  \n",
      " |  get_hom_refs(self)\n",
      " |      The list of hom ref genotypes\n",
      " |  \n",
      " |  get_unknowns(self)\n",
      " |      The list of unknown genotypes\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  aaf\n",
      " |      A list of allele frequencies of alternate alleles.\n",
      " |      NOTE: Denominator calc'ed from _called_ genotypes.\n",
      " |  \n",
      " |  call_rate\n",
      " |      The fraction of genotypes that were actually called.\n",
      " |  \n",
      " |  heterozygosity\n",
      " |      Heterozygosity of a site. Heterozygosity gives the probability that\n",
      " |      two randomly chosen chromosomes from the population have different\n",
      " |      alleles, giving a measure of the degree of polymorphism in a population.\n",
      " |      \n",
      " |      If there are i alleles with frequency p_i, H=1-sum_i(p_i^2)\n",
      " |  \n",
      " |  is_deletion\n",
      " |      Return whether or not the INDEL is a deletion\n",
      " |  \n",
      " |  is_indel\n",
      " |      Return whether or not the variant is an INDEL\n",
      " |  \n",
      " |  is_monomorphic\n",
      " |      Return True for reference calls\n",
      " |  \n",
      " |  is_snp\n",
      " |      Return whether or not the variant is a SNP\n",
      " |  \n",
      " |  is_sv\n",
      " |      Return whether or not the variant is a structural variant\n",
      " |  \n",
      " |  is_sv_precise\n",
      " |      Return whether the SV cordinates are mapped\n",
      " |      to 1 b.p. resolution.\n",
      " |  \n",
      " |  is_transition\n",
      " |      Return whether or not the SNP is a transition\n",
      " |  \n",
      " |  nucl_diversity\n",
      " |      pi_hat (estimation of nucleotide diversity) for the site.\n",
      " |      This metric can be summed across multiple sites to compute regional\n",
      " |      nucleotide diversity estimates.  For example, pi_hat for all variants\n",
      " |      in a given gene.\n",
      " |      \n",
      " |      Derived from:\n",
      " |      \"Population Genetics: A Concise Guide, 2nd ed., p.45\"\n",
      " |      John Gillespie.\n",
      " |  \n",
      " |  num_called\n",
      " |      The number of called samples\n",
      " |  \n",
      " |  num_het\n",
      " |      The number of heterozygous genotypes\n",
      " |  \n",
      " |  num_hom_alt\n",
      " |      The number of homozygous for alt allele genotypes\n",
      " |  \n",
      " |  num_hom_ref\n",
      " |      The number of homozygous for ref allele genotypes\n",
      " |  \n",
      " |  num_unknown\n",
      " |      The number of unknown genotypes\n",
      " |  \n",
      " |  sv_end\n",
      " |      Return the end position for the SV\n",
      " |  \n",
      " |  var_subtype\n",
      " |      Return the subtype of variant.\n",
      " |      \n",
      " |      - For SNPs and INDELs, yeild one of: [ts, tv, ins, del]\n",
      " |      - For SVs yield either \"complex\" or the SV type defined in the ALT\n",
      " |        fields (removing the brackets). E.g.::\n",
      " |      \n",
      " |            <DEL>       -> DEL\n",
      " |            <INS:ME:L1> -> INS:ME:L1\n",
      " |            <DUP>       -> DUP\n",
      " |      \n",
      " |      The logic is meant to follow the rules outlined in the following\n",
      " |      paragraph at:\n",
      " |      \n",
      " |      http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41\n",
      " |      \n",
      " |      \"For precisely known variants, the REF and ALT fields should contain\n",
      " |      the full sequences for the alleles, following the usual VCF conventions.\n",
      " |      For imprecise variants, the REF field may contain a single base and the\n",
      " |      ALT fields should contain symbolic alleles (e.g. <ID>), described in more\n",
      " |      detail below. Imprecise variants should also be marked by the presence\n",
      " |      of an IMPRECISE flag in the INFO field.\"\n",
      " |  \n",
      " |  var_type\n",
      " |      Return the type of variant [snp, indel, unknown]\n",
      " |      TO DO: support SVs\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __hash__ = None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
