{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Various GWAS genesets are tested for enrichment in SCZ somatic variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%reload_ext autoreload\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import scipy.stats as stats\n",
    "from bsmcalls import SNPnexus\n",
    "from bsmcalls import operations\n",
    "from matplotlib import pyplot as plt\n",
    "import attila_utils\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = SNPnexus.load_data('/home/attila/projects/bsm/results/2020-09-07-annotations/annotated-calls.p')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read autism GWAS loci from the GWAS Catalog to a DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwaspath = '/home/attila/projects/bsm/resources/GWAS/autism-spectrum-disorder.tsv'\n",
    "asdloci = pd.read_csv(gwaspath, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will select the study ``Identification of common genetic risk variants for autism spectrum disorder.'' (PMID [30804558](www.ncbi.nlm.nih.gov/pubmed/30804558))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR_ID</th>\n",
       "      <th>CHR_POS</th>\n",
       "      <th>REPORTED GENE(S)</th>\n",
       "      <th>MAPPED_GENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>1</td>\n",
       "      <td>96096246</td>\n",
       "      <td>LOC102723661, PTBP2</td>\n",
       "      <td>LINC02790 - RNU1-130P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>1</td>\n",
       "      <td>96513405</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>EEF1A1P11 - RN7SL831P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>1</td>\n",
       "      <td>98627228</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>LINC01776 - AC095031.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>1</td>\n",
       "      <td>104249635</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>AC092506.1 - AL591888.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>1</td>\n",
       "      <td>193533479</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>AL136456.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>20</td>\n",
       "      <td>21267478</td>\n",
       "      <td>KIZ, XRN2, NKX2-2, NKX2-4</td>\n",
       "      <td>ZNF877P - AL117332.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>5</td>\n",
       "      <td>104676602</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AC099520.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>14</td>\n",
       "      <td>103551616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RNU7-160P - AL139300.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>1</td>\n",
       "      <td>72263459</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEGR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>4</td>\n",
       "      <td>42121711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BEND4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHR_ID    CHR_POS           REPORTED GENE(S)              MAPPED_GENE\n",
       "94       1   96096246        LOC102723661, PTBP2    LINC02790 - RNU1-130P\n",
       "95       1   96513405                 Intergenic    EEF1A1P11 - RN7SL831P\n",
       "96       1   98627228                 Intergenic   LINC01776 - AC095031.1\n",
       "97       1  104249635                 Intergenic  AC092506.1 - AL591888.1\n",
       "98       1  193533479                 Intergenic               AL136456.1\n",
       "..     ...        ...                        ...                      ...\n",
       "151     20   21267478  KIZ, XRN2, NKX2-2, NKX2-4     ZNF877P - AL117332.1\n",
       "152      5  104676602                        NaN               AC099520.1\n",
       "153     14  103551616                        NaN   RNU7-160P - AL139300.2\n",
       "154      1   72263459                        NaN                    NEGR1\n",
       "155      4   42121711                        NaN                    BEND4\n",
       "\n",
       "[62 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asdloci = asdloci.loc[asdloci['PUBMEDID'] == 30804558]\n",
    "selcols = ['CHR_ID', 'CHR_POS', 'REPORTED GENE(S)', 'MAPPED_GENE']\n",
    "asdloci[selcols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'C8orf74',\n",
       " 'CADPS',\n",
       " 'KIZ',\n",
       " 'KMT2E',\n",
       " 'LOC102723661',\n",
       " 'MACROD2',\n",
       " 'MROH5',\n",
       " 'NKX2-2',\n",
       " 'NKX2-4',\n",
       " 'PINX1',\n",
       " 'PTBP2',\n",
       " 'SOX7',\n",
       " 'SRPK2',\n",
       " 'XRN2'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reported = set(asdloci['REPORTED GENE(S)'].dropna().str.split(', ').sum())\n",
    "reported.remove('Intergenic')\n",
    "reported"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "79 genes\n",
      "{'LINC01776', 'EEF1A1P11', 'AL589740.1', 'PAUPAR', 'MAPT', 'ESRRB', 'AC021192.2', 'LINC02790', 'AL117332.1', 'SERPINA2', 'PTBP2', 'XRN2', 'EGR2', 'RNU1-130P', 'BEND4', 'AL139300.2', 'AC095031.1', 'AC120036.1', 'RSU1', 'CADPS', 'GALNT10', 'C8orf74', 'GALNT1', 'AC013287.1', 'NRGN', 'U95743.1', 'OLFM4', 'AC090987.1', 'RN7SL831P', 'NEGR1', 'MROH5', 'AL133270.1', 'KCNN2', 'NKX2-4', 'AL391117.1', 'AC092506.1', 'RSRC1', 'RNU7-160P', 'SRPK2', 'AL591888.1', 'AC099520.1', 'ITIH4', 'ZNF568', 'AL022724.3', 'AC026320.3', 'AC005070.3', 'NKX2-2', 'ZNF877P', 'LINC02058', 'AL032822.1', 'ASAP1', 'PINX1', 'AL136456.1', 'GUCY1A2', 'AC025839.1', 'PCDH9', 'LINC00461', 'KMT2E', 'SGO1-AS1', 'KANSL1', 'ADTRP', 'AL035078.4', 'RBFOX1', 'SOX7', 'MACROD2', 'AC022387.1', 'FGF12', 'SERPINA1', 'KIZ', 'GABBR1', 'ITIH3', 'LOC102723661', 'LINC02210-CRHR1', 'AL139093.1', 'RP1L1', 'PKP4', 'U91319.1', 'NEDD4L', 'SORCS3'}\n"
     ]
    }
   ],
   "source": [
    "mappeds = asdloci['MAPPED_GENE'].dropna()\n",
    "commas = mappeds.str.match('^.*, .*$')\n",
    "mappeds1 = mappeds.loc[commas]\n",
    "mapped1 = set(mappeds1.str.split(', ').sum())\n",
    "mappeds2 = mappeds.loc[~ commas]\n",
    "mapped2 = set(mappeds2.str.split(' - ').sum())\n",
    "mapped = mapped2.union(mapped1)\n",
    "asd_gwas_genes = mapped.union(reported)\n",
    "print(len(asd_gwas_genes), 'genes')\n",
    "print(asd_gwas_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us read in the set of SCZ GWAS genes!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "clozukpath = '/home/attila/projects/bsm/resources/CLOZUK/supp-table-4.csv'\n",
    "scz_gwas_genes = operations.get_geneset(df=pd.read_csv(clozukpath, skiprows=7), col='Gene(s) tagged')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/attila/anaconda3/lib/python3.7/site-packages/pandas/core/generic.py:3889: PerformanceWarning: dropping on a non-lexsorted multi-index without a level parameter may impact performance.\n",
      "  obj = obj._drop_axis(labels, axis, level=level, errors=errors)\n",
      "/home/attila/anaconda3/lib/python3.7/site-packages/scipy/stats/stats.py:5982: RuntimeWarning: invalid value encountered in true_divide\n",
      "  terms = (f_obs.astype(np.float64) - f_exp)**2 / f_exp\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Control</th>\n",
       "      <th>SCZ</th>\n",
       "      <th>ASD</th>\n",
       "      <th>chisq stat</th>\n",
       "      <th>chisq p</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feature</th>\n",
       "      <th>Query</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">near_gens_Annotation</th>\n",
       "      <th>coding nonsyn</th>\n",
       "      <td>23</td>\n",
       "      <td>43</td>\n",
       "      <td>23</td>\n",
       "      <td>4.480340</td>\n",
       "      <td>0.106440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coding syn</th>\n",
       "      <td>16</td>\n",
       "      <td>20</td>\n",
       "      <td>15</td>\n",
       "      <td>1.177849</td>\n",
       "      <td>0.554924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>missense</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop-gain</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.887477</td>\n",
       "      <td>0.236044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intronic (splice_site)</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.582797</td>\n",
       "      <td>0.453210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_Predicted Function</th>\n",
       "      <th>coding</th>\n",
       "      <td>39</td>\n",
       "      <td>64</td>\n",
       "      <td>40</td>\n",
       "      <td>3.794387</td>\n",
       "      <td>0.149989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">sift_Prediction</th>\n",
       "      <th>Deleterious</th>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>1.315056</td>\n",
       "      <td>0.518131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Deleterious - Low Confidence</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.429823</td>\n",
       "      <td>0.806613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">polyphen_Prediction</th>\n",
       "      <th>Probably Damaging</th>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1.080625</td>\n",
       "      <td>0.582566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Possibly Damaging</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2.470275</td>\n",
       "      <td>0.290795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tfbs_TFBS Name</th>\n",
       "      <th>any</th>\n",
       "      <td>48</td>\n",
       "      <td>78</td>\n",
       "      <td>50</td>\n",
       "      <td>4.120011</td>\n",
       "      <td>0.127453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phast_Score</th>\n",
       "      <th>any</th>\n",
       "      <td>102</td>\n",
       "      <td>174</td>\n",
       "      <td>141</td>\n",
       "      <td>1.591145</td>\n",
       "      <td>0.451323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gerp_Element RS Score</th>\n",
       "      <th>any</th>\n",
       "      <td>140</td>\n",
       "      <td>262</td>\n",
       "      <td>208</td>\n",
       "      <td>5.012216</td>\n",
       "      <td>0.081585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cpg_CpG Island</th>\n",
       "      <th>any</th>\n",
       "      <td>26</td>\n",
       "      <td>47</td>\n",
       "      <td>19</td>\n",
       "      <td>9.614759</td>\n",
       "      <td>0.008169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">near_gens_Overlapped Gene</th>\n",
       "      <th>SCZ GWAS genes</th>\n",
       "      <td>20</td>\n",
       "      <td>66</td>\n",
       "      <td>42</td>\n",
       "      <td>10.800565</td>\n",
       "      <td>0.004515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASD GWAS genes</th>\n",
       "      <td>2</td>\n",
       "      <td>20</td>\n",
       "      <td>7</td>\n",
       "      <td>11.912735</td>\n",
       "      <td>0.002589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASD reported GWAS genes</th>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>9.496785</td>\n",
       "      <td>0.008666</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         Control  SCZ  ASD  \\\n",
       "Feature                    Query                                             \n",
       "near_gens_Annotation       coding nonsyn                      23   43   23   \n",
       "                           coding syn                         16   20   15   \n",
       "                           missense                            0    0    0   \n",
       "                           stop-gain                           1    0    0   \n",
       "                           intronic (splice_site)              0    1    0   \n",
       "ensembl_Predicted Function coding                             39   64   40   \n",
       "sift_Prediction            Deleterious                         8    9    6   \n",
       "                           Deleterious - Low Confidence        1    1    2   \n",
       "polyphen_Prediction        Probably Damaging                   3    5    2   \n",
       "                           Possibly Damaging                   2    2    0   \n",
       "tfbs_TFBS Name             any                                48   78   50   \n",
       "phast_Score                any                               102  174  141   \n",
       "gerp_Element RS Score      any                               140  262  208   \n",
       "cpg_CpG Island             any                                26   47   19   \n",
       "near_gens_Overlapped Gene  SCZ GWAS genes                     20   66   42   \n",
       "                           ASD GWAS genes                      2   20    7   \n",
       "                           ASD reported GWAS genes             0    6    0   \n",
       "\n",
       "                                                         chisq stat   chisq p  \n",
       "Feature                    Query                                               \n",
       "near_gens_Annotation       coding nonsyn                   4.480340  0.106440  \n",
       "                           coding syn                      1.177849  0.554924  \n",
       "                           missense                             NaN       NaN  \n",
       "                           stop-gain                       2.887477  0.236044  \n",
       "                           intronic (splice_site)          1.582797  0.453210  \n",
       "ensembl_Predicted Function coding                          3.794387  0.149989  \n",
       "sift_Prediction            Deleterious                     1.315056  0.518131  \n",
       "                           Deleterious - Low Confidence    0.429823  0.806613  \n",
       "polyphen_Prediction        Probably Damaging               1.080625  0.582566  \n",
       "                           Possibly Damaging               2.470275  0.290795  \n",
       "tfbs_TFBS Name             any                             4.120011  0.127453  \n",
       "phast_Score                any                             1.591145  0.451323  \n",
       "gerp_Element RS Score      any                             5.012216  0.081585  \n",
       "cpg_CpG Island             any                             9.614759  0.008169  \n",
       "near_gens_Overlapped Gene  SCZ GWAS genes                 10.800565  0.004515  \n",
       "                           ASD GWAS genes                 11.912735  0.002589  \n",
       "                           ASD reported GWAS genes         9.496785  0.008666  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "querydict = {'near_gens_Annotation': ['coding nonsyn', 'coding syn', 'missense', 'stop-gain', 'intronic (splice_site)'],\n",
    "             'ensembl_Predicted Function': ['coding'],\n",
    "             'sift_Prediction': ['Deleterious', 'Deleterious - Low Confidence'],\n",
    "             'polyphen_Prediction': ['Probably Damaging', 'Possibly Damaging'],\n",
    "             'tfbs_TFBS Name': None,\n",
    "             'phast_Score': None,\n",
    "             'gerp_Element RS Score': None,\n",
    "             'cpg_CpG Island': None,\n",
    "             'near_gens_Overlapped Gene': {'SCZ GWAS genes': scz_gwas_genes,\n",
    "                                           'ASD GWAS genes': asd_gwas_genes,\n",
    "                                           'ASD reported GWAS genes': reported,\n",
    "                                          },\n",
    "            }\n",
    "results = operations.multiquery(querydict, data, do_sum=False, do_sort=False)\n",
    "summary = operations.summarize_query_results(results, data, chisq=True, margin=False)\n",
    "summary#.style.bar(subset='chisq stat')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"shell_port\": 41725,\n",
      "  \"iopub_port\": 36167,\n",
      "  \"stdin_port\": 60409,\n",
      "  \"control_port\": 40337,\n",
      "  \"hb_port\": 52123,\n",
      "  \"ip\": \"127.0.0.1\",\n",
      "  \"key\": \"96401c45-fa6385b36e8d4ec92ba12e38\",\n",
      "  \"transport\": \"tcp\",\n",
      "  \"signature_scheme\": \"hmac-sha256\",\n",
      "  \"kernel_name\": \"\"\n",
      "}\n",
      "\n",
      "Paste the above JSON into a file, and connect with:\n",
      "    $> jupyter <app> --existing <file>\n",
      "or, if you are local, you can connect with just:\n",
      "    $> jupyter <app> --existing kernel-3022da89-61f5-46b7-b6e2-9f710614c66e.json\n",
      "or even just:\n",
      "    $> jupyter <app> --existing\n",
      "if this is the most recent Jupyter kernel you have started.\n"
     ]
    }
   ],
   "source": [
    "%connect_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
