\documentclass{beamer}
%\documentclass[handout]{beamer}

% language settings
%\usepackage{fontspec, polyglossia}
%\setdefaultlanguage{magyar}

% common packages
\usepackage{amsmath, multimedia, hyperref, color, multirow}
%\usepackage{graphicx}

% TikZ
\usepackage{tikz}
%\usetikzlibrary{arrows.meta, decorations.pathmorphing, decorations.pathreplacing, shapes.geometric,mindmap}
%\usetikzlibrary{shapes.geometric,fadings,bayesnet}

% beamer styles
\mode<presentation>{
\usetheme{Boadilla}
%\usetheme{Antibes}
%\usecolortheme{beaver}
%\usecolortheme{seahorse}
%\usefonttheme{structureitalicserif}
\setbeamercovered{transparent}
}
\setbeamertemplate{blocks}[rounded][shadow=true]
\AtBeginSubsection[]{
  \begin{frame}<beamer>{Contents}
    \tableofcontents[currentsection,currentsubsection]
  \end{frame}
}
%\useoutertheme[]{tree}

% title, etc
\title{Proposed Benchmark Study}
\subtitle{Biological and technical limitations of somatic variant calling}
\author{Brain Somatic Mosaicism project}
\date{}

\begin{document}

\maketitle

\section{Motivation}

\begin{frame}{Terms}
\small
\begin{tabular}{c|l}
variant & a SNV or small indel, germline or somatic \\
allele frequency & of a somatic variant in a tissue sample \\
paired sample & NeuN+ DLPFC nuclei and reference tissue \\
data set & NGS data from a paired sample \\
caller & a classifier and scoring method \\
filter & a tunable classifier \\
%compound caller & a combination of a \emph{single} caller and multiple filters \\
var.~calling workflow & data set(s) + caller(s) + filters \\
\end{tabular}

\vfill

\begin{tabular}{c|c|c}
 & caller & filter \\
\hline
local info & base (including gap) & caller/quality score, depth,... \\
stat.~model & yes & no \\
yields & unfiltered call set\footnote{VCF record} & final call set \\
designed for & sensitivity & specificity \\
\end{tabular}
\end{frame}

\begin{frame}{The ``Reference tissue project''}
\begin{itemize}
\item a paired sample shared by labs
\item NGS technology and compound or combined caller specific to each lab
\end{itemize}
\end{frame}

\begin{frame}{Example: the combined caller of U.~Michigan}
\begin{center}
\includegraphics[height=0.8\textheight]{figures/UM-Pipeline-3.pdf}
\end{center}
\end{frame}

\begin{frame}{Amplicon-seq validation}
{validation attempted for 7 of 11 candidate variants}
\begin{center}
\includegraphics[height=0.9\textheight]{figures/BSMN_data_call_101617-summary.pdf}
\end{center}
\end{frame}

\begin{frame}{Concordance of call sets}
\begin{center}
\includegraphics[height=0.7\textheight]{figures/tbae-call-set-comparison.png}
\end{center}
\end{frame}

\begin{frame}{Conclusion: low accuracy workflows \(\Rightarrow\) need
optimization}
\begin{center}
{\Large Two approaches to optimization}
\vfill
\begin{tabular}{r|c|c}
 & 1. & 2. \\
\hline
known true variants & benchmark data & targeted re-seq. \\
\# of true variants & high & low \\
allows machine learning? & yes & no \\
bias & prior knowledge & ascertainment \\
ideal for & initial optim. & final validation \\
\end{tabular}
\end{center}
\end{frame}

\begin{frame}[label=umworkflow]{Optimization without machine learning}
{Example: U.~Michigan workflow}
\begin{itemize}
\item 
exhaustive search
\begin{itemize}
\item \(\ge 9\) tuning parameters
\item if only 2 settings for each \(\Rightarrow 2^9 = 512\) call sets
\item \(512 \gg\) what targeted re-sequencing allows
\end{itemize}
\item<2-> inspection
\begin{itemize}
\item 5 out of 9 are annotation/quality based filters
\item 5 additional dimensions when inspecting paired alignments
\end{itemize}
\end{itemize}
\end{frame}

\section{Proposal}

\begin{frame}[label=plan]{Plan}
\begin{enumerate}
\item<1> obtain benchmark data: several data sets
\begin{itemize}
\item various sequencing technology and depth 
\item no gold standard for BSM \(\Rightarrow\) must be generated
\item little prior knowledge \(\Rightarrow\) various BSM models
\end{itemize}
\item<2> optimize workflow parameters
\begin{itemize}
\item for each data set
\item machine learning: VariantMetaCaller 
\end{itemize}
\item<3> estimate sensitivity-specificity 
\begin{itemize}
\item demonstrate improvement by VariantMetaCaller
\item FDR (specificity) control
\item dependence on BSM model
\item dependence on seq.~technology and depth
\end{itemize}
\item<4> validation of VariantMetaCaller call set with re-sequencing
\begin{itemize}
\item collaboration within BSM network?
\end{itemize}
\end{enumerate}
\end{frame}

\againframe<1>{plan}

\begin{frame}{Relative depth of NeuN+ and reference}
\includegraphics[width=0.2\textwidth]{/home/attila/figures/from-others/Hofmann-2017-Fig3legend.png}
\includegraphics[width=0.4\textwidth]{/home/attila/figures/from-others/Hofmann-2017-Fig3a.png}
\includegraphics[width=0.4\textwidth]{/home/attila/figures/from-others/Hofmann-2017-Fig3c.png}
\vfill
\footnotesize{Hofmann et al 2017 BMC Bioinformatics}
\end{frame}

\againframe<1>{plan}

\begin{frame}{How many mitoses before and after neurulation?}
\begin{itemize}
\item \(86.1 \pm 8.1\) billion NeuN+ cells\footnote{Azevedo et al 2009 J Comp
Neur}
\item taking \(\log_2 \Rightarrow\) avg: 36.33 mitoses (error: 36.17, 36.46)
\item not accounted for:
\begin{itemize}
\item cell death
\item mitoses before and after neurulation  
\end{itemize}
\item mutation rate \(=\)?
\end{itemize}
\end{frame}

\begin{frame}{Mutation rate: latest estimate}
\includegraphics[width=1.0\textwidth]{/home/attila/figures/from-others/milholland-2017-fig1b.png}
\vfill
\footnotesize{Milholland et al 2017 Nat Commun}
\end{frame}

\begin{frame}{Human embryogenesis}
\includegraphics[height=0.8\textheight]{/home/attila/figures/from-others/human-embryogenesis-wikipedia.png}
\end{frame}

\begin{frame}{Movies on human embryonic development}
\begin{description}
\item[2nd week]
https://www.youtube.com/watch?v=bIdJOiXpp9g
\item[gastrulation]
https://www.youtube.com/watch?v=3AOoikTEfeo
\item[neurulation] 
https://www.youtube.com/watch?v=lGLexQR9xGs
\end{description}
\end{frame}

\againframe<1>{plan}

\againframe{umworkflow}

\begin{frame}{Combination benefits from weighting}
\includegraphics[width=0.8\textwidth]{/home/attila/figures/from-others/Hofmann-2017-Fig4cd.png}
\vfill
\footnotesize{Hofmann et al 2017 BMC Bioinformatics}
\end{frame}

\begin{frame}{VariantMetaCaller}
\includegraphics[height=0.7\textheight]{/home/attila/figures/from-others/vmc-fig1.png}
\vfill
\footnotesize{Gezsi et al 2015 BMC Genomics}
\end{frame}

\end{document}


\begin{columns}[t]
\begin{column}{0.5\textwidth}

\end{column}

\begin{column}{0.5\textwidth}

\end{column}
\end{columns}

