\documentclass[usenames,dvipsnames]{beamer}
%\documentclass[handout]{beamer}

% language settings
%\usepackage{fontspec, polyglossia}
%\setdefaultlanguage{magyar}

% common packages
\usepackage{amsmath, multimedia, url, hyperref, xcolor, multirow}
%\usepackage{graphicx}

% TikZ
\usepackage{tikz}
%\usetikzlibrary{arrows.meta, decorations.pathmorphing, decorations.pathreplacing, shapes.geometric,mindmap}
%\usetikzlibrary{shapes.geometric,fadings,bayesnet}

% beamer styles
\mode<presentation>{
\usetheme{Warsaw}
%\usetheme{Antibes}
%\usecolortheme{beaver}
\usecolortheme{seahorse}
%\usefonttheme{structureitalicserif}
\setbeamercovered{transparent}
}
\setbeamertemplate{blocks}[rounded][shadow=true]
\AtBeginSubsection[]{
  \begin{frame}<beamer>{Contents}
    \tableofcontents[currentsection,currentsubsection]
  \end{frame}
}
%\useoutertheme[]{tree}

% title, etc
\title{Somatic SNVs from WGS data in Schizophrenic Brains}
%\subtitle{A subtitle may be shorter and more technical}
\author{Attila Jones, Chaggai Rosenbluh, Andy Chess}
\date{Mount Sinai School of Medicine}

\begin{document}

\maketitle

%\section{Introduction}

\begin{frame}{Genetic risk in schizophrenia}
\begin{columns}[t]
\begin{column}{0.5\textwidth}
\begin{itemize}
\item germline variants
\begin{itemize}
\item 80\% heritability
\item thousands of risk variants
\end{itemize}
\item de novo variants
\begin{itemize}
        \item synaptic gene networks
	\item exome investigated only%\footnote{Nature 2014, 506, 179-84}
\end{itemize}
\end{itemize}
\end{column}
\begin{column}{0.5\textwidth}
\vfill
This study: somatic variants
\begin{enumerate}
\item role in schizophrenia
\item whole genome
%\item relative risk: germline vs somatic variants	
\end{enumerate}
\end{column}
\end{columns}
%\includegraphics[scale=0.5]{from-others/fromer-denovo-figA.png}
\end{frame}

%\section{Our Study: Data}

\begin{frame}{Our study}
\begin{center}
61 SCZ and 25 Control samples, bulk WGS
\end{center}
\includegraphics[width=1.0\textwidth]{from-others/bsm-marker-fig1.jpg}
\end{frame}

%\begin{frame}{Cortical bulk WGS data sets in BSMN}
%\includegraphics[width=1.0\textwidth]{../../notebook/2020-09-28-wgs-data-other-groups/named-figure/bsmn-case-control-ssizes.pdf}
%\end{frame}

\begin{frame}{Calling somatic SNVs and small indels}
\footnotesize
Taejeong Bae et al \\
\url{https://github.com/bsmn/bsmn-pipeline}
\vfill
\includegraphics[width=1.0\textwidth]{from-others/bsmn-pipeline-calling.png}

\includegraphics[scale=0.5]{../../results/2020-06-05-filtered-callsets/aws-ec2-master-cpu-usage.png}
\end{frame}

\begin{frame}{Call sets}
\begin{columns}[t]
\begin{column}{0.7\textwidth}

\includegraphics[scale=0.5]{../../notebook/2020-06-10-annotate-explore-variants/named-figure/num-indiv-calls-Dx.pdf}
\end{column}

\begin{column}{0.5\textwidth}

\includegraphics[scale=0.5]{../../notebook/2020-06-10-annotate-explore-variants/named-figure/eCDF-calls.pdf}
\end{column}
\end{columns}
\end{frame}

\begin{frame}[label=numcalls-depth]{Number of calls per sample}
\includegraphics[scale=0.5]{../../notebook/2020-08-19-indiv-based-analysis/named-figure/numcalls-depth-Dx.pdf}
\end{frame}

\begin{frame}{Occurrence of calls}{Interplay between read depth and allele frequency}
\includegraphics<1>[scale=0.5]{../../notebook/2020-08-13-select-vars/named-figure/DP-AF-jointdensity-calls.pdf}
\includegraphics<2>[scale=0.5]{../../notebook/2020-08-13-select-vars/named-figure/DP-AF-jointdensity-conddensity.pdf}
\end{frame}

%\section{Results}

\begin{frame}{Annotating somatic variants with features}
\includegraphics[width=1.0\textwidth]{../../results/2020-09-07-annotations/named-figure/MSSM_224-NeuN_pl-karyotype-of-genomic-consequences.png}
\end{frame}

\begin{frame}{Call features}
\begin{columns}[t]
\begin{column}{0.5\textwidth}
	Data specific
\begin{itemize}
\item VCF fields: DP, AF, ...
\item clinical info: Dx, ethnicity, ...
\end{itemize}
\end{column}
\begin{column}{0.5\textwidth}
	Knowledge bases
\begin{itemize}
\item \url{www.szdb.org}
\begin{itemize}
\item GWAS loci
\item SCZ-CNV loci
\end{itemize}
%\item Roadmap Epigenomics
%\begin{itemize}
%\item chromatin state in DLPC; 15 state HMM
%\end{itemize}
\item \url{www.snp-nexus.org}
\end{itemize}
\end{column}
\end{columns}
\end{frame}

\begin{frame}{Call features: SNPnexus}
	\tiny
	\begin{description}
		\item[Gene annotation systems]  Ensembl, RefSeq, UCSC, CCDS, coding, intronic, non-coding
                \item[Protein deleterious effects] SIFT and PolyPhen
                \item[Population data] gnomAD, 1000Genomes, HapMap
                \item[Regulatory elements] miRBASE, CpG islands, TarBase, microRNAs / snoRNAs / scaRNAs, Ensembl Regulatory Build, ENCODE Project, Roadmap Epigenomics 
                \item[Conservation scores] GERP++, PHAST 
                \item[Disease studies] GAD, COSMIC, NHGRI-GWAS, ClinVar
                \item[Non-coding scoring] fitCons, EIGEN, FATHMM GWAVA, DeepSEA, ReMM, CADD, FunSeq2 
                \item[Structural variations] Gain, Loss, Gain+Loss, Duplication, Deletion, Insertion, Complex, Inversion, Tandem duplication, Novel sequence insertion, Mobile element insertion, Sequence alteration 
                \item[Pathway Analysis] Reactome pathways 
                \item[Biological / Clinical Interpretation] Cancer genome interpreter 
	\end{description}
\end{frame}

\begin{frame}{Dx: no association with schizophrenia GWAS loci}
\includegraphics[scale=0.7]{../../notebook/2020-06-10-annotate-explore-variants/named-figure/gwas-Dx.pdf}
\end{frame}

%\begin{frame}{Dx: weak association with transcription start sites}
%\includegraphics[scale=0.7]{../../notebook/2020-06-10-annotate-explore-variants/named-figure/chromatin-state-DLPFC-Dx-10.pdf}
%\end{frame}

\begin{frame}{Correlation analysis}
\begin{columns}[t]
\begin{column}{0.4\textwidth}

\includegraphics[scale=0.35]{../../notebook/2020-09-07-annotations/named-figure/corr-Dx-snpnexus.pdf}
\end{column}

\begin{column}{0.5\textwidth}

\includegraphics[scale=0.35]{../../notebook/2020-09-07-annotations/named-figure/corr-mat-snpnexus.pdf}
\end{column}
\end{columns}
\end{frame}

\begin{frame}{Statistical modeling challenge}
	\(\mathrm{Dx} = X\beta_x + Y\beta_y + \epsilon\)
\footnotesize
\begin{columns}[t]
\begin{column}{0.6\textwidth}
\begin{center}
Nested observations
\end{center}
\begin{center}
\begin{tabular}{cc|c|c}
& & clinical feat. & genomic feat. \\
sample & variant & \(X\) & \(Y\) \\
\hline
\multirow{4}{*}{\colorbox{Apricot}{\(s_1\)}} & \colorbox{Yellow}{\(v_1\)} & \colorbox{Apricot}{\(x_1\)} & \colorbox{Yellow}{\(y_1\)} \\
& \colorbox{Periwinkle}{\(v_2\)} & \colorbox{Apricot}{\(x_1\)} & \colorbox{Periwinkle}{\(y_2\)} \\
& \colorbox{CarnationPink}{\(v_3\)} & \colorbox{Apricot}{\(x_1\)} & \colorbox{CarnationPink}{\(y_3\)} \\
& \colorbox{CornflowerBlue}{\(v_4\)} & \colorbox{Apricot}{\(x_1\)} & \colorbox{CornflowerBlue}{\(y_4\)} \\
\hline
\multirow{3}{*}{{\colorbox{LimeGreen}{\(s_2\)}}} & \colorbox{Dandelion}{\(v_5\)} & \colorbox{LimeGreen}{\(x_2\)} & \colorbox{Dandelion}{\(y_5\)} \\
& \colorbox{Emerald}{\(v_6\)} & \colorbox{LimeGreen}{\(x_2\)} & \colorbox{Emerald}{\(y_6\)} \\
& \colorbox{Gray}{\(v_7\)} & \colorbox{LimeGreen}{\(x_2\)} & \colorbox{Gray}{\(y_7\)} \\
\end{tabular}
\end{center}
\end{column}

\begin{column}{0.4\textwidth}
\begin{center}
Correlated errors
\end{center}
\includegraphics[width=1.0\columnwidth]{from-others/regrression-correlated-error.png}
\end{column}
\end{columns}
\end{frame}

\begin{frame}[label=regr-models]{Regression models for inference}
\footnotesize
\begin{tabular}{cccc}
approach & regression formula & least squares & effects \\
\hline
A & \(\mathrm{Dx} = X\beta_x + Y\beta_y + \epsilon\) & ordinary & fixed \\
B &\(\mathrm{Dx} = X\beta_x + Y\beta_y + \epsilon\) & generalized & fixed \\
C & \(\mathrm{Dx} = X\gamma_x + Y\beta_y + \epsilon\) & ordinary & mixed\\
D & \(\mathrm{\#calls} = \mathrm{Dx}\beta_\mathrm{Dx} + X\beta_x + a(Y)\beta'_y + \epsilon\) & ordinary & fixed\\
\end{tabular}
\vfill
\begin{flushright}
approach D
\begin{tabular}{cc|c|c}
& & clinical feat. & genomic feat. \\
sample & variant & \(X\) & \(a(Y)\) \\
\hline
\multirow{4}{*}{\colorbox{Apricot}{\(s_1\)}} & \colorbox{Yellow}{\(v_1\)} & \multirow{4}{*}{\colorbox{Apricot}{\(x_1\)}} &
\multirow{4}{*}{\(\colorbox{Apricot}{\(a_1\)} = a(\colorbox{Yellow}{\(y_1\)},
		\colorbox{Periwinkle}{\(y_2\)},
		\colorbox{CarnationPink}{\(y_3\)},
		\colorbox{CornflowerBlue}{\(y_4\)})\)} \\
& \colorbox{Periwinkle}{\(v_2\)} & & \\
& \colorbox{CarnationPink}{\(v_3\)} & & \\
& \colorbox{CornflowerBlue}{\(v_4\)} & & \\
\hline
\multirow{3}{*}{\colorbox{LimeGreen}{\(s_2\)}} & \colorbox{Dandelion}{\(v_5\)} & \multirow{3}{*}{\colorbox{LimeGreen}{\(x_2\)}} &
\multirow{3}{*}{\(\colorbox{LimeGreen}{\(a_2\)} = a(\colorbox{Dandelion}{\(y_5\)}, \colorbox{Emerald}{\(y_6\)}, \colorbox{Gray}{\(y_7\)})\)} \\
& \colorbox{Emerald}{\(v_6\)} & & \\
& \colorbox{Gray}{\(v_7\)} & & \\
\end{tabular}
\end{flushright}
\end{frame}

\begin{frame}{27 from 55 features selected by forward algorithm}
\tiny

\begin{tabular}{rcc}
feature & p value & direction \\
\hline
DP mean & 0.003 & + \\
BaseQRankSum mean & 0.003 & - \\
phast Score bin mean & 0.005 & + \\
Intercept & 0.007 & + \\
Antidepress[T.1] & 0.007 & - \\
AF mean & 0.012 & - \\
\color{NavyBlue}{gerp ElementRSScore bin mean} & 0.016 & - \\
\color{NavyBlue}{ensembl PredictedFunction entropy} & 0.017 & + \\
MQRankSum mean & 0.025 & - \\
\color{NavyBlue}{GWASpval mean} & 0.036 & - \\
\color{NavyBlue}{szdbCNVcount mean} & 0.045 & + \\
\color{NavyBlue}{cpg CpGIsland bin mean} & 0.053 & + \\
ageOfDeath & 0.069 & - \\
\color{NavyBlue}{encode FeatureTypeClass entropy} & 0.072 & + \\
EV2 & 0.079 & - \\
Ethnicity[T.Caucasian] & 0.086 & - \\
culprit entropy & 0.132 & + \\
Alcohol[T.1] & 0.147 & - \\
ReportedGender[T.Male] & 0.148 & - \\
Lithium[T.1] & 0.162 & - \\
YearofAutopsy & 0.165 & - \\
QD mean & 0.178 & + \\
EV1 & 0.221 & - \\
Anticonvulsants[T.1] & 0.224 & - \\
AntipsychAtyp[T.1] & 0.326 & + \\
Ethnicity[T.Hispanic] & 0.372 & - \\
\color{NavyBlue}{sift Prediction entropy} & 0.428 & - \\
\end{tabular}
\end{frame}

\begin{frame}{Summary and early conclusions}
\begin{enumerate}
\item somatic SNVs and small indels called from bulk WGS 
\begin{enumerate}
\item 2488 calls from 61 SCZ brains
\item 895 calls from 25 Control brains
\end{enumerate}
\item sensitivity: joint dependence on read depth and allele freq
\item \(> 100\) biological and technical features to each call
\item preliminary statistical modeling \(\mathrm{\#calls} \sim \mathrm{Dx} + X\)
\begin{itemize}
\item some genomic features \(X\) affect \(\mathrm{\#calls}\)
\item \(\mathrm{Dx}\) (SCZ) does not affect \(\mathrm{\#calls}\)
\end{itemize}
\end{enumerate}
\end{frame}

\begin{frame}{Further work}
\begin{enumerate}
\item close look at genes implicated in SCZ and psychiatric disorders
\item statistical modeling \(\mathrm{Dx} \sim X\)
\begin{itemize}
\item confirm findings
\item feature selection
\end{itemize}
\end{enumerate}
\begin{enumerate}
\item copy number variation (Eduardo M, Walsh lab)
\item joint analysis of somatic and germline variants
\begin{enumerate}
\item polygenic risk score
\end{enumerate}
\end{enumerate}
\end{frame}

\begin{frame}{Acknowledgements}
\begin{itemize}
\item Taejeong Bae, Abyzov lab, Mayo Clinic
\item Park \& Walsh labs, Harvard %CNV control
\item Urban lab, Stanford %transposable elements
\item Akbarian lab, Mount Sinai % nuclei sorting
\item David Obensheim, NIMH
\item Mette Peters, Kenny Daily, Cindy Molitor, Sage
\end{itemize}
\end{frame}

\end{document}

\begin{columns}[t]
\begin{column}{0.5\textwidth}

\end{column}

\begin{column}{0.5\textwidth}

\end{column}
\end{columns}

% Sequencing this Spring
% We did the calling with Alexej and TJ
% shorten the intro and contrast with autism that SCZ risk is spread out
% among thousands of germline variants
% study question: somatic mutations that are distinct?
% muscle and NeuN-: early study design "a remnant"
% first pass analysis: focus on all available features; final analysis:
% hypothesis driven, gene centric
% emphasize hypothesis driven and present the large feature set as a basis for that
% DONTs: don't say the data science way is superior to the hypothesis driven way
% we don't know the answer to our main question yet
