{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Submitting data from Synapse scratch space"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%reload_ext autoreload\n",
    "import synapseclient\n",
    "import cmc_submit2ndar as s2n\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pickle\n",
    "import os.path\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome, Attila Jones!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "syn = synapseclient.login()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data files to submit\n",
    "These data files have different types marked by their extension "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading file from /home/attila/projects/bsm/results/2021-02-02-submit-to-nda/files_not_uploaded_yet.MSSM.txt.p\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>synapseID</th>\n",
       "      <th>filename</th>\n",
       "      <th>data_file1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>indivID</th>\n",
       "      <th>filetype</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CMC_MSSM_027</th>\n",
       "      <th>cram</th>\n",
       "      <td>syn22007834</td>\n",
       "      <td>MSSM_027_NeuN_pl.cram</td>\n",
       "      <td>5229ec83-c2eb-4f44-8d2b-f688b476bbda/MSSM_027_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMC_MSSM_033</th>\n",
       "      <th>cram</th>\n",
       "      <td>syn22090557</td>\n",
       "      <td>MSSM_033_NeuN_pl.cram</td>\n",
       "      <td>475392a7-5f20-4a07-8694-9dc86a677032/MSSM_033_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMC_MSSM_055</th>\n",
       "      <th>cram</th>\n",
       "      <td>syn22162964</td>\n",
       "      <td>MSSM_055_NeuN_pl.cram</td>\n",
       "      <td>e55eb946-1335-4dd5-a26b-85d74b5e3a36/MSSM_055_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMC_MSSM_056</th>\n",
       "      <th>cram</th>\n",
       "      <td>syn22093909</td>\n",
       "      <td>MSSM_056_NeuN_pl.cram</td>\n",
       "      <td>af069550-3c69-4e3a-bbb3-48d532d2c21c/MSSM_056_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMC_MSSM_063</th>\n",
       "      <th>cram</th>\n",
       "      <td>syn22150192</td>\n",
       "      <td>MSSM_063_NeuN_pl.cram</td>\n",
       "      <td>d98961ae-eef8-4eb3-8be8-5a79f9fda507/MSSM_063_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMC_PITT_072</th>\n",
       "      <th>unmapped.bam</th>\n",
       "      <td>syn22385266</td>\n",
       "      <td>PITT_072_NeuN_pl.unmapped.bam</td>\n",
       "      <td>bc52db52-ba9d-49e4-a9fb-5bc2ec7e22e3/PITT_072_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMC_PITT_082</th>\n",
       "      <th>unmapped.bam</th>\n",
       "      <td>syn22384272</td>\n",
       "      <td>PITT_082_NeuN_pl.unmapped.bam</td>\n",
       "      <td>a877be33-4ca7-4165-b2a3-178bbd818d0f/PITT_082_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMC_PITT_098</th>\n",
       "      <th>unmapped.bam</th>\n",
       "      <td>syn22385231</td>\n",
       "      <td>PITT_098_NeuN_pl.unmapped.bam</td>\n",
       "      <td>abbcfd1c-e2f8-4cdd-8bb3-8b26c0a26788/PITT_098_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMC_PITT_101</th>\n",
       "      <th>unmapped.bam</th>\n",
       "      <td>syn22381683</td>\n",
       "      <td>PITT_101_NeuN_pl.unmapped.bam</td>\n",
       "      <td>2b8d3b25-3654-43e8-b188-a4089b1afd93/PITT_101_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMC_PITT_113</th>\n",
       "      <th>unmapped.bam</th>\n",
       "      <td>syn22385324</td>\n",
       "      <td>PITT_113_NeuN_pl.unmapped.bam</td>\n",
       "      <td>8dc9dc6f-61df-4d90-a36b-f4a32bdff594/PITT_113_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>809 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             synapseID                       filename  \\\n",
       "indivID      filetype                                                   \n",
       "CMC_MSSM_027 cram          syn22007834          MSSM_027_NeuN_pl.cram   \n",
       "CMC_MSSM_033 cram          syn22090557          MSSM_033_NeuN_pl.cram   \n",
       "CMC_MSSM_055 cram          syn22162964          MSSM_055_NeuN_pl.cram   \n",
       "CMC_MSSM_056 cram          syn22093909          MSSM_056_NeuN_pl.cram   \n",
       "CMC_MSSM_063 cram          syn22150192          MSSM_063_NeuN_pl.cram   \n",
       "...                                ...                            ...   \n",
       "CMC_PITT_072 unmapped.bam  syn22385266  PITT_072_NeuN_pl.unmapped.bam   \n",
       "CMC_PITT_082 unmapped.bam  syn22384272  PITT_082_NeuN_pl.unmapped.bam   \n",
       "CMC_PITT_098 unmapped.bam  syn22385231  PITT_098_NeuN_pl.unmapped.bam   \n",
       "CMC_PITT_101 unmapped.bam  syn22381683  PITT_101_NeuN_pl.unmapped.bam   \n",
       "CMC_PITT_113 unmapped.bam  syn22385324  PITT_113_NeuN_pl.unmapped.bam   \n",
       "\n",
       "                                                                  data_file1  \n",
       "indivID      filetype                                                         \n",
       "CMC_MSSM_027 cram          5229ec83-c2eb-4f44-8d2b-f688b476bbda/MSSM_027_...  \n",
       "CMC_MSSM_033 cram          475392a7-5f20-4a07-8694-9dc86a677032/MSSM_033_...  \n",
       "CMC_MSSM_055 cram          e55eb946-1335-4dd5-a26b-85d74b5e3a36/MSSM_055_...  \n",
       "CMC_MSSM_056 cram          af069550-3c69-4e3a-bbb3-48d532d2c21c/MSSM_056_...  \n",
       "CMC_MSSM_063 cram          d98961ae-eef8-4eb3-8be8-5a79f9fda507/MSSM_063_...  \n",
       "...                                                                      ...  \n",
       "CMC_PITT_072 unmapped.bam  bc52db52-ba9d-49e4-a9fb-5bc2ec7e22e3/PITT_072_...  \n",
       "CMC_PITT_082 unmapped.bam  a877be33-4ca7-4165-b2a3-178bbd818d0f/PITT_082_...  \n",
       "CMC_PITT_098 unmapped.bam  abbcfd1c-e2f8-4cdd-8bb3-8b26c0a26788/PITT_098_...  \n",
       "CMC_PITT_101 unmapped.bam  2b8d3b25-3654-43e8-b188-a4089b1afd93/PITT_101_...  \n",
       "CMC_PITT_113 unmapped.bam  8dc9dc6f-61df-4d90-a36b-f4a32bdff594/PITT_113_...  \n",
       "\n",
       "[809 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fpath = '/home/attila/projects/bsm/results/2021-02-02-submit-to-nda/files_not_uploaded_yet.MSSM.txt'\n",
    "picklepath = fpath + '.p'\n",
    "if not os.path.exists(picklepath):\n",
    "    dfiles = s2n.read_dfiles(syn, fpath)\n",
    "    pickle.dump(dfiles, open(picklepath, 'wb'))\n",
    "else:\n",
    "    print('loading file from', picklepath)\n",
    "    with open(picklepath, 'rb') as f:\n",
    "        dfiles = pickle.load(f)\n",
    "dfiles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `prefixes` dictionary contains the AWS S3 prefix for each file type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prefixes = {'cram': '3338602',\n",
    "            'cram.crai': '3338602',\n",
    "            'flagstat.txt': '3340241',\n",
    "            'ploidy_2.vcf.gz': '3340241',\n",
    "            'ploidy_2.vcf.gz.tbi': '3340241',\n",
    "            'ploidy_12.vcf.gz': '3340241',\n",
    "            'ploidy_12.vcf.gz.tbi': '3340241',\n",
    "            'ploidy_50.vcf.gz': '3340241',\n",
    "            'ploidy_50.vcf.gz.tbi': '3340241',\n",
    "            'unmapped.bam': '3338602'\n",
    "           }\n",
    "ftypes = dfiles.index.get_level_values(1).unique()\n",
    "set(ftypes) == set(prefixes.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating `genomics_subject02` and `nichd_btb02`\n",
    "These refer to the same samples and subjects as the ones submitted with the FASTQ files.  Since there were two FASTQ submissions for these set of samples I merge them "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def foo(maniftype='genomics_subject02'):\n",
    "    basen = '/home/attila/projects/bsm/results/2021-02-02-submit-to-nda/' + maniftype + '-'\n",
    "    fpaths = [basen + x + '.csv' for x in ['2019-12-09', '2020-06-08']]\n",
    "    manif = pd.concat([pd.read_csv(x, skiprows=1) for x in fpaths], axis=0)\n",
    "    target_path = '/home/attila/projects/bsm/results/2021-02-02-submit-to-nda/' + maniftype +'-2021-02-02.csv'\n",
    "    template_path = '/home/attila/projects/bsm/results/2020-04-22-upload-to-ndar-from-s3/chess-' + maniftype + '.csv'\n",
    "    s2n.write_manifest(manif, template_path=template_path, target_path=target_path)\n",
    "    return(manif)\n",
    "\n",
    "gsub = foo('genomics_subject02')\n",
    "btb = foo('nichd_btb02')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating `genomics_sample03`\n",
    "\n",
    "### `genomics_sample03` filled template\n",
    "\n",
    "I take `genomics_sample03` from a previous submission that was used for the FASTQ files for the same set of 81 samples.  I name this manifest `oldgsam`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subjectkey</th>\n",
       "      <th>experiment_id</th>\n",
       "      <th>src_subject_id</th>\n",
       "      <th>interview_age</th>\n",
       "      <th>interview_date</th>\n",
       "      <th>sample_description</th>\n",
       "      <th>sample_id_original</th>\n",
       "      <th>organism</th>\n",
       "      <th>sample_amount</th>\n",
       "      <th>sample_unit</th>\n",
       "      <th>...</th>\n",
       "      <th>patient_id_biorepository</th>\n",
       "      <th>sample_id_biorepository</th>\n",
       "      <th>cell_id_original</th>\n",
       "      <th>cell_id_biorepository</th>\n",
       "      <th>comments_misc</th>\n",
       "      <th>site</th>\n",
       "      <th>rat280</th>\n",
       "      <th>rat230</th>\n",
       "      <th>gqn</th>\n",
       "      <th>seq_batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NDAR_INV0971H4H4</td>\n",
       "      <td>1223</td>\n",
       "      <td>CMC_MSSM_033</td>\n",
       "      <td>972</td>\n",
       "      <td>04/13/2018</td>\n",
       "      <td>frontal cortex</td>\n",
       "      <td>MSSM_033.DLPFC_1355.np1</td>\n",
       "      <td>human</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CMC_MSSM_033</td>\n",
       "      <td>MSSM_033.DLPFC_1355.np1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>U01MH106891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NDAR_INV0971H4H4</td>\n",
       "      <td>1223</td>\n",
       "      <td>CMC_MSSM_033</td>\n",
       "      <td>972</td>\n",
       "      <td>04/13/2018</td>\n",
       "      <td>frontal cortex</td>\n",
       "      <td>MSSM_033.DLPFC_1355.np1</td>\n",
       "      <td>human</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CMC_MSSM_033</td>\n",
       "      <td>MSSM_033.DLPFC_1355.np1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>U01MH106891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NDAR_INV1VPUF5CL</td>\n",
       "      <td>1223</td>\n",
       "      <td>CMC_MSSM_056</td>\n",
       "      <td>804</td>\n",
       "      <td>04/13/2018</td>\n",
       "      <td>frontal cortex</td>\n",
       "      <td>MSSM_056.DLPFC_1181.np1</td>\n",
       "      <td>human</td>\n",
       "      <td>1543.0</td>\n",
       "      <td>ng</td>\n",
       "      <td>...</td>\n",
       "      <td>CMC_MSSM_056</td>\n",
       "      <td>MSSM_056.DLPFC_1181.np1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>U01MH106891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NDAR_INV1VPUF5CL</td>\n",
       "      <td>1223</td>\n",
       "      <td>CMC_MSSM_056</td>\n",
       "      <td>804</td>\n",
       "      <td>04/13/2018</td>\n",
       "      <td>frontal cortex</td>\n",
       "      <td>MSSM_056.DLPFC_1181.np1</td>\n",
       "      <td>human</td>\n",
       "      <td>1543.0</td>\n",
       "      <td>ng</td>\n",
       "      <td>...</td>\n",
       "      <td>CMC_MSSM_056</td>\n",
       "      <td>MSSM_056.DLPFC_1181.np1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>U01MH106891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NDAR_INV1VPUF5CL</td>\n",
       "      <td>1223</td>\n",
       "      <td>CMC_MSSM_056</td>\n",
       "      <td>804</td>\n",
       "      <td>04/13/2018</td>\n",
       "      <td>frontal cortex</td>\n",
       "      <td>MSSM_056.DLPFC_1181.np1</td>\n",
       "      <td>human</td>\n",
       "      <td>1543.0</td>\n",
       "      <td>ng</td>\n",
       "      <td>...</td>\n",
       "      <td>CMC_MSSM_056</td>\n",
       "      <td>MSSM_056.DLPFC_1181.np1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>U01MH106891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>NDAR_INVTY432EB9</td>\n",
       "      <td>1223</td>\n",
       "      <td>CMC_MSSM_379</td>\n",
       "      <td>852</td>\n",
       "      <td>06/08/2020</td>\n",
       "      <td>frontal cortex</td>\n",
       "      <td>1238.np1</td>\n",
       "      <td>human</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CMC_MSSM_379</td>\n",
       "      <td>1238.np1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>U01MH106891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>NDAR_INVLH141LX6</td>\n",
       "      <td>1223</td>\n",
       "      <td>CMC_MSSM_340</td>\n",
       "      <td>912</td>\n",
       "      <td>06/08/2020</td>\n",
       "      <td>frontal cortex</td>\n",
       "      <td>1247.np1</td>\n",
       "      <td>human</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CMC_MSSM_340</td>\n",
       "      <td>1247.np1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>U01MH106891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>NDAR_INVAK206LGH</td>\n",
       "      <td>1223</td>\n",
       "      <td>CMC_MSSM_362</td>\n",
       "      <td>876</td>\n",
       "      <td>06/08/2020</td>\n",
       "      <td>frontal cortex</td>\n",
       "      <td>1346.np1</td>\n",
       "      <td>human</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CMC_MSSM_362</td>\n",
       "      <td>1346.np1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>U01MH106891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>NDAR_INVTM018EUJ</td>\n",
       "      <td>1223</td>\n",
       "      <td>CMC_MSSM_405</td>\n",
       "      <td>708</td>\n",
       "      <td>06/08/2020</td>\n",
       "      <td>frontal cortex</td>\n",
       "      <td>1357.np1</td>\n",
       "      <td>human</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CMC_MSSM_405</td>\n",
       "      <td>1357.np1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>U01MH106891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>NDAR_INVDZ834FVN</td>\n",
       "      <td>1223</td>\n",
       "      <td>CMC_PITT_052</td>\n",
       "      <td>708</td>\n",
       "      <td>06/08/2020</td>\n",
       "      <td>frontal cortex</td>\n",
       "      <td>1454.np1</td>\n",
       "      <td>human</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>CMC_PITT_052</td>\n",
       "      <td>1454.np1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>U01MH106891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>186 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          subjectkey  experiment_id src_subject_id  interview_age  \\\n",
       "0   NDAR_INV0971H4H4           1223   CMC_MSSM_033            972   \n",
       "1   NDAR_INV0971H4H4           1223   CMC_MSSM_033            972   \n",
       "2   NDAR_INV1VPUF5CL           1223   CMC_MSSM_056            804   \n",
       "3   NDAR_INV1VPUF5CL           1223   CMC_MSSM_056            804   \n",
       "4   NDAR_INV1VPUF5CL           1223   CMC_MSSM_056            804   \n",
       "..               ...            ...            ...            ...   \n",
       "64  NDAR_INVTY432EB9           1223   CMC_MSSM_379            852   \n",
       "65  NDAR_INVLH141LX6           1223   CMC_MSSM_340            912   \n",
       "66  NDAR_INVAK206LGH           1223   CMC_MSSM_362            876   \n",
       "67  NDAR_INVTM018EUJ           1223   CMC_MSSM_405            708   \n",
       "68  NDAR_INVDZ834FVN           1223   CMC_PITT_052            708   \n",
       "\n",
       "   interview_date sample_description       sample_id_original organism  \\\n",
       "0      04/13/2018     frontal cortex  MSSM_033.DLPFC_1355.np1    human   \n",
       "1      04/13/2018     frontal cortex  MSSM_033.DLPFC_1355.np1    human   \n",
       "2      04/13/2018     frontal cortex  MSSM_056.DLPFC_1181.np1    human   \n",
       "3      04/13/2018     frontal cortex  MSSM_056.DLPFC_1181.np1    human   \n",
       "4      04/13/2018     frontal cortex  MSSM_056.DLPFC_1181.np1    human   \n",
       "..            ...                ...                      ...      ...   \n",
       "64     06/08/2020     frontal cortex                 1238.np1    human   \n",
       "65     06/08/2020     frontal cortex                 1247.np1    human   \n",
       "66     06/08/2020     frontal cortex                 1346.np1    human   \n",
       "67     06/08/2020     frontal cortex                 1357.np1    human   \n",
       "68     06/08/2020     frontal cortex                 1454.np1    human   \n",
       "\n",
       "    sample_amount sample_unit  ...  patient_id_biorepository  \\\n",
       "0             1.0         NaN  ...              CMC_MSSM_033   \n",
       "1             1.0         NaN  ...              CMC_MSSM_033   \n",
       "2          1543.0          ng  ...              CMC_MSSM_056   \n",
       "3          1543.0          ng  ...              CMC_MSSM_056   \n",
       "4          1543.0          ng  ...              CMC_MSSM_056   \n",
       "..            ...         ...  ...                       ...   \n",
       "64            1.0         NaN  ...              CMC_MSSM_379   \n",
       "65            1.0         NaN  ...              CMC_MSSM_340   \n",
       "66            1.0         NaN  ...              CMC_MSSM_362   \n",
       "67            1.0         NaN  ...              CMC_MSSM_405   \n",
       "68            1.0         NaN  ...              CMC_PITT_052   \n",
       "\n",
       "    sample_id_biorepository cell_id_original cell_id_biorepository  \\\n",
       "0   MSSM_033.DLPFC_1355.np1              NaN                   NaN   \n",
       "1   MSSM_033.DLPFC_1355.np1              NaN                   NaN   \n",
       "2   MSSM_056.DLPFC_1181.np1              NaN                   NaN   \n",
       "3   MSSM_056.DLPFC_1181.np1              NaN                   NaN   \n",
       "4   MSSM_056.DLPFC_1181.np1              NaN                   NaN   \n",
       "..                      ...              ...                   ...   \n",
       "64                 1238.np1              NaN                   NaN   \n",
       "65                 1247.np1              NaN                   NaN   \n",
       "66                 1346.np1              NaN                   NaN   \n",
       "67                 1357.np1              NaN                   NaN   \n",
       "68                 1454.np1              NaN                   NaN   \n",
       "\n",
       "   comments_misc         site  rat280  rat230  gqn  seq_batch  \n",
       "0            NaN  U01MH106891     NaN     NaN  NaN        NaN  \n",
       "1            NaN  U01MH106891     NaN     NaN  NaN        NaN  \n",
       "2            NaN  U01MH106891     NaN     NaN  NaN        NaN  \n",
       "3            NaN  U01MH106891     NaN     NaN  NaN        NaN  \n",
       "4            NaN  U01MH106891     NaN     NaN  NaN        NaN  \n",
       "..           ...          ...     ...     ...  ...        ...  \n",
       "64           NaN  U01MH106891     NaN     NaN  NaN        NaN  \n",
       "65           NaN  U01MH106891     NaN     NaN  NaN        NaN  \n",
       "66           NaN  U01MH106891     NaN     NaN  NaN        NaN  \n",
       "67           NaN  U01MH106891     NaN     NaN  NaN        NaN  \n",
       "68           NaN  U01MH106891     NaN     NaN  NaN        NaN  \n",
       "\n",
       "[186 rows x 32 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "basen = '/home/attila/projects/bsm/results/2021-02-02-submit-to-nda/genomics_sample03-'\n",
    "fpaths = [basen + x + '.csv' for x in ['2019-12-09', '2020-06-08']]\n",
    "oldgsam = pd.concat([pd.read_csv(x, skiprows=1) for x in fpaths], axis=0)\n",
    "oldgsam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Editing `genomics_sample03`\n",
    "\n",
    "The `edit_gsam` function takes the `oldgsam` template and simplifies it to one row for each sample.  Then it replaces the `data_file1_type` and `data_file1` fields with the file type (e.g `unmapped` and `295ba82f-8f14-4e50-8945-4496f5131c9e/MSSM_027_NeuN_pl.unmapped.bam`, respectively).\n",
    "\n",
    "`edit_gsam` edits `data_file1` in a certain way; see Cindy Molitor's email from Nov 30, 2020, 1:29 PM\n",
    "\n",
    "> Note that the s3 prefix for the scratch bucket should not be included in the file path in the data_file columns, e.g. the value of the column should be \n",
    ">\n",
    "> 89eac311-84e3-4be0-9f0e-c6186104f71c/NC1-CX-ASTMIG.cram\n",
    ">\n",
    "> instead of\n",
    ">\n",
    "> 3340241/89eac311-84e3-4be0-9f0e-c6186104f71c/NC1-CX-ASTMIG.cram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "template_path = '/home/attila/projects/bsm/results/2020-04-22-upload-to-ndar-from-s3/genomics_sample03_template.csv'\n",
    "def foo(ftype):\n",
    "    gsam = s2n.edit_gsam(oldgsam, dfiles, gender=gsub.set_index('src_subject_id')['gender'], dftype=ftype)\n",
    "    target_path = '/home/attila/projects/bsm/results/2021-02-02-submit-to-nda/genomics_sample03-' + ftype + '.csv'\n",
    "    s2n.write_manifest(gsam, template_path=template_path, target_path=target_path)\n",
    "    return(gsam)\n",
    "\n",
    "gsamd = {ftype: foo(ftype) for ftype in ftypes}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submission\n",
    "First submit everything except for `unmapped.bam`s because `PITT_117_NeuN_pl.unmapped.bam` is still missing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cram\n",
      "Running NDATools Version 0.2.3\n",
      "Opening log: /home/attila/NDAValidationResults/debug_log_20210203T221214.txt\n",
      "\n",
      "Validating files...\n",
      "Validation report output to: /home/attila/NDAValidationResults/validation_results_20210203T221214.csv\n",
      "Warnings output to: /home/attila/NDAValidationResults/validation_results_20210203T221214.csv\n",
      "\n",
      "All files have finished validating.\n",
      "\n",
      "The following files passed validation:\n",
      "UUID 257914cb-445a-4398-9dc5-16d2668232f2: genomics_sample03-cram.csv\n",
      "UUID a20f5739-0230-4f0c-9a99-c1f0667910dc: genomics_subject02-2021-02-02.csv\n",
      "UUID 779c13a5-183f-4f7d-b961-e1df2ed85373: nichd_btb02-2021-02-02.csv\n",
      "Enter title for dataset name:----\n",
      "cram.crai\n",
      "Running NDATools Version 0.2.3\n",
      "Opening log: /home/attila/NDAValidationResults/debug_log_20210203T221216.txt\n",
      "\n",
      "Validating files...\n",
      "Validation report output to: /home/attila/NDAValidationResults/validation_results_20210203T221216.csv\n",
      "Warnings output to: /home/attila/NDAValidationResults/validation_results_20210203T221216.csv\n",
      "\n",
      "All files have finished validating.\n",
      "\n",
      "The following files passed validation:\n",
      "UUID 38cf65e4-2776-4abf-b855-99df20f9f7b7: nichd_btb02-2021-02-02.csv\n",
      "UUID ed89c15f-5c20-4b6d-a651-70780fc7f689: genomics_subject02-2021-02-02.csv\n",
      "UUID f9ba0dbc-9568-4abe-a12c-93c76b4acee4: genomics_sample03-cram.crai.csv\n",
      "Enter title for dataset name:----\n",
      "flagstat.txt\n",
      "Running NDATools Version 0.2.3\n",
      "Opening log: /home/attila/NDAValidationResults/debug_log_20210203T221218.txt\n",
      "\n",
      "Validating files...\n",
      "Validation report output to: /home/attila/NDAValidationResults/validation_results_20210203T221218.csv\n",
      "Warnings output to: /home/attila/NDAValidationResults/validation_results_20210203T221218.csv\n",
      "\n",
      "All files have finished validating.\n",
      "\n",
      "The following files passed validation:\n",
      "UUID 2e83f00b-1fda-40be-bdbc-36bd48db4d22: nichd_btb02-2021-02-02.csv\n",
      "UUID 138137b8-88ad-4b8b-bfa2-7cea0ae6b1b5: genomics_sample03-flagstat.txt.csv\n",
      "UUID 8f9f9287-a0c9-4d57-8f73-338e31f1dd0b: genomics_subject02-2021-02-02.csv\n",
      "Enter title for dataset name:----\n",
      "ploidy_2.vcf.gz\n",
      "Running NDATools Version 0.2.3\n",
      "Opening log: /home/attila/NDAValidationResults/debug_log_20210203T221220.txt\n",
      "\n",
      "Validating files...\n",
      "Validation report output to: /home/attila/NDAValidationResults/validation_results_20210203T221220.csv\n",
      "Warnings output to: /home/attila/NDAValidationResults/validation_results_20210203T221220.csv\n",
      "\n",
      "All files have finished validating.\n",
      "\n",
      "The following files passed validation:\n",
      "UUID 08d837d5-7a34-442d-affd-e5a14cf1e011: nichd_btb02-2021-02-02.csv\n",
      "UUID c9cc3bb2-214d-49fa-bd2d-8bd05fedaec3: genomics_subject02-2021-02-02.csv\n",
      "UUID c37d25b4-9122-4500-bca3-0a38d40efe15: genomics_sample03-ploidy_2.vcf.gz.csv\n",
      "Enter title for dataset name:----\n",
      "ploidy_2.vcf.gz.tbi\n",
      "Running NDATools Version 0.2.3\n",
      "Opening log: /home/attila/NDAValidationResults/debug_log_20210203T221221.txt\n",
      "\n",
      "Validating files...\n",
      "Validation report output to: /home/attila/NDAValidationResults/validation_results_20210203T221222.csv\n",
      "Warnings output to: /home/attila/NDAValidationResults/validation_results_20210203T221222.csv\n",
      "\n",
      "All files have finished validating.\n",
      "\n",
      "The following files passed validation:\n",
      "UUID f7cec924-d9ca-4629-a4e5-ec82f32129ff: genomics_sample03-ploidy_2.vcf.gz.tbi.csv\n",
      "UUID bf63de8d-2e2a-40dd-b9e9-d2d713db876c: genomics_subject02-2021-02-02.csv\n",
      "UUID f38fc5c5-4ee7-48eb-b2f8-3602aed6d922: nichd_btb02-2021-02-02.csv\n",
      "Enter title for dataset name:----\n",
      "ploidy_12.vcf.gz\n",
      "Running NDATools Version 0.2.3\n",
      "Opening log: /home/attila/NDAValidationResults/debug_log_20210203T221224.txt\n",
      "\n",
      "Validating files...\n",
      "Validation report output to: /home/attila/NDAValidationResults/validation_results_20210203T221224.csv\n",
      "Warnings output to: /home/attila/NDAValidationResults/validation_results_20210203T221224.csv\n",
      "\n",
      "All files have finished validating.\n",
      "\n",
      "The following files passed validation:\n",
      "UUID 601305eb-02d4-4871-bbd4-c831c2870773: nichd_btb02-2021-02-02.csv\n",
      "UUID 8b98d726-8a06-42c6-aa2e-18e1d1063b49: genomics_subject02-2021-02-02.csv\n",
      "UUID 571064e7-cc1c-4e30-93f2-ab87e6516ff8: genomics_sample03-ploidy_12.vcf.gz.csv\n",
      "Enter title for dataset name:----\n",
      "ploidy_12.vcf.gz.tbi\n",
      "Running NDATools Version 0.2.3\n",
      "Opening log: /home/attila/NDAValidationResults/debug_log_20210203T221226.txt\n",
      "\n",
      "Validating files...\n",
      "Validation report output to: /home/attila/NDAValidationResults/validation_results_20210203T221226.csv\n",
      "Warnings output to: /home/attila/NDAValidationResults/validation_results_20210203T221226.csv\n",
      "\n",
      "All files have finished validating.\n",
      "\n",
      "The following files passed validation:\n",
      "UUID 84ba246b-daff-4813-82cf-1d00c695f431: genomics_subject02-2021-02-02.csv\n",
      "UUID af05e18f-273e-4653-b35b-10336997d2d8: nichd_btb02-2021-02-02.csv\n",
      "UUID fa1e7843-a6a4-41b2-b0d4-9b4e1aac1961: genomics_sample03-ploidy_12.vcf.gz.tbi.csv\n",
      "Enter title for dataset name:----\n",
      "ploidy_50.vcf.gz\n",
      "Running NDATools Version 0.2.3\n",
      "Opening log: /home/attila/NDAValidationResults/debug_log_20210203T221227.txt\n",
      "\n",
      "Validating files...\n",
      "Validation report output to: /home/attila/NDAValidationResults/validation_results_20210203T221228.csv\n",
      "Warnings output to: /home/attila/NDAValidationResults/validation_results_20210203T221228.csv\n",
      "\n",
      "All files have finished validating.\n",
      "\n",
      "The following files passed validation:\n",
      "UUID d60f8b1f-7c55-4d8a-bba0-4f7a7a753915: genomics_subject02-2021-02-02.csv\n",
      "UUID 556e0996-d3d9-46d5-a494-58667c002240: nichd_btb02-2021-02-02.csv\n",
      "UUID bea23421-6307-4bd4-b5f8-7397b5538f5c: genomics_sample03-ploidy_50.vcf.gz.csv\n",
      "Enter title for dataset name:----\n",
      "ploidy_50.vcf.gz.tbi\n",
      "Running NDATools Version 0.2.3\n",
      "Opening log: /home/attila/NDAValidationResults/debug_log_20210203T221229.txt\n",
      "\n",
      "Validating files...\n",
      "Validation report output to: /home/attila/NDAValidationResults/validation_results_20210203T221229.csv\n",
      "Warnings output to: /home/attila/NDAValidationResults/validation_results_20210203T221229.csv\n",
      "\n",
      "All files have finished validating.\n",
      "\n",
      "The following files passed validation:\n",
      "UUID f4bfa448-6f90-43e5-ba28-1eabc5d5c3b5: genomics_subject02-2021-02-02.csv\n",
      "UUID 01de61e3-6258-43d4-be4f-40abedd69b0a: nichd_btb02-2021-02-02.csv\n",
      "UUID 73927e79-564e-4a35-a9c0-a90bfc3a1d02: genomics_sample03-ploidy_50.vcf.gz.tbi.csv\n",
      "Enter title for dataset name:----\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r",
      "  0%|          | 0/3 [00:00<?, ?it/s]\r",
      " 33%|â–ˆâ–ˆâ–ˆâ–Ž      | 1/3 [00:01<00:02,  1.38s/it]\r",
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:01<00:00,  1.00it/s]\r",
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:01<00:00,  1.87it/s]\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/attila/.local/bin/vtcmd\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 290, in main\n",
      "    package_results = build_package(uuid, associated_files, config=config)\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 226, in build_package\n",
      "    config.title = input('Enter title for dataset name:')\n",
      "EOFError: EOF when reading a line\n",
      "\r",
      "  0%|          | 0/3 [00:00<?, ?it/s]\r",
      " 33%|â–ˆâ–ˆâ–ˆâ–Ž      | 1/3 [00:00<00:01,  1.13it/s]\r",
      " 67%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–‹   | 2/3 [00:01<00:00,  1.40it/s]\r",
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:01<00:00,  1.84it/s]\r",
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:01<00:00,  2.13it/s]\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/attila/.local/bin/vtcmd\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 290, in main\n",
      "    package_results = build_package(uuid, associated_files, config=config)\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 226, in build_package\n",
      "    config.title = input('Enter title for dataset name:')\n",
      "EOFError: EOF when reading a line\n",
      "\r",
      "  0%|          | 0/3 [00:00<?, ?it/s]\r",
      " 33%|â–ˆâ–ˆâ–ˆâ–Ž      | 1/3 [00:00<00:01,  1.20it/s]\r",
      " 67%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–‹   | 2/3 [00:01<00:00,  1.52it/s]\r",
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:01<00:00,  1.91it/s]\r",
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:01<00:00,  2.30it/s]\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/attila/.local/bin/vtcmd\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 290, in main\n",
      "    package_results = build_package(uuid, associated_files, config=config)\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 226, in build_package\n",
      "    config.title = input('Enter title for dataset name:')\n",
      "EOFError: EOF when reading a line\n",
      "\r",
      "  0%|          | 0/3 [00:00<?, ?it/s]\r",
      " 33%|â–ˆâ–ˆâ–ˆâ–Ž      | 1/3 [00:00<00:01,  1.20it/s]\r",
      " 67%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–‹   | 2/3 [00:01<00:00,  1.56it/s]\r",
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:01<00:00,  2.71it/s]\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/attila/.local/bin/vtcmd\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 290, in main\n",
      "    package_results = build_package(uuid, associated_files, config=config)\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 226, in build_package\n",
      "    config.title = input('Enter title for dataset name:')\n",
      "EOFError: EOF when reading a line\n",
      "\r",
      "  0%|          | 0/3 [00:00<?, ?it/s]\r",
      " 33%|â–ˆâ–ˆâ–ˆâ–Ž      | 1/3 [00:01<00:02,  1.11s/it]\r",
      " 67%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–‹   | 2/3 [00:01<00:00,  1.15it/s]\r",
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:01<00:00,  1.48it/s]\r",
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:01<00:00,  1.76it/s]\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/attila/.local/bin/vtcmd\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 290, in main\n",
      "    package_results = build_package(uuid, associated_files, config=config)\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 226, in build_package\n",
      "    config.title = input('Enter title for dataset name:')\n",
      "EOFError: EOF when reading a line\n",
      "\r",
      "  0%|          | 0/3 [00:00<?, ?it/s]\r",
      " 33%|â–ˆâ–ˆâ–ˆâ–Ž      | 1/3 [00:00<00:01,  1.06it/s]\r",
      " 67%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–‹   | 2/3 [00:01<00:00,  1.38it/s]\r",
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:01<00:00,  2.49it/s]\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/attila/.local/bin/vtcmd\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 290, in main\n",
      "    package_results = build_package(uuid, associated_files, config=config)\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 226, in build_package\n",
      "    config.title = input('Enter title for dataset name:')\n",
      "EOFError: EOF when reading a line\n",
      "\r",
      "  0%|          | 0/3 [00:00<?, ?it/s]\r",
      " 33%|â–ˆâ–ˆâ–ˆâ–Ž      | 1/3 [00:00<00:01,  1.02it/s]\r",
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:01<00:00,  1.38it/s]\r",
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:01<00:00,  2.29it/s]\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/attila/.local/bin/vtcmd\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 290, in main\n",
      "    package_results = build_package(uuid, associated_files, config=config)\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 226, in build_package\n",
      "    config.title = input('Enter title for dataset name:')\n",
      "EOFError: EOF when reading a line\n",
      "\r",
      "  0%|          | 0/3 [00:00<?, ?it/s]\r",
      " 33%|â–ˆâ–ˆâ–ˆâ–Ž      | 1/3 [00:00<00:01,  1.01it/s]\r",
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:01<00:00,  1.38it/s]\r",
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:01<00:00,  2.30it/s]\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/attila/.local/bin/vtcmd\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 290, in main\n",
      "    package_results = build_package(uuid, associated_files, config=config)\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 226, in build_package\n",
      "    config.title = input('Enter title for dataset name:')\n",
      "EOFError: EOF when reading a line\n",
      "\r",
      "  0%|          | 0/3 [00:00<?, ?it/s]\r",
      " 33%|â–ˆâ–ˆâ–ˆâ–Ž      | 1/3 [00:00<00:01,  1.04it/s]\r",
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:01<00:00,  1.44it/s]\r",
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:01<00:00,  2.71it/s]\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/attila/.local/bin/vtcmd\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 290, in main\n",
      "    package_results = build_package(uuid, associated_files, config=config)\n",
      "  File \"/home/attila/.local/lib/python3.7/site-packages/NDATools/clientscripts/vtcmd.py\", line 226, in build_package\n",
      "    config.title = input('Enter title for dataset name:')\n",
      "EOFError: EOF when reading a line\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/projects/bsm/results/2021-02-02-submit-to-nda\n",
    "validate=\"vtcmd -b -s3 nda-bsmn-scratch -pre 3338602 -u $NDA_USER -p $NDA_PASSWORD -c 2965 -w  -t title -d description\"\n",
    "manifests=\"nichd_btb02-2021-02-02.csv genomics_subject02-2021-02-02.csv\"\n",
    "dftypes=\"cram cram.crai flagstat.txt ploidy_2.vcf.gz ploidy_2.vcf.gz.tbi ploidy_12.vcf.gz ploidy_12.vcf.gz.tbi ploidy_50.vcf.gz ploidy_50.vcf.gz.tbi\"\n",
    "for dftype in $dftypes; do\n",
    "echo $dftype\n",
    "$validate $manifests genomics_sample03-$dftype.csv\n",
    "echo '----'\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"shell_port\": 34869,\n",
      "  \"iopub_port\": 42471,\n",
      "  \"stdin_port\": 58131,\n",
      "  \"control_port\": 38257,\n",
      "  \"hb_port\": 58923,\n",
      "  \"ip\": \"127.0.0.1\",\n",
      "  \"key\": \"dd2d1bbc-23ffdb9ef3b30854b6460fed\",\n",
      "  \"transport\": \"tcp\",\n",
      "  \"signature_scheme\": \"hmac-sha256\",\n",
      "  \"kernel_name\": \"\"\n",
      "}\n",
      "\n",
      "Paste the above JSON into a file, and connect with:\n",
      "    $> jupyter <app> --existing <file>\n",
      "or, if you are local, you can connect with just:\n",
      "    $> jupyter <app> --existing kernel-b2759800-5071-41de-be61-ce4144813a56.json\n",
      "or even just:\n",
      "    $> jupyter <app> --existing\n",
      "if this is the most recent Jupyter kernel you have started.\n"
     ]
    }
   ],
   "source": [
    "%connect_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
