Illumina short sequencing reads were processed by a pipeline described in detail in (Comprehensive identification of somatic nucleotide variants in human brain tissue; Genome Biology, in press, DOI : 10.1186/s13059-021-02285-3).  Briefly, for each sample, the reads were aligned to the reference sequence GRCh37d5 using bwa (version 3.7.16a).  Duplicate reads were marked with  PICARD (v2.12.1).  Indel realignment and base quality score recalibration using GATK (v3.7-0) gave rise to the final BAM file for each sample.

From the final BAM files single nucleotide variants and small indels were called using GATK (v3.7-0) HaplotypeCaller with --ploidy option 2, 12 or 50 resulting in three raw call sets per sample.  The raw call sets were filtered in the following sequential steps: (1) Only PASS calls were retained; (2) known germline variants were excluded based on the 1000 Genomes, ExAC, GnomAD,ESP5600, and Kaviar databases; (3) calls located in problematic genomic regions defined as non-P bases of the 1000 Genomes were removed; (4) a binomial test was used to remove likely heterozygous variants: for each variant the p and N parameter of the binomial null distribution was set to 0.5 and the total read count (read depth) at the variant position, respectively, and we removed the variant supporting the null hypothesis at alpha = 0.00001; (5) a similar binomial hypothesis test was used to filter variant calls at sites with strand bias; (6) similarly, Fisher's exact test was used to filter out calls with imbalance of strand ratios between reference (REF) and alternative (ALT) bases; (7) multi allelic calls were removed; (8) calls at sites with > 2.5 copy number, estimated using CNVnator, were also removed.

The implementation of the pipeline used for the present work was designed to run on AWS EC-2 compute instances managed by the AWS parallelCluster software. The source code is available at https://github.com/bsmn/bsmn-pipeline.
