---
layout: default
title: "Comparing Callsets: Callers and Referenfe Tissues"
---

```{r echo=FALSE, warning=FALSE, message=FALSE}
library(lme4)
library(lattice)
library(VennDiagram)
#library(latticeExtra)
opts_chunk$set(dpi = 144)
opts_chunk$set(out.width = "400px")
opts_chunk$set(fig.width = 4)
opts_chunk$set(fig.asp = 1)
opts_chunk$set(dev = c("png", "pdf"))
lattice.options(default.args = list(as.table = TRUE))
lattice.options(default.theme = "standard.theme")
source("2017-05-29-vcf-comparisons.R")
```

## Goals

Assess concordance of call sets

1. between callers
   * compare call set by mutect2 and that by strelka for a given tissue-pair, variant type (snvs or indels), and filter setting
1. between reference tissues
   * compare call set with muscle reference to that with NeuN_mn reference given a variant type
   * use call sets obtained by mutect2 $$\cap$$ strelka given a tissue pair and variant type and filter setting

## Results

### Computation and summary

The `myisec.sh` script does all the job: various filtering, conversion, indexing and comparison using `bcftools`, as well as summarizing the results by calculating the size of call sets in files named `callset-sizes.tsv`, which are printed below.  The only filter setting tested was the *PASS* value of mutect2; otherwise unfiltered sets are reported.  Note that none of the strelka records have *PASS* filter value.

```{r engine="bash"}
./myisec.sh $HOME/projects/bsm/results/2017-05-03-strelka-mutect2-pilot/32MB
./myisec.sh $HOME/projects/bsm/results/2017-05-03-strelka-mutect2-pilot/32MB PASS
find results/ -name callset-sizes.tsv | xargs head
```

Import sizes of various callsets to R:

```{r}
indirs <- paste0("results/mutect2-", c("unfilt", "PASS"), "/")
indirs <- paste0(rep(indirs, each = 3), c(paste0("1_isec-callers/", c("NeuN_mn-NeuN_pl", "muscle-NeuN_pl")), "2_cmp-reftissues"))
indirs <- paste0(rep(indirs, each = 2), c("/indels/", "/snvs/"))
names(indirs) <- LETTERS[seq_along(indirs)]
clsets <-
    lapply(indirs, function(x) {
           df <- read.delim(paste0(x, "callset-sizes.tsv"), header = FALSE,
                            col.names = c("set.size", "file.vcf", "set.operator", "set.operand"))
           df$directory <- factor(x)
           return(df)
       })
clsets <- do.call(rbind, clsets)
```

### Comparing callers

#### NeuN_mn reference tissue, SNVs

Without filtering mutect2 calls many more variants than strelka.  With filtering (PASS) mutect2 still calls slightly more variants than stelka without filtering.  The overlap between the two callers is substantial and filtering makes mutect2 calls even more concordant with strelka calls.

```{r venn-caller-neun_mn-snvs}
my.venn3(dir = "1_isec-callers/NeuN_mn-NeuN_pl/snvs/", cls = clsets)
```

#### NeuN_mn reference tissue, indels

Callsets of indels share some but not all of the above properties with SNV callsets.  Importantly, the concordance seems lower for indels than for NSPs, but the much lower number of indel records leaves a greater room for certain patterns emerging by chance.

```{r venn-caller-neun_mn-indels, echo=FALSE}
my.venn3(dir = "1_isec-callers/NeuN_mn-NeuN_pl/indels/", cls = clsets)
```


#### muscle reference tissue, SNVs

These results are very similar to the previous ones; the muscle reference gives very slightly more candidate variants than NeuN_mn reference, which aligns with the expectation that muscle cell lineage is further away from the NeuN_pl lineage compared to NeuN_mn.  But the differences are very subtle.

```{r venn-caller-muscle-snvs, echo=FALSE}
my.venn3(dir = "1_isec-callers/muscle-NeuN_pl/snvs/", cls = clsets)
```

#### muscle reference tissue, indels

```{r venn-caller-muscle-indels, echo=FALSE}
my.venn3(dir = "1_isec-callers/muscle-NeuN_pl/indels/", cls = clsets)
```

### Comparing reference tissues: unfiltered

The next Venn-Euler diagrams feature two sets corresponding to muscle and NeuN_mn reference tissue.  Both sets are **intersections** of mutect2 and strelka call sets.  This first set of diagrams was obtained **without** filtering mutect2 calls.  The overlap is relatively larger than either set difference---records that are private to either muscle or NeuN_mn reference.  This suggests that the choice of reference tissue is not critical.  Fewer records are private to NeuN_mn than to muscle.  This is again consistent with the muscle lineage being more distant to the NeuN_pl lineage.

### SNVs

```{r venn-ref-tissue-snvs, echo=FALSE}
my.venn2(dir = "results/mutect2-unfilt/2_cmp-reftissues/snvs/", cls = clsets,
             category = c("NeuN_mn", "muscle"), col = my.cols <- c("gray", "rosybrown"), fill = my.cols)
```

#### indels

```{r venn-ref-tissue-indels, echo=FALSE}
my.venn2(dir = "results/mutect2-unfilt/2_cmp-reftissues/indels/", cls = clsets,
             category = c("NeuN_mn", "muscle"), col = my.cols <- c("gray", "rosybrown"), fill = my.cols)
```

### Comparing reference tissues: filtered

The next set of diagrams was obtained **with** filtered mutect2 calls.  These results show the same qualitative picture as the ones without filtering.

### SNVs

```{r venn-ref-tissue-snvs-PASS, echo=FALSE}
my.venn2(dir = "results/mutect2-PASS/2_cmp-reftissues/snvs/", cls = clsets,
             category = c("NeuN_mn", "muscle"), col = my.cols <- c("gray", "rosybrown"), fill = my.cols)
```

#### indels

```{r venn-ref-tissue-indels-PASS, echo=FALSE}
my.venn2(dir = "results/mutect2-PASS/2_cmp-reftissues/indels/", cls = clsets,
             category = c("NeuN_mn", "muscle"), col = my.cols <- c("gray", "rosybrown"), fill = my.cols)
```
