```{r echo=FALSE, warning=FALSE, message=FALSE}
library(lme4)
library(lattice)
library(VennDiagram)
#library(latticeExtra)
opts_chunk$set(dpi = 144)
opts_chunk$set(out.width = "400px")
opts_chunk$set(fig.width = 4)
opts_chunk$set(fig.asp = 1)
opts_chunk$set(dev = c("png", "pdf"))
lattice.options(default.args = list(as.table = TRUE))
lattice.options(default.theme = "standard.theme")
source("2017-05-29-vcf-comparisons.R")
```

## Goals

Assess concordance of call sets

1. between callers
   * compare call set by mutect2 and that by strelka for a given tissue-pair and variant type (snvs or indels)
1. between reference tissues
   * compare call set with muscle reference to that with NeuN_mn reference given a variant type
   * use call sets obtained by mutect2 $\cap$ strelka given a tissue pair and variant type

The first

## Results

### Computation and summary

```{r engine="bash"}
./myisec.sh $HOME/projects/bsm/results/2017-05-03-strelka-mutect2-pilot/32MB
./myisec.sh $HOME/projects/bsm/results/2017-05-03-strelka-mutect2-pilot/32MB PASS
find results/ -name callset-sizes.tsv | xargs head
```

```{r}
indirs <- paste0("results/mutect2-", c("unfilt", "PASS"), "/")
indirs <- paste0(rep(indirs, each = 3), c(paste0("1_isec-callers/", c("NeuN_mn-NeuN_pl", "muscle-NeuN_pl")), "2_cmp-reftissues"))
indirs <- paste0(rep(indirs, each = 2), c("/indels/", "/snvs/"))
names(indirs) <- LETTERS[seq_along(indirs)]
clsets <-
    lapply(indirs, function(x) {
           df <- read.delim(paste0(x, "callset-sizes.tsv"), header = FALSE,
                            col.names = c("set.size", "file.vcf", "set.operator", "set.operand"))
           df$directory <- factor(x)
           return(df)
       })
clsets <- do.call(rbind, clsets)
```

### Comparing callers

#### NeuN_mn reference tissue, SNVs

```{r venn-caller-neun_mn-snvs}
my.venn3(dir = "1_isec-callers/NeuN_mn-NeuN_pl/snvs/", cls = clsets)
```

#### NeuN_mn reference tissue, indels

```{r venn-caller-neun_mn-indels, echo=FALSE}
my.venn3(dir = "1_isec-callers/NeuN_mn-NeuN_pl/indels/", cls = clsets)
```


#### muscle reference tissue, SNVs

```{r venn-caller-muscle-snvs, echo=FALSE}
my.venn3(dir = "1_isec-callers/muscle-NeuN_pl/snvs/", cls = clsets)
```

#### muscle reference tissue, indels

```{r venn-caller-muscle-indels, echo=FALSE}
my.venn3(dir = "1_isec-callers/muscle-NeuN_pl/indels/", cls = clsets)
```

### Comparing reference tissues: unfiltered

### SNVs

```{r venn-ref-tissue-snvs, echo=FALSE}
my.venn2(dir = "results/mutect2-unfilt/2_cmp-reftissues/snvs/", cls = clsets,
             category = c("NeuN_mn", "muscle"), col = my.cols <- c("gray", "rosybrown"), fill = my.cols)
```

#### indels

```{r venn-ref-tissue-indels, echo=FALSE}
my.venn2(dir = "results/mutect2-unfilt/2_cmp-reftissues/indels/", cls = clsets,
             category = c("NeuN_mn", "muscle"), col = my.cols <- c("gray", "rosybrown"), fill = my.cols)
```

### Comparing reference tissues: filtered

mutect2 PASS only

### SNVs

```{r venn-ref-tissue-snvs-PASS, echo=FALSE}
my.venn2(dir = "results/mutect2-PASS/2_cmp-reftissues/snvs/", cls = clsets,
             category = c("NeuN_mn", "muscle"), col = my.cols <- c("gray", "rosybrown"), fill = my.cols)
```

#### indels

```{r venn-ref-tissue-indels-PASS, echo=FALSE}
my.venn2(dir = "results/mutect2-PASS/2_cmp-reftissues/indels/", cls = clsets,
             category = c("NeuN_mn", "muscle"), col = my.cols <- c("gray", "rosybrown"), fill = my.cols)
```
