\documentclass{beamer}
%\documentclass[handout]{beamer}

% language settings
%\usepackage{fontspec, polyglossia}
%\setdefaultlanguage{magyar}

% common packages
\usepackage{amsmath, multimedia, hyperref, color, multirow}
%\usepackage{graphicx}

% TikZ
\usepackage{tikz}
%\usetikzlibrary{arrows.meta, decorations.pathmorphing, decorations.pathreplacing, shapes.geometric,mindmap}
%\usetikzlibrary{shapes.geometric,fadings,bayesnet}

% beamer styles
\mode<presentation>{
\usetheme{Warsaw}
%\usetheme{Antibes}
\usecolortheme{beaver}
%\usecolortheme{seahorse}
%\usefonttheme{structureitalicserif}
\setbeamercovered{transparent}
}
\setbeamertemplate{blocks}[rounded][shadow=true]
\AtBeginSubsection[]{
  \begin{frame}<beamer>{Contents}
    \tableofcontents[currentsection,currentsubsection]
  \end{frame}
}
%\useoutertheme[]{tree}

% title, etc
\title{Truth sets of somatic variants}
%\subtitle{A subtitle may be shorter and more technical}
\author{Attila Jones, Chaggai Rosenbluh, Andy Chess}
\date{Chess Lab, Mount Sinai School of Medicine}

\begin{document}

\maketitle

\begin{frame}{Measuring performance of somatic variant calling}
%{Compare call set \(C\) to truth set \(V\)}
\begin{description}
\item [\(C\)] some call set of somatic variants
\item [\(V\)] the truth set of somatic variants
\end{description}
\vfill
\alert{Precision}\\
the fraction of calls that are true
\begin{equation}
p(C) = \frac{|C \cap V|}{|C|} = \frac{\#\text{true calls}}{\#\mathrm{calls}}
\end{equation}
\alert{Recall}\\
the fraction of variants that are called
\begin{equation}
r(C) = \frac{|C \cap V|}{|V|} = \frac{\#\text{true calls}}{\#\mathrm{variants}}
\end{equation}
\end{frame}

\begin{frame}<1>[label=ourapproach]{The Chess lab's approach to truth set}
\begin{enumerate}
\item<1> accept germline call sets as germline truth sets
\item<2> mix genomes in different proportions 
\item<3> map joint genotypes to alternative allele frequencies (AAF) 
\item<3> partition variants according to AAF 
\item<4-> set \# of variants for each AAF (modeling)
\only<5>{\huge \alert{How?}}
\end{enumerate}
\begin{columns}[t]
\begin{column}{0.5\textwidth}

\includegraphics[scale=0.3]{../../notebook/2018-05-01-ceph-vars-greally/figure/venn-S1-1.pdf}
\end{column}

\begin{column}{0.5\textwidth}

{\tiny
\only<2>{
\begin{tabular}{r|ccc}
genome & mix1 & mix2 & mix3 \\
\hline
NA12889 & 4 & 2 & 0 \\
NA12891 & 8 & 4 & 0 \\
NA12890 & 16 & 8 & 0 \\
NA12892 & 72 & 86 & 100 \\
\end{tabular}
}
\vfill
\only<3>{
\begin{tabular}{c|ccc}
genotype & \multicolumn{3}{c}{AAF (\%)} \\
& mix1 &  mix2 & mix3 \\
\hline
0001 & 36 & 43 & 50\\
0010 & 8 & 4 & 0\\
0100 & 4 & 2 & 0\\
1000 & 2 & 1 & 0\\
0002 & 72 & 86 & 100\\
... & ... & ... & ...\\
1222 & 98 & 99 & 100\\
2122 & 96 & 98 & 100\\
2212 & 92 & 96 & 100\\
2221 & 64 & 57 & 50\\
2222 & 100 & 100 & 100\\
\end{tabular}
}
}
\end{column}
\end{columns}
\end{frame}

\begin{frame}{Size of germline callset looks OK}
\begin{center}
{\large all autosomes}
\vfill

\includegraphics[scale=0.6]{../../notebook/2019-03-18-truth-sets/figure/2019-03-18-truth-sets_14_0.pdf}
\end{center}
\end{frame}

\againframe<2-3>{ourapproach}

\begin{frame}[plain]
\includegraphics[scale=0.3]{../../notebook/2019-04-04-truth-sets-aaf/figure/2019-04-04-truth-sets-aaf_7_1.pdf}
\end{frame}

\begin{frame}{Small region for test runs}
\includegraphics[scale=0.35]{../../notebook/2019-04-04-truth-sets-aaf/figure/2019-04-04-truth-sets-aaf_10_1.pdf}
\end{frame}

\againframe<4>{ourapproach}

\begin{frame}[label=chaggais_model]{Initial ``models''}
\includegraphics[scale=0.35]{../../notebook/2019-04-04-truth-sets-aaf/figure/2019-04-04-truth-sets-aaf_24_1.pdf}
\end{frame}

\begin{frame}{Model: M\_Chaggai}
\includegraphics[height=0.7\textheight]{/home/attila/figures/from-others/chaggai-pr-prec-recall.png}
\end{frame}

\againframe<5>{ourapproach}

\begin{frame}[plain]
Four simple models
\includegraphics[scale=0.3]{../../notebook/2019-04-04-truth-sets-aaf/figure/2019-04-04-truth-sets-aaf_21_1.pdf}
\end{frame}

\end{document}



\begin{columns}[t]
\begin{column}{0.5\textwidth}

\end{column}

\begin{column}{0.5\textwidth}

\end{column}
\end{columns}
