\documentclass{beamer} %\documentclass[handout]{beamer}


%\includeonlyframes{bsm,title,toc,imprint-mouse-devel,igf2-imprint-evol,sister-disorders,previous-age-studies,our-study,cmc,toc-current,filtering-calling,fitting-models,ll-surface,all-betas,orthogonality,identifiability,anova,betas-cluster,signif-genes,cf-overall-expression,summary,improving-model,p-val,imprinted-brain,chess-lab}

% language settings
%\usepackage{fontspec, polyglossia}
%\setdefaultlanguage{magyar}

% common packages
\usepackage{amsmath, multimedia, hyperref, color, multirow}
%\usepackage{graphicx}

% TikZ
\usepackage{tikz}
%\usetikzlibrary{arrows.meta, decorations.pathmorphing, decorations.pathreplacing, shapes.geometric,mindmap}
%\usetikzlibrary{shapes.geometric,fadings,bayesnet}

% beamer styles
\mode<presentation>{
\usetheme{Boadilla}
%\usetheme{Copenhagen}
%\usecolortheme{beaver}
\usecolortheme{default}
%\usefonttheme{structureitalicserif}
\setbeamercovered{transparent}
}
\setbeamertemplate{blocks}[rounded][shadow=true]
\AtBeginSection[]{
  \begin{frame}<beamer>{Contents}
    %\tableofcontents[currentsection]
    \tableofcontents[currentsection]
  \end{frame}
}
%\useoutertheme[]{tree}

\title{Accurate variant calling with FDR control}
\author{Attila Guly\'{a}s-Kov\'{a}cs}
\date{Chess Lab at Mount Sinai}

\newcommand{\platefigscale}[0]{0.7}
\newcommand{\ownfigscale}[0]{0.4}

\begin{document}

\begin{frame}[plain, label=title]
\maketitle
\end{frame}

\begin{frame}{Challenges of variant calling}
\begin{columns}[t]
\begin{column}{0.5\textwidth}

\includegraphics[width=1.0\columnwidth]{figures/from-others/kim-2013-bmcbioinf-fig1a.png}

%\includegraphics[width=1.0\columnwidth]{figures/from-others/kim-2013-bmcbioinf-fig1b.png}
\end{column}
\begin{column}{0.5\textwidth}
{\large accuracy}
\begin{itemize}
\item no perfect caller
\item depends on quality annotations
\end{itemize}
\bigskip
{\large approaches}
\begin{enumerate}
\item heuristic (conventional)
\begin{itemize}
\item hard filters
\item combine call sets
\end{itemize}
\item machine learning-based
\begin{itemize}
\item optimally weighted combination
\item soft filters 
\item FDR control
\item probability that a call is true
\end{itemize}
\end{enumerate}
%\includegraphics[width=1.0\columnwidth]{figures/from-others/kim-2013-bmcbioinf-fig1b.png}
\end{column}
\end{columns}
\end{frame}


\begin{frame}{Probability that a called variant is true}

A set \(\mathcal{K}\) of \(K\) callers was applied to data \(D\).\\
Candidate variant \(i\) was called by a subset \(\mathcal{C}\subseteq\mathcal{K}\)
of callers. \\
Quality annotations \(X_{i\mathcal{K}}\).

\[
P(i \text{ is true } | D, i \text{ by }
\mathcal{C}\subseteq\mathcal{K}, X_{i\mathcal{K}}) \equiv \pi_{i}
\]
\begin{itemize}
\item would be useful for FDR control and downstream analysis
\item but it is unknown
\end{itemize}

\bigskip
{\onslide<2>
\begin{center}
\Large We need a good estimator \(\hat{\pi}_{i}\) for \(\pi_{i}\).
\end{center}
}
\end{frame}

\begin{frame}{FDR control (precision-based filtering)}
\begin{columns}[t]
\small
\begin{column}{0.45\textwidth}
\begin{center}
{let \(j = \text{\# called positives}\) }
\smallskip
\end{center}
\begin{eqnarray*}
\text{precision}_j &=& \text{E} \left[ \frac{\text{\# true positives}}{j} \right] \\
&=& \frac{\pi_{(1)} + ... + \pi_{(j)}}{j} \\
&=& 1 - \text{FDR}_j
\end{eqnarray*}
\smallskip
\[
\overbrace{\text{recall}_j}^{\text{sensitivity}} = \text{E} \left[
\frac{\text{\# true positives}}{\text{\# actual positives}} \right]
\]
\end{column}
\begin{column}{0.6\textwidth}

\includegraphics[width=1.0\columnwidth]{figures/from-others/PRROC-R-package-fig2.png}
\end{column}
\end{columns}
\end{frame}

%\begin{frame}{Algorithm for FDR control}
%\begin{enumerate}
%\item set FDR to some desired level \(\alpha\)
%\item rank candidate variants by \(\pi_{i|\mathcal{K}}\) 
%\item for each rank \(j\) calculate
%\begin{equation*}
%\overbrace{\text{FDR}_{(j)|\mathcal{K}}}^{q\text{-value}} = 1 -
%\text{precision}_{(j)|\mathcal{K}} = 1 -
%%\frac{1}{j} \sum_{k=1}^j \pi_{(k)}
%\frac{\pi_{(1)|\mathcal{K}} + ... + \pi_{(j)|\mathcal{K}}}{j}
%\end{equation*}
%\item call the \(j'\) highest ranking candidates if
%\(\text{FDR}_{(j')|\mathcal{K}} <
%\alpha < \text{FDR}_{(j'+1)|\mathcal{K}}\)
%\end{enumerate}
%
%\bigskip
%{\footnotesize
%\begin{tabular}{cl}
%\hline
%%\(j\) & the rank of candidate variant \(i\) \\
%\(\text{FDR}_{(j)|\mathcal{K}}\) & the FDR when we filter just below rank \(j\) \\
%%\(\pi_i\) & the probability that \(i\) is a true call \\
%\end{tabular}
%}
%\end{frame}


%\begin{frame}{Optimal combination of callers}
%\begin{columns}[t]
%\begin{column}{0.45\textwidth}
%Using a set \(\mathcal{K}\) of \(K\) callers \\
%we have
%\begin{itemize}
%\item \(K\) VCFs with annot.~\(\mathbf{x}_{i\mathcal{K}}\)
%\item \(\hat{\pi}_{ik | \mathcal{K}}\) analogous to \(\hat{\pi}_{i | \{k\}}\)
%\end{itemize}
%\bigskip
%
%%optimal combination:
%\[\underbrace{\hat{\pi}_{i | \mathcal{K}}}_{\text{metacaller}} = \frac{\hat{\pi}_{i1 | \mathcal{K}} + ... + \hat{\pi}_{iK | \mathcal{K}}}{K}\]
%\end{column}
%\begin{column}{0.6\textwidth}
%
%\includegraphics[width=1.0\columnwidth]{figures/from-others/PRROC-R-package-fig7.png}
%\end{column}
%\end{columns}
%\end{frame}

\begin{frame}{An estimator for \(\pi_i\): VariantMetaCaller (VMC)}
{G\'{e}zsi,..., Antal, BMC Genomics. 2015; 16: 875.}
\begin{eqnarray*}
\hat{\pi}_{i} &=& \frac{\hat{\pi}_{i1} + ... + \hat{\pi}_{iK}}{K} \\
\hat{\pi}_{ik} &=& \hat{P}_{\text{svm}}(i \text{ is true } |
\mathbf{x}_{ik}, \text{ training sets based on } \mathcal{K})
\end{eqnarray*}
\bigskip

input \(\mathbf{x}_{ik}\) to SVM
\begin{itemize}
\item \(\mathbf{x}_{ik} = (x_{ik1},...,x_{ikA})\) using \(A\) quality annotations \\
\item \emph{caller's score}, read depth, variant quality, mapping quality,...
\end{itemize}
\end{frame}

\begin{frame}{Support vector machines}
\begin{columns}[t]
\begin{column}{0.45\textwidth}

\includegraphics[width=1.0\columnwidth]{figures/from-others/ben-hur-2008-ploscompbio-fig6middle.png}
\medskip

{\footnotesize Ben-Hur et al 2008}
\end{column}
\begin{column}{0.55\textwidth}

\includegraphics<2>[width=1.0\columnwidth]{figures/from-others/kim-2013-bmcbioinf-fig1a.png}
\end{column}
\end{columns}
\end{frame}


%\begin{frame}{Training sets based on set \(\mathcal{K}\) of \(K\) callers}
%\begin{center}
%\begin{eqnarray*}
%\text{positive training set} &=& \{i \text{ called by all callers}\} \\
%\text{negative training set} &=& \{i \text{ called by any single caller}\}
%\end{eqnarray*}
%\bigskip
%
%\includegraphics[height=0.5\textheight]{figures/from-others/kim-2013-bmcbioinf-fig1a.png}
%\end{center}
%\end{frame}

\begin{frame}{Accuracy and goodness of precision estimation}
\begin{columns}[t]
\begin{column}{0.7\textwidth}
precision-recall curve
\bigskip

\includegraphics[width=1.0\columnwidth]{figures/from-others/vmc-fig4a-bwa-indel.png}
%\includegraphics[height=0.8\textheight]{figures/from-others/vmc-fig4b-bwa.png}
\end{column}

\begin{column}{0.35\textwidth}
mean abs.~error of \(\widehat{\text{FDR}}\)
\bigskip

\includegraphics[height=0.5\textheight]{figures/from-others/vmc-fig4c-bwa-indel.png}
\end{column}
\end{columns}
\end{frame}

\begin{frame}{Somatic variants from normal--tumor samples}
\includegraphics[height=0.7\textheight]{figures/from-others/baysic-fig6.png}
\bigskip

{\footnotesize BMC Bioinformatics. 2014; 15: 104.}
\end{frame}

\begin{frame}{Summary}
\includegraphics[width=0.9\textwidth]{figures/from-others/vmc-fig1.png}
\bigskip

{\footnotesize G\'{e}zsi,..., Antal, BMC Genomics. 2015; 16: 875.}
\end{frame}

%\begin{frame}{Summary}
%\begin{enumerate}
%\item useful information in
%\begin{itemize}
%\item calling pattern of mulitple callers
%\item quality annotations 
%\end{itemize}
%\item traditional approach
%\begin{itemize}
%\item heuristics: hard filter + combination
%\item Inaccurate. No FDR control. 
%\end{itemize} 
%\item VariantMetaCaller
%\begin{itemize}
%\item machine learning (SVM)
%\item best precision-recall
%\item well-estimated precision \(\Rightarrow\) FDR control
%\end{itemize}
%\end{enumerate}
%\end{frame}

\end{document}

\begin{columns}[t]
\begin{column}{0.5\textwidth}
\end{column}
\begin{column}{0.5\textwidth}
\end{column}
\end{columns}

%\begin{frame}{Brain Somatic Mosaicism}
%\begin{columns}[t]
%\begin{column}{0.6\textwidth}
%
%\includegraphics[width=1.0\columnwidth]{figures/from-others/gage-curropsysbio-2016-1.jpg}
%
%{\tiny Paquola, Erwin, Gage 2016}
%\end{column}
%
%\begin{column}{0.4\textwidth}
%challenges with somatic variants:
%\begin{enumerate}
%\item detection\\
%{\footnotesize allelic fraction}
%\item prioritization\\{\footnotesize multiple info}
%\item integration\\{\footnotesize germline vars.}
%\end{enumerate}
%\end{column}
%\end{columns}
%\end{frame}
