\documentclass{beamer} %\documentclass[handout]{beamer}


%\includeonlyframes{bsm,title,toc,imprint-mouse-devel,igf2-imprint-evol,sister-disorders,previous-age-studies,our-study,cmc,toc-current,filtering-calling,fitting-models,ll-surface,all-betas,orthogonality,identifiability,anova,betas-cluster,signif-genes,cf-overall-expression,summary,improving-model,p-val,imprinted-brain,chess-lab}

% language settings
%\usepackage{fontspec, polyglossia}
%\setdefaultlanguage{magyar}

% common packages
\usepackage{amsmath, multimedia, hyperref, color, multirow}
%\usepackage{graphicx}

% TikZ
\usepackage{tikz}
%\usetikzlibrary{arrows.meta, decorations.pathmorphing, decorations.pathreplacing, shapes.geometric,mindmap}
%\usetikzlibrary{shapes.geometric,fadings,bayesnet}

% beamer styles
\mode<presentation>{
\usetheme{Boadilla}
%\usetheme{Copenhagen}
%\usecolortheme{beaver}
\usecolortheme{default}
%\usefonttheme{structureitalicserif}
\setbeamercovered{transparent}
}
\setbeamertemplate{blocks}[rounded][shadow=true]
\AtBeginSection[]{
  \begin{frame}<beamer>{Contents}
    %\tableofcontents[currentsection]
    \tableofcontents[currentsection]
  \end{frame}
}
%\useoutertheme[]{tree}

\title{Accurate variant calling with FDR control}
\author{Attila Guly\'{a}s-Kov\'{a}cs}
\date{Chess Lab at Mount Sinai}

\newcommand{\platefigscale}[0]{0.7}
\newcommand{\ownfigscale}[0]{0.4}

\begin{document}

\begin{frame}[plain, label=title]
\maketitle
\end{frame}

\section{The problem}

\begin{frame}{Challenges of variant calling}
\begin{columns}[t]
\begin{column}{0.4\textwidth}

\includegraphics[width=1.0\columnwidth]{figures/from-others/kim-2013-bmcbioinf-fig1a.png}

\includegraphics[width=1.0\columnwidth]{figures/from-others/kim-2013-bmcbioinf-fig1b.png}
\end{column}
\begin{column}{0.5\textwidth}
\begin{enumerate}
\item accuracy
\begin{itemize}
\item no perfect caller
\item not even best caller
\end{itemize}
\item filtering
\begin{itemize}
\item quality annotations in VCF
\item sensitivity vs.~specificity
\end{itemize}
\item combination: \alert{meta}callers
\end{enumerate}
%\includegraphics[width=1.0\columnwidth]{figures/from-others/kim-2013-bmcbioinf-fig1b.png}
\end{column}
\end{columns}
\end{frame}


%\begin{frame}{The probability that \(i\) is a true call}
%\begin{center}
%
%\[P(i \text{ is true } | D, i \text{ called by } k) \equiv \pi_i\]
%
%\bigskip
%
%{\small
%\begin{tabular}{cl}
%\hline
%\(i\) & candidate variant \\
%\(D\) & sequence data \\
%\(k\) & caller \\
%\end{tabular}
%}
%\end{center}
%\end{frame}

\begin{frame}{}

Candidate variant \(i\) was called by (meta)caller \(k\) based on data \(D\).
\begin{itemize}
\item Is \(i\) a true variant?
\item Is \(i\) also causal?
\end{itemize}

\bigskip

\onslide<2>{
\begin{eqnarray*}
P(i \text{ is true and causal } | D, E_i, i \text{ cal'd by } k)
&=&
\overbrace{P(i \text{ is true } | D, i \text{ cal'd by } k)}^{\pi_i} \\
&\times&
P(i \text{ is causal } | i \text{ is true}, E_i)
\end{eqnarray*}

\bigskip

{\small
\begin{tabular}{cl}
\hline
%\(\mathcal{T}\) & true variants \\
%\(v\) & a called variant \\
%\(\mathcal{C}\) & called variants \\
%\(\mathcal{V}\) & true (but unknown) variants \\
%\(i\) & candidate variant \\
\(E_v\) & prior evidence that a true variant \(v\) is causal \\
\end{tabular}
}
}
\end{frame}

\begin{frame}{Performance and precision-based filtering}
\begin{columns}[t]
\small
\begin{column}{0.45\textwidth}
\begin{eqnarray*}
\text{precision} &=& \text{E} \left[ \frac{\text{\# true calls}}{\text{\# calls}} \right] \\
&=& 1 - \underbrace{\text{E} \left[ \frac{\text{\# false calls}}{\text{\# calls}} \right]}_{\text{FDR}} \\
\text{recall} &=& \text{E} \left[ \frac{\text{\# true calls}}{\text{\# variants}} \right]
\end{eqnarray*}
\end{column}
\begin{column}{0.6\textwidth}

\includegraphics[width=1.0\columnwidth]{figures/from-others/PRROC-R-package-fig2.png}
\end{column}
\end{columns}

\end{frame}

\begin{frame}{Algorithm for precision-based filtering (FDR control)}

\begin{enumerate}
\item set FDR to some desired level \(\alpha\)
\item rank candidate variants by \(\pi_i\) 
\item for each rank \(j\) calculate
\begin{equation*}
\overbrace{\text{FDR}_{(j)}}^{q\text{-value}} = 1 - \text{precision}_{(j)} = 1 -
%\frac{1}{j} \sum_{k=1}^j \pi_{(k)}
\frac{\pi_{(1)} + ... + \pi_{(j)}}{j}
\end{equation*}
\item call the \(j'\) highest ranking candidates if \(\text{FDR}_{(j')} <
\alpha < \text{FDR}_{(j'+1)}\)

\end{enumerate}

\bigskip
{\footnotesize
\begin{tabular}{cl}
\hline
%\(j\) & the rank of candidate variant \(i\) \\
\(\text{FDR}_{(j)}\) & FDR when we filter just below rank \(j\) \\
%\(\pi_i\) & the probability that \(i\) is a true call \\
\end{tabular}
}
\end{frame}


\begin{frame}{Optimal soft filtering}
we have
\begin{itemize}
\item a VCF file by a typical caller \(k\)
\item quality annotations \(\mathbf{x}_{ik} = (x_{ik1},...,x_{ikA})\)
\end{itemize}
\bigskip
\begin{description}
\item[Problem:] \(\{\pi_{ik}\}\) is unknown and depends on \(\mathbf{x}_{ik}\)
\item[Task:]
\begin{enumerate}
\item find a good estimator \(\hat{\pi}_{i}\) for \(\pi_{i}\) \\
\item maximize \(\hat{\pi}_{ik}\) given \(\mathbf{x}_{ik}\)
\end{enumerate}

\end{description}
\end{frame}


\begin{frame}{Optimal metacaller}
\begin{columns}[t]
\begin{column}{0.45\textwidth}
we have
\begin{itemize}
\item \(K\) VCFs by \(K\) callers
\item maximized \(\hat{\pi}_{ik}\)
\end{itemize}
\bigskip

optimal metacaller:
\[\underbrace{\hat{\pi}_{i}}_{\text{metacaller}} = \frac{\hat{\pi}_{i1} + ... + \hat{\pi}_{iK}}{K}\]
\end{column}
\begin{column}{0.6\textwidth}

\includegraphics[width=1.0\columnwidth]{figures/from-others/PRROC-R-package-fig7.png}
\end{column}
\end{columns}

\end{frame}

\section{Solutions}

\begin{frame}{Callers and metacallers}
\includegraphics[width=0.9\textwidth]{figures/from-others/vmc-fig1.png}
\bigskip

{\footnotesize G\'{e}zsi,..., Antal, BMC Genomics. 2015; 16: 875.}
\end{frame}

\begin{frame}{BAYSIC}
\includegraphics[height=0.7\textheight]{figures/from-others/baysic-fig6.png}
\bigskip

{\footnotesize BMC Bioinformatics. 2014; 15: 104.}
\end{frame}

\begin{frame}{VariantMetaCaller (VMC)}
recall that

\(\pi_{ik} = P(i \text{ is true } | \mathbf{x}_{ik}, D, i \text{ called by
caller } k)\)
\bigskip

estimator

\(\hat{\pi}_{ik} = \hat{P}_{\text{svm}}(i \text{ is true } | \mathbf{x}_{ik},
D, i \text{ called by \emph{any caller}} )\)
\begin{itemize}
\item optimal soft filtering
\item good estimator 
\item svm: support vector machine
\end{itemize}
\end{frame}

\begin{frame}{Support vector machines}

\end{frame}

\begin{frame}{Performance and precision}
\begin{columns}[t]
\begin{column}{0.5\textwidth}
performance

\includegraphics[height=0.8\textheight]{figures/from-others/vmc-fig4b-bwa.png}
\end{column}

\begin{column}{0.5\textwidth}
error of precision

\includegraphics[height=0.8\textheight]{figures/from-others/vmc-fig4c-bwa.png}
\end{column}
\end{columns}
\end{frame}

\end{document}

\begin{columns}[t]
\begin{column}{0.5\textwidth}
\end{column}
\begin{column}{0.5\textwidth}
\end{column}
\end{columns}

%\begin{frame}{Brain Somatic Mosaicism}
%\begin{columns}[t]
%\begin{column}{0.6\textwidth}
%
%\includegraphics[width=1.0\columnwidth]{figures/from-others/gage-curropsysbio-2016-1.jpg}
%
%{\tiny Paquola, Erwin, Gage 2016}
%\end{column}
%
%\begin{column}{0.4\textwidth}
%challenges with somatic variants:
%\begin{enumerate}
%\item detection\\
%{\footnotesize allelic fraction}
%\item prioritization\\{\footnotesize multiple info}
%\item integration\\{\footnotesize germline vars.}
%\end{enumerate}
%\end{column}
%\end{columns}
%\end{frame}
