\documentclass[usenames,dvipsnames]{beamer}
%\documentclass[handout]{beamer}

% language settings
%\usepackage{fontspec, polyglossia}
%\setdefaultlanguage{magyar}

% common packages
\usepackage{amsmath, multimedia, url, hyperref, multirow, colortbl}
\usepackage[table]{xcolor}
%\usepackage{graphicx}

% TikZ
\usepackage{tikz}
%\usetikzlibrary{arrows.meta, decorations.pathmorphing, decorations.pathreplacing, shapes.geometric,mindmap}
%\usetikzlibrary{shapes.geometric,fadings,bayesnet}

% beamer styles
\mode<presentation>{
\usetheme{Singapore}
%\usetheme{Warsaw}
%\usetheme{Antibes}
%\usecolortheme{beaver}
%\usecolortheme{crane}
%\usecolortheme{seahorse}
\usecolortheme{default}
%\usefonttheme{structureitalicserif}
\setbeamercovered{transparent}
}
\setbeamertemplate{blocks}[rounded][shadow=true]
\AtBeginSubsection[]{
  \begin{frame}<beamer>{Contents}
    \tableofcontents[currentsection,currentsubsection]
  \end{frame}
}
%\useoutertheme[]{tree}

% title, etc
\title{Two Genomic Studies on Schizophrenia}
\subtitle{and Related Psychiatric Disorders}
%\subtitle{A subtitle may be shorter and more technical}
\author{Attila Jones}
\date{Brain Somatic Mosaicism Network}

\begin{document}

\maketitle

\section{Genetics of Schizophrenia}

\begin{frame}{What is schizophrenia (SCZ)?}
\begin{columns}[t]
\begin{column}{0.55\textwidth}
	\begin{itemize}
		\item \(\approx\)1\% prevalence
                \item psychosis
                \item social, cognitive dysfunction
                \item environmental risk
		\item associations with MHC and synaptic genes
                \item a neurodevelopmental disorder?
	\end{itemize}
\end{column}

\begin{column}{0.45\textwidth}

\end{column}
\end{columns}
%symptoms
%highly heritable
%unclear aetiology:
\end{frame}

\begin{frame}{Heritability in common disorders}
\begin{columns}[t]
\begin{column}{0.4\textwidth}

\includegraphics[width=1.0\columnwidth]{from-others/sullivan2012-fig1a.png}

{\tiny Sullivan 2012 Nature Reviews Genetics}
\end{column}

\begin{column}{0.6\textwidth}

\includegraphics[width=1.0\columnwidth]{from-others/brainstorm-consortium-science2018-fig2.jpg}

{\tiny Brainstorm Consortium 2018 Science}
\end{column}
\end{columns}
\end{frame}

%\begin{frame}{Some studies of schizophrenia genetics}
%\footnotesize
%\begin{tabular}{lll}
%common variants & GWAS & 145 loci, 500 genes, MHC, ...\\
%rare \& de novo variants & WES, gene set enrichment & synaptic, LoF intolerant, ...\\
%somatic variants & WGS, gene set enrichment & \alert{our ongoing work}\\
%gene expression & allelic bias, imprinted genes & \alert{our 2018 Nat Comm article}\\
%\end{tabular}
%\end{frame}

\begin{frame}{Genomic studies on schizophrenia}
\begin{center}
\includegraphics[width=1.1\textwidth]{from-others/schizophrenia-Foley2017-Fig1.png}
\end{center}
\end{frame}

\begin{frame}{Our work on schizophrenia genetics}
\begin{enumerate}
\item Imprinted genes, allelic expression bias
\begin{itemize}
\item Jones et al, Nat Comm 2018
\end{itemize}	
\item Somatic variants, brain somatic mosaicism
\begin{itemize}
\item McConnell et al, 2017 Science
\item Wang et al 2021 Genome Biol, in press
\item Jones et al, unpublished
\end{itemize}
\end{enumerate}
\end{frame}

\section{Imprinted genes, allelic expression bias}

\begin{frame}<1>[label=imprinting]{Imprinted genes exhibit allelic bias}
\begin{columns}[t]
\begin{column}{0.5\textwidth}

\includegraphics[width=1.0\columnwidth]{from-others/renfree-2012-fig2.jpg}

{\tiny Renfree et al 2012 Philos Trans R Soc Lond B}
\end{column}
\begin{column}{0.5\textwidth}
Biological roles
\begin{itemize}
\item mother--child interaction
\item neurodevelopmental 
\item psychiatric?
\begin{itemize}
\item CNVs of imprinted genes
\item<2> inprinted brain hypothesis\\
kinship hypothesis
\end{itemize}
\end{itemize}
\end{column}
\end{columns}
\end{frame}

\begin{frame}{Rare CNVs with high penetrance for schizophrenia}
\includegraphics[width=1.0\textwidth]{from-others/sullivan-natrevgenet-2012-fig1b.jpg}
\end{frame}

\againframe<2>{imprinting}

\begin{frame}{The inprinted brain hypothesis}
\includegraphics[width=1.0\textwidth]{from-others/crespi-2008-fig3.png}

\tiny Crespi Biol Reviews 2008
\end{frame}

\begin{frame}{Study design}

\includegraphics[height=0.75\textheight]{by-me/commonmind-rna-seq/commonmind-rna-seq.pdf}
\end{frame}

\begin{frame}{Allelic bias in imprinted genes of the adult DLPFC}
\includegraphics[height=0.9\textheight]{figures/S-Dx-strip-1.png}
\end{frame}

\begin{frame}{The effect of age seemed to depend on the imprinted gene}
\includegraphics[height=0.9\textheight]{figures/S-Dx-age-1.png}
\end{frame}

\begin{frame}{Regression model based inference}
\footnotesize
\begin{eqnarray*}
\overbrace{t(S)}^{\text{allelic bias}} &\sim& \overbrace{(1 \,|\, \mathrm{Dx}) + (1 \,|\, \mathrm{Dx}:\mathrm{Gene})}^{\text{random effects of Dx}} + \\
&+&  \overbrace{(1 + \mathrm{Age} + \mathrm{Ancestry.1} \,|\, \mathrm{Gene})}^{\text{random, gene dependent, effects}} + \\
&+&  \overbrace{1 + \mathrm{Age} + \mathrm{Ancestry.1}}^{\text{fixed effects}}
+ \overbrace{\dotsb}^{\text{other variables}}
\end{eqnarray*}

\vfill

\begin{tabular}{|l|l|rl|}
\hline
Added term & Interpretation                             & \(\Delta\)AIC   & p-value \\
\hline
\((1\,|\,\mathrm{Dx})\) & effect of Dx (SCZ and ASD) & \(2.0\)         & \(1.0\) \\
\((1\,|\,\mathrm{Dx}:\mathrm{Gene})\) & gene dep.~effect of Dx  & \(0.4\)         & \(0.21\) \\
\hline
\((1\,|\,\mathrm{Gene})\) & effect of gene & \(-126.8\)      & \(8.5\times 10^{-28}\) \\
\((\mathrm{Age}\,|\,\mathrm{Gene})\) & gene dep.~effect of Age         & \(-18.9\)       & \(2.5\times 10^{-5}\) \\
\((\mathrm{Ancestry.1}\,|\,\mathrm{Gene})\) & gene dep.~effect of Ancestry & \(-71.2\) & \(4.6\times 10^{-16}\) \\
\hline
\(\mathrm{Age}\) & gene indep.~effect of Dx & \(1.3\) & \(0.39\) \\
\(\mathrm{Ancestry.1}\) & gene indep.~effect of Ancestry & \(0.6\) & \(0.24\) \\
\hline
\end{tabular}
\end{frame}

\begin{frame}{Summary}
\begin{itemize}
\item we found 30 imprinted genes in the DLPFC
\item determinants of allelic bias
	\begin{itemize}
		\item the gene \(\Rightarrow\) different epigenetic mechanisms
		\item age \(\Rightarrow\) epigenetic changes.  Causal to aging?
                \item ancestry \(\Rightarrow\) ``imprinting QTLs''
		\item \alert{not} schizophrenia \alert{nor} affective spectrum
			\(\Rightarrow\) no support for imprinted brain
			hypothesis
	\end{itemize}
\end{itemize}
\end{frame}

\section{Somatic variants, brain somatic mosaicism}

\begin{frame}{Somatic variants, brain somatic mosaicism}
\begin{columns}[t]
\begin{column}{0.3\textwidth}
\begin{center}
	\footnotesize no mosaicism
\end{center}

\includegraphics[width=1.0\columnwidth]{from-others/poduri-2013-fig1a.png}
\begin{center}
\end{center}
\end{column}

\begin{column}{0.3\columnwidth}
\begin{center}
	\footnotesize high freq.~mosaicism
\end{center}

\includegraphics[width=1.0\columnwidth]{from-others/poduri-2013-fig1e.png}
\end{column}
\begin{column}{0.3\columnwidth}
\begin{center}
	\footnotesize low freq.~mosaicism
\end{center}

\includegraphics[width=1.0\columnwidth]{from-others/poduri-2013-fig1g.png}
\end{column}
\end{columns}
{\tiny Poduri et al 2013 Science}
\end{frame}

\begin{frame}{Brain somatic mosaicism in disease}
\begin{columns}[t]
\begin{column}{0.6\textwidth}

\includegraphics[width=1.0\columnwidth]{from-others/bsmn-visual-summary.jpg}

{\tiny McConnell et al 2017 Science}
\end{column}

\begin{column}{0.4\textwidth}
\begin{itemize}
\item 7 NIH projects
	\begin{itemize}
	\item sequence brain cells/samples
	\item call somatic variants
	\end{itemize}
\item collaboration
	\begin{itemize}
	\item methods \& tools 
        \item validation
        \item functional experiments
	\end{itemize}
\item public data
	\begin{itemize}
		\item NIMH Data Archive
	\end{itemize}
\end{itemize}
\end{column}
\end{columns}
\end{frame}

\begin{frame}{Calling somatic variants}
\includegraphics[width=1.0\textwidth]{from-others/bsm-marker-fig1.jpg}

{\tiny McConnell et al 2017 Science}
\end{frame}

\begin{frame}{Somatic SNVs: traditional calling practices failed}
\includegraphics[height=0.8\textheight]{from-others/bsmn-best-practices-fig2ab.png}

{\tiny Wang et al 2021 Genome Biology}
\end{frame}

\begin{frame}{Validation results suggested new best practices}
\begin{center}
\includegraphics[height=0.7\textheight]{from-others/bsmn-best-practices-fig2c.png}
\end{center}

{\tiny Wang et al 2021 Genome Biology}
\end{frame}

\begin{frame}{Best practices variant calling pipeline}
\includegraphics[width=1.0\textwidth]{from-others/bsmn-pipeline-calling.png}
\begin{center}
\begin{itemize}
\item software tool
\begin{itemize}
        \item cloud computing: compute resotinyurceb auto-scale with data
	\item  \url{https://github.com/bsmn/bsmn-pipeline}
\end{itemize}
\item validation rate (specificity): 85\%
\item what about sensitivity?
\end{itemize}
\end{center}

{\tiny Wang et al 2021 Genome Biology}
\end{frame}

\begin{frame}[label=sensitivity]{Allele frequency distribution looks as
		expected}
\begin{columns}[t]
\begin{column}{0.35\textwidth}
\begin{center}
Sensitivity

\footnotesize
toy data, ``labeled'' variants
\end{center}

\includegraphics[width=1.0\textwidth]{from-others/bsmn-best-practices-fig4b.png}

{\tiny Wang et al 2021 Genome Biology}
\end{column}

\begin{column}{0.65\textwidth}
\begin{center}
Allele frequency distribution

\footnotesize
real data, ``unlabeled'' variants
\end{center}

\includegraphics<1>[width=1.0\textwidth]{../../notebook/2020-08-13-select-vars/named-figure/DP-AF-jointdensity-calls.pdf}

\includegraphics<2>[width=1.0\textwidth]{../../notebook/2020-08-13-select-vars/named-figure/DP-AF-jointdensity-conddensity.pdf}

{\tiny A.~Jones, unpublished }
\end{column}
\end{columns}
\end{frame}

\begin{frame}{Genome wide somatic mutation rate unaffected}
\begin{center}
\includegraphics[height=0.8\textheight]{../../notebook/2020-12-09-walsh-data/named-figure/ncalls-depth-chess-walsh.pdf}
\end{center}
\end{frame}

\begin{frame}{Annotation with functional categories}
\begin{itemize}
\item coding categories
\item regulatory
\item genesets
\end{itemize}
\end{frame}

\begin{frame}{Modeling number of calls in a geneset}
\begin{columns}[t]
\begin{column}{0.5\textwidth}

\includegraphics[width=1.0\columnwidth]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/scz_gwas_genes-log10_ncalls-poisson-fit.pdf}
\end{column}

\begin{column}{0.5\textwidth}

\includegraphics[width=1.0\columnwidth]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/scz_gwas_genes-ageOfDeath-binom-fit.pdf}
\end{column}
\end{columns}
\begin{center}
\end{center}
\end{frame}

\begin{frame}{Finding the best fitting model for SCZ GWAS genes}{Forward variable selection algorithm}
\includegraphics[width=1.0\textwidth]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/scz_gwas_genes-fw-varsel-binom-onlyIC.pdf}
\end{frame}

\begin{frame}{Best models for various categories}
\begin{center}
\tiny
\begin{tabular}{rrl}
\hline
category  & size\footnote{\(n_\mathrm{indiv} \,|\, n_{\mathrm{calls} \in
			\mathrm{category}} \ge 1\)} & best fitting model  \\
\hline
missense  & 0 & None  \\
stop\_gain  & 1 & None  \\
splice\_site  & 1 & None  \\
deleterious\_low\_confidence  & 4 & 1 \\
possibly\_damaging  & 4 & 1 + BaseQRankSum  \\
probably\_damaging  & 9 & 1 + DP  \\
deleterious  & 19 & 1 + BaseQRankSum + Dataset  \\
coding\_syn  & 38 & 1 + ageOfDeath + AF  \\
coding\_nonsyn  & 55 & 1 + Dataset + BaseQRankSum  \\
cpg\_island  & 57 & 1 + Dx + AF  \\
scz\_gwas\_genes  & 72 & 1 + Dx + ageOfDeath  \\
coding  & 80 & 1 + ageOfDeath  \\
tfbs  & 93 & 1 + Dx + AF  \\
phast  & 117 & 1 + DP + ageOfDeath + AF  \\
gerp  & 127 & 1 + Dx + BaseQRankSum + AF + DP  \\
\hline
\end{tabular}
\end{center}
\end{frame}

\begin{frame}{Fit improves with with size of category}
\includegraphics[width=1.0\textwidth]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/resid-binom.pdf}
\end{frame}

\begin{frame}{Enrichment of somatic SNVs in various categories}
\includegraphics[width=1.0\textwidth]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/scz-effect-pois-binom.pdf}

\includegraphics[width=1.0\textwidth]{../../notebook/2021-01-26-indiv-counts-modeling/named-figure/asd-effect-pois-binom.pdf}
\end{frame}

\end{document}

\begin{columns}[t]
\begin{column}{0.5\textwidth}

\end{column}

\begin{column}{0.5\textwidth}

\end{column}
\end{columns}
