---
title: Somatic SNVs and Indels by HaplotypeCaller
layout: default
featimg: "barchart-1.png"
---

TODO

## Introduction

From Alison Barton in Peter Park's lab we received a set of annotated SNV and indel variants called by the GATK [HaplotypeCaller] based on three samples from the same individual:

1. Fibro_Common_7 (fibroblasts)
1. Common_7_NeuN (sorted NeuN+ neurons from the dorsolateral prefrontal cortex DLFPC)
1. LBRCE-pfc-1b123 (tissue from the PFC)

This analysis summarizes the more than 4 million variants in the vcf file `Common_experiment_HC.vqsr.snp.indel.vcf.gz` provided by Alison on 2017-01-24.  In particular, those variants are in focus that are specific to either of the three samples since these represent some type of somatic mozaicism.  See the *Discussion* below on how somatic mozaicism may be interpreted given the present statistical approach (HaplotypeCaller).

Open [snv-2017-01-24.1-1000.tsv] to look at the first thousand variants and their annotations in an Excel-like program.  Similarly, open [snv-2017-01-24.Fibro_Common_7.recode.tsv] for somatic variants specific to the fibroblast sample (equivalently, brain specific variants).


[snv-2017-01-24.1-1000.tsv]: {{ site.baseurl}}/assets/projects/bsm/snv-2017-01-24.1-1000.tsv
[snv-2017-01-24.Fibro_Common_7.recode.tsv]: {{ site.baseurl}}/assets/projects/bsm/snv-2017-01-24.Fibro_Common_7.recode.tsv
[HaplotypeCaller]: https://software.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_gatk_tools_walkers_haplotypecaller_HaplotypeCaller.php
```{r echo=FALSE, message=FALSE, warning=FALSE}
library(lattice)
library(latticeExtra)
opts_chunk$set(dpi = 144)
opts_chunk$set(out.width = "700px")
opts_chunk$set(dev = c("png", "pdf"))
lattice.options(default.args = list(as.table = TRUE))
lattice.options(default.theme = "standard.theme")
```

## Definitions

**(all) kept variants**
* these were kept by the quality filter (see `PASS` in the `FILTER` field of the vcf file)
* they include specific and non-specific variants

**sample-specific variants**
* these are specific to either of the three samples
* thus they represent a certain type of somatic mozaicism (see *Discussion* below)

## Analysis

Extract the number of kept variants:

```{r engine='bash'}
# after running "runme.sh"
cd ../../results/common-experiment/snv/2017-01-24-alison
BN="snv-2017-01-24"
grep '^After filtering.*Sites' $${BN}.log | sed 's/.*kept \([[:digit:]]\+\) out of.*/\1/' > kept.vars
cat $${BN}.log
```

Read number of kept variants and also read the sites of singletons:

```{r}
kept.vars <- scan(file = "../../results/common-experiment/snv/2017-01-24-alison/kept.vars")
singletons <- read.delim("../../results/common-experiment/snv/2017-01-24-alison/snv-2017-01-24.singletons")
head(singletons)
```

Print first the number sample-specific variants:

```{r}
(singleton.freq <- with(singletons, table(INDV)))
```

Now print their fraction in all `r kept.vars` kept variants:

```{r}
singleton.freq / kept.vars
```

The same information as a barchart:

```{r barchart}
barchart(singleton.freq, main = "Sample-specific variants")
```

## Discussion
