\documentclass{beamer}
%\documentclass[handout]{beamer}

% language settings
%\usepackage{fontspec, polyglossia}
%\setdefaultlanguage{magyar}

% common packages
\usepackage{amsmath, multimedia, url, hyperref, color, multirow}
%\usepackage{graphicx}

% TikZ
\usepackage{tikz}
%\usetikzlibrary{arrows.meta, decorations.pathmorphing, decorations.pathreplacing, shapes.geometric,mindmap}
%\usetikzlibrary{shapes.geometric,fadings,bayesnet}

% beamer styles
\mode<presentation>{
\usetheme{Warsaw}
%\usetheme{Antibes}
\usecolortheme{beaver}
%\usecolortheme{seahorse}
%\usefonttheme{structureitalicserif}
\setbeamercovered{transparent}
}
\setbeamertemplate{blocks}[rounded][shadow=true]
\AtBeginSubsection[]{
  \begin{frame}<beamer>{Contents}
    \tableofcontents[currentsection,currentsubsection]
  \end{frame}
}
%\useoutertheme[]{tree}

% title, etc
\title{Somatic SNVs from WGS data from Schizophrenic Brains}
\subtitle{A subtitle may be shorter and more technical}
\author{Attila Jones}
\date{Mount Sinai School of Medicine}

\begin{document}

\maketitle

\section{Introduction}

\begin{frame}{Schizophrenia}
\begin{itemize}
\item 80\% heritability
\item extreme genetic complexity
	\begin{itemize}
		\item common variants: shared risk with other psychiatric
			disorders
	\end{itemize}
\item recurrent deletions% \(\approx 1\%\)
\end{itemize}
%\includegraphics[width=1.0\textwidth]{from-others/sullivan-natrevgenet-2012-fig1b.jpg}
\end{frame}


\begin{frame}{De novo variants in schizophrenia}
\begin{columns}[t]
\begin{column}{0.5\textwidth}
\begin{itemize}
        \item Fromer et al Nature 2014
        \item exome sequencing
	\item \(>600\) schizophrenia trios
\end{itemize}
\end{column}

\begin{column}{0.5\textwidth}

\includegraphics[scale=0.5]{from-others/fromer-denovo-figA.png}
\end{column}
\end{columns}
\end{frame}

\section{Our Study: Data}

\begin{frame}{Our study}
\begin{columns}[t]
\begin{column}{0.5\textwidth}
Questions
\begin{enumerate}
\item role of somatic variants in schizophrenia
\item relative risk: germline vs somatic variants	
\item WGS: coding, intronic and intergenic regions
\end{enumerate}
\end{column}

\begin{column}{0.5\textwidth}
Approach
\begin{enumerate}
\item SCZ and Control samples
	\begin{itemize}
		\item cortical neurons (NeuN+)
                \item cortical non-neurons (NeuN-)
                \item muscle
	\end{itemize}
\item bulk
\end{enumerate}
\end{column}
\end{columns}
\end{frame}


\begin{frame}{Samples and sequencing}
\includegraphics[width=1.0\textwidth]{from-others/bsm-marker-fig1.jpg}
\end{frame}

\begin{frame}{Samples from the CommonMind Consortium}
\begin{enumerate}
\item \(> 700\) brains
\item SCZ, AFF, Control
\end{enumerate}
\vfill
\begin{columns}[t]
\begin{column}{0.5\textwidth}
Data sets
\begin{itemize}
\item Clinical
\item RNAseq
\item ATACseq
\item Genotypes
\end{itemize}
\end{column}

\begin{column}{0.5\textwidth}
Analyses
\begin{itemize}
\item Ancestry
\item Differential Expression
\item Expression outlier
\item eQTL
\item Sherlock scores
\item Gene co-expression Networks
\end{itemize}
\end{column}
\end{columns}

\end{frame}

\begin{frame}{Cortical bulk WGS data sets in BSMN}
\includegraphics[width=1.0\textwidth]{../../notebook/2020-09-28-wgs-data-other-groups/named-figure/bsmn-case-control-ssizes.pdf}
\end{frame}

\begin{frame}{The bsmn-pipeline}
\begin{itemize}
\item Taejeong Bae \url{https://github.com/bsmn/bsmn-pipeline}
\item autoscaling AWS parallelCluster
\item somatic variant calling
\includegraphics[width=1.0\textwidth]{from-others/bsmn-pipeline-calling.png}
\end{itemize}
\end{frame}

\begin{frame}{Call sets}
\begin{columns}[t]
\begin{column}{0.7\textwidth}

\includegraphics[scale=0.5]{../../notebook/2020-06-10-annotate-explore-variants/named-figure/num-indiv-calls-Dx.pdf}
\end{column}

\begin{column}{0.5\textwidth}

\includegraphics[scale=0.5]{../../notebook/2020-06-10-annotate-explore-variants/named-figure/eCDF-calls.pdf}
\end{column}
\end{columns}
\end{frame}

\begin{frame}[label=numcalls-depth]{Number of calls per sample}
\includegraphics[scale=0.5]{../../notebook/2020-08-19-indiv-based-analysis/named-figure/numcalls-depth-Dx.pdf}
\end{frame}

\begin{frame}{Low frequency variants are missed at low depth}
\includegraphics[scale=0.5]{../../notebook/2020-08-13-select-vars/named-figure/DP-AF-jointdensity-conddensity.pdf}
\end{frame}

\section{Results}

\begin{frame}{Annotating somatic variants with features}
\includegraphics[width=1.0\textwidth]{../../results/2020-09-07-annotations/named-figure/MSSM_224-NeuN_pl-karyotype-of-genomic-consequences.png}
\end{frame}

\begin{frame}{Call features}
\begin{columns}[t]
\begin{column}{0.5\textwidth}
	Knowledge bases
\begin{itemize}
\item \url{www.szdb.org}
\begin{itemize}
\item GWAS loci
\item SCZ-CNV loci
\end{itemize}
\item Roadmap Epigenomics
\begin{itemize}
\item chromatin state in DLPC; 15 state HMM
\end{itemize}
\item \url{www.snp-nexus.org}
\end{itemize}
\end{column}

\begin{column}{0.5\textwidth}
	Data specific
\begin{itemize}
\item VCF fields: DP, AF, ...
\item clinical info: Dx, ethnicity, ...
\end{itemize}
\end{column}
\end{columns}
\end{frame}

\begin{frame}{Call features: SNPnexus}
	\tiny
	\begin{description}
		\item[Gene annotation systems]  Ensembl, RefSeq, UCSC, CCDS, coding, intronic, non-coding
                \item[Protein deleterious effects] SIFT and PolyPhen
                \item[Population data] gnomAD, 1000Genomes, HapMap
                \item[Regulatory elements] miRBASE, CpG islands, TarBase, microRNAs / snoRNAs / scaRNAs, Ensembl Regulatory Build, ENCODE Project, Roadmap Epigenomics 
                \item[Conservation scores] GERP++, PHAST 
                \item[Disease studies] GAD, COSMIC, NHGRI-GWAS, ClinVar
                \item[Non-coding scoring] fitCons, EIGEN, FATHMM GWAVA, DeepSEA, ReMM, CADD, FunSeq2 
                \item[Structural variations] Gain, Loss, Gain+Loss, Duplication, Deletion, Insertion, Complex, Inversion, Tandem duplication, Novel sequence insertion, Mobile element insertion, Sequence alteration 
                \item[Pathway Analysis] Reactome pathways 
                \item[Biological / Clinical Interpretation] Cancer genome interpreter 
	\end{description}
\end{frame}

\begin{frame}{Dx: no association with GWAS loci}
\includegraphics[scale=0.7]{../../notebook/2020-06-10-annotate-explore-variants/named-figure/gwas-Dx.pdf}
\end{frame}

%\begin{frame}{Dx: weak association with transcription start sites}
%\includegraphics[scale=0.7]{../../notebook/2020-06-10-annotate-explore-variants/named-figure/chromatin-state-DLPFC-Dx-10.pdf}
%\end{frame}

\begin{frame}{Correlation analysis}
\begin{columns}[t]
\begin{column}{0.4\textwidth}

\includegraphics[scale=0.35]{../../notebook/2020-09-07-annotations/named-figure/corr-Dx-snpnexus.pdf}
\end{column}

\begin{column}{0.5\textwidth}

\includegraphics[scale=0.35]{../../notebook/2020-09-07-annotations/named-figure/corr-mat-snpnexus.png}
\end{column}
\end{columns}
\end{frame}

\begin{frame}{Statistical modeling challenge}
\footnotesize
\begin{columns}[t]
\begin{column}{0.6\textwidth}
\begin{center}
Nested observations
\end{center}
\begin{center}
\begin{tabular}{cc|c|c}
& & clinical feat. & genomic feat. \\
sample & variant & \(X\) & \(Y\) \\
\hline
\multirow{4}{*}{\(s_1\)} & \(v1\) & \(x_1\) & \(y_1\) \\
& \(v2\) & \(x_1\) & \(y_2\) \\
& \(v3\) & \(x_1\) & \(y_3\) \\
& \(v4\) & \(x_1\) & \(y_4\) \\
\hline
\multirow{3}{*}{\(s_2\)} & \(v5\) & \(x_2\) & \(y_5\) \\
& \(v6\) & \(x_2\) & \(y_6\) \\
& \(v7\) & \(x_2\) & \(y_7\) \\
\end{tabular}
\end{center}
\end{column}

\begin{column}{0.4\textwidth}
\begin{center}
Correlated errors
\end{center}
\includegraphics[width=1.0\columnwidth]{from-others/regrression-correlated-error.png}
\end{column}
\end{columns}
\end{frame}

\begin{frame}{Addressing the challenge}
\begin{tabular}{cccc}
approach & formula & least squares & effects & issue \\
\hline
A & \(\mathrm{Dx} = X\beta_x + Y\beta_y + \epsilon\) & ordinary & fixed \\
B &\(\mathrm{Dx} = X\beta_x + Y\beta_y + \epsilon\) & generalized & fixed \\
C & \(\mathrm{Dx} = X\gamma_x + Y\beta_y + \epsilon\) & ordinary & mixed\\
D & \(\mathrm{\#calls} = X\gamma_x + Y'\beta'_y + \epsilon\) & ordinary & mixed\\
\end{tabular}
\end{frame}

\begin{frame}{Features selected by forward algorithm}
\tiny
\begin{tabular}{rcc}
	variable & \(P>|t|\) & direction \\
\hline
DP std & 0.007 & + \\
AF mean & 0.011 & - \\
culprit entropy & 0.011 & + \\
ReportedGender[T.Male] & 0.016 & - \\
AntipsychAtyp[T.1] & 0.033 & + \\
Anticonvulsants[T.1] & 0.043 & - \\
REF entropy & 0.054 & + \\
szdbCNVcount std & 0.065 & + \\
YearofAutopsy & 0.069 & - \\
Intercept & 0.114 & + \\
EV1 & 0.147 & - \\
DP mean & 0.183 & + \\
Alcohol[T.1] & 0.252 & - \\
Lithium[T.1] & 0.284 & - \\
EV3 & 0.292 & + \\
Ethnicity[T.Caucasian] & 0.498 & - \\
Ethnicity[T.Hispanic] & 0.897 & - \\
\end{tabular}
\end{frame}

\begin{frame}{Germline and somatic risk variants in cancer}
\includegraphics[height=0.8\textheight]{from-others/sommutcount-vs-polygenicriskscore.png}

\tiny
British Journal of Cancer volume 115, pages752–760(2016)
\end{frame}


\begin{columns}[t]
\begin{column}{0.5\textwidth}

\end{column}

\begin{column}{0.5\textwidth}

\end{column}
\end{columns}


\end{document}
